<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A Toolkit for Simulation-Based Inference in Population Genetics • demografr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A Toolkit for Simulation-Based Inference in Population Genetics">
<meta name="description" content="A toolkit for simulation-based inferences of demographic models in population genetics, such as Approximate Bayesian Computation, grid-based parameter sweeps, but also simulation-based parameter optimization. Population genetic models are specified using the functionality of the slendr R package, and summary statistics are computed using its interface to the tskit tree-sequence Python module. Together with features for specifying prior distributions and implicit parallelism, users can use demografr to build efficient, reproducible, and fully automated population genetic inference pipelines entirely in R, while leveraging the available tools for model selection and diagnostics of the abc package.">
<meta property="og:description" content="A toolkit for simulation-based inferences of demographic models in population genetics, such as Approximate Bayesian Computation, grid-based parameter sweeps, but also simulation-based parameter optimization. Population genetic models are specified using the functionality of the slendr R package, and summary statistics are computed using its interface to the tskit tree-sequence Python module. Together with features for specifying prior distributions and implicit parallelism, users can use demografr to build efficient, reproducible, and fully automated population genetic inference pipelines entirely in R, while leveraging the available tools for model selection and diagnostics of the abc package.">
<meta property="og:image" content="https://bodkan.net/demografr/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">demografr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html"><span class="fa fa-question-circle"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/vignette-01-basics.html">A basic ABC workflow</a></li>
    <li><a class="dropdown-item" href="articles/vignette-02-priors.html">Specifying prior distributions</a></li>
    <li><a class="dropdown-item" href="articles/vignette-03-grids.html">Support for grid simulations</a></li>
    <li><a class="dropdown-item" href="articles/vignette-04-parallelization.html">Parallelization options</a></li>
    <li><a class="dropdown-item" href="articles/vignette-05-custom.html">Custom inference using SLiM or Python</a></li>
    <li><a class="dropdown-item" href="articles/vignette-06-diagnostics.html">Model selection and other diagnostics</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bodkan/demografr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="demografr-simulation-based-inference-for-population-genetics">
<em>demografr</em>: Simulation-based inference for population genetics<a class="anchor" aria-label="anchor" href="#demografr-simulation-based-inference-for-population-genetics"></a>
</h1>
</div>
<!-- badges: start -->
<!--[![CRAN-version](https://www.r-pkg.org/badges/version/slendr)](https://cran.r-project.org/package=slendr) [![CRAN-downloads](https://cranlogs.r-pkg.org/badges/grand-total/slendr)](https://cran.r-project.org/package=slendr) -->

<!-- badges: end -->
<p>⚠️⚠️⚠️</p>
<p><strong>Please note that until a peer reviewed publication describing the <em>demografr</em> package is published, the software should be regarded as experimental and potentially unstable.</strong></p>
<p>⚠️️⚠️⚠️</p>
<p>The goal of <em>demografr</em> is to simplify and streamline the development of simulation-based inference pipelines in population genetics and evolutionary biology, such as <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation" class="external-link">Approximate Bayesian Computation</a> (ABC) or parameter grid inferences, and make them more reproducible. <em>demografr</em> also aims to make these inferences orders of magnitude faster and more efficient by leveraging the <a href="https://tskit.dev/learn/" class="external-link">tree sequences</a> as an internal data structure and computation engine.</p>
<p>Unlike traditional ABC and other simulation-based approaches, which generally involve custom-built pipelines and scripts for population genetic simulation and computation of summary statistics, <em>demografr</em> makes it possible to perform simulation, computation of summary statistics, and the inference itself entirely in R within a single reproducible analysis script. By eliminating the need to write custom simulation code and scripting for integration of various population genetic tools for computing summary statistics, it lowers the barrier for new users and facilitates reproducibility for everyone regardless of their level of experience by eliminating many common sources of bugs.</p>
<div class="section level3">
<h3 id="how-does-demografr-help-with-abc">How does <em>demografr</em> help with ABC?<a class="anchor" aria-label="anchor" href="#how-does-demografr-help-with-abc"></a>
</h3>
<p><em>demografr</em> streamlines every step of a typical ABC pipeline by leveraging the <a href="https://github.com/bodkan/slendr/" class="external-link"><em>slendr</em></a> framework as a building block for simulation and data analysis, making it possible to write complete simulation-based workflows entirely in R. Specifically:</p>
<ol style="list-style-type: decimal">
<li><p><em>slendr</em>’s intuitive, interactive <a href="https://slendr.net/articles/vignette-04-nonspatial-models.html" class="external-link">interface for definning population genetic models</a> makes it easy to encode even complex demographic models with only bare minimum of R knowledge needed.</p></li>
<li><p><em>demografr</em> makes it possible to encode prior distributions of parameters using familiar R interface resembling standard probabilistic statements, and provides an automated function which simulates ABC replicates drawing parameters from priors in a trivial, one-step manner.</p></li>
<li><p>Because <em>slendr</em> embraces <a href="https://tskit.dev/learn/" class="external-link">tree sequence</a> as its default internal data structure, most population genetic statistics can be computed directly on such tree sequences using R functions which are part of <em>slendr</em>’s statistical library. A tree sequence is never saved to disk and no conversion between file formats is required, which significantly speeds up every workflow.</p></li>
<li><p><em>demografr</em> facilitates tight integration with the powerful R package <a href="https://cran.r-project.org/package=abc" class="external-link"><em>abc</em></a> by automatically feeding it simulation data for inference and diagnostics.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of <em>demografr</em> from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"bodkan/demografr"</span><span class="op">)</span></span></code></pre></div>
<p>Note that this requires an R package <em>devtools</em>, which you can obtain simply by running <code>install.packages("devtools")</code>.</p>
<p>Because <em>demografr</em> is tightly linked to the <em>slendr</em> simulation package (in fact, new developments in <em>slendr</em> ale currently driven by requirements of <em>demografr</em>), you will also need the development version of <em>slendr</em> itself:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"bodkan/slendr"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="note-on-stability">Note on stability<a class="anchor" aria-label="anchor" href="#note-on-stability"></a>
</h3>
<p><em>demografr</em> is very much in an experimental stage at this point. Although inference of “standard” demographic models (i.e. estimating <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>, split times gene-flow parameters, etc.) already works very nicely, our long-term ambitions for the project are much higher and extend, for instance, towards inferences of spatial models or models which include selection. As such, please be aware that the interface might change on a rather short notice to accomodate features for estimating parameters of much more complex models.</p>
<p>If you want to follow updates on <em>demografr</em>, you can do this also on my <a href="https://bsky.app/profile/bodkan.bsky.social" class="external-link">social media</a> and by checking out the <a href="https://github.com/bodkan/demografr/blob/main/NEWS.md" class="external-link">changelog</a> from time to time.</p>
</div>
<div class="section level3">
<h3 id="testing-the-r-package-in-an-online-rstudio-session">Testing the R package in an online RStudio session<a class="anchor" aria-label="anchor" href="#testing-the-r-package-in-an-online-rstudio-session"></a>
</h3>
<p>You can open an RStudio session and test examples from the <a href="https://bodkan.net/demografr/articles/">vignettes</a> directly in your web browser by clicking this button (no installation is needed!):</p>
<p><a href="http://beta.mybinder.org/v2/gh/bodkan/demografr/main?urlpath=rstudio" class="external-link"><img src="http://mybinder.org/badge.svg" alt="Binder"></a></p>
<p><strong>In case the RStudio instance appears to be starting very slowly, please be patient (Binder is a freely available service with limited computational resources).</strong> If Binder crashes, try reloading the web page, which will restart the cloud session.</p>
<p>Once you get a browser-based RStudio session, you can navigate to the <code>vignettes/</code> directory and test the examples on your own!</p>
</div>
</div>
<div class="section level2">
<h2 id="an-example-abc-pipeline">An example ABC pipeline<a class="anchor" aria-label="anchor" href="#an-example-abc-pipeline"></a>
</h2>
<p><strong>Note:</strong> A much more detailed explanation of this toy example can be found in the <a href="https://bodkan.net/demografr/articles/vignette-01-basics.html">following vignette</a>.</p>
<p>Imagine that we sequenced genomes of individuals from populations “A”, “B”, “C”, and “D”.</p>
<p>Let’s also assume that we know that the populations are phylogenetically related in the following way, with an indicated gene-flow event at a certain time in the past, but we don’t know anything else (i.e., we have no idea about their <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>, their split times, or the proportion of gene flow):</p>
<p><img src="reference/figures/README-ape_tree-1.png"></p>
<p>After sequencing the genomes of individuals from these populations, we computed the nucleotide diversity in these populations, their pairwise genetic divergence, and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>4</mn></msub><annotation encoding="application/x-tex">f_4</annotation></semantics></math> statistic, and observed the following values (which we saved in two standard R data frames):</p>
<ol style="list-style-type: decimal">
<li>Nucleotide diversity in each population:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_diversity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/basics_diversity.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_diversity</span></span>
<span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt; 1   A 8.030512e-05</span></span>
<span><span class="co">#&gt; 2   B 3.288576e-05</span></span>
<span><span class="co">#&gt; 3   C 1.013804e-04</span></span>
<span><span class="co">#&gt; 4   D 8.910909e-05</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Pairwise divergence d_X_Y between populations X and Y:</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_divergence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/basics_divergence.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_divergence</span></span>
<span><span class="co">#&gt;   x y   divergence</span></span>
<span><span class="co">#&gt; 1 A B 0.0002378613</span></span>
<span><span class="co">#&gt; 2 A C 0.0002375761</span></span>
<span><span class="co">#&gt; 3 A D 0.0002379385</span></span>
<span><span class="co">#&gt; 4 B C 0.0001088217</span></span>
<span><span class="co">#&gt; 5 B D 0.0001157056</span></span>
<span><span class="co">#&gt; 6 C D 0.0001100633</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Value of the following <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>4</mn></msub><annotation encoding="application/x-tex">f_4</annotation></semantics></math>-statistic:</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_f4</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/basics_f4.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_f4</span></span>
<span><span class="co">#&gt;   W X Y Z            f4</span></span>
<span><span class="co">#&gt; 1 A B C D -3.262146e-06</span></span></code></pre></div>
<div class="section level3">
<h3 id="a-complete-abc-analysis-in-a-single-r-script">A complete ABC analysis in a single R script<a class="anchor" aria-label="anchor" href="#a-complete-abc-analysis-in-a-single-r-script"></a>
</h3>
<p>This is how we would use <em>demografr</em> to estimate the <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>, split times for all populations, as well as the rate of the indicated gene-flow event with Approximate Bayesian Computation in a single R script:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/demografr" class="external-link">demografr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span><span class="co"># running setup_env() first might be necessary to set up slendr's internal</span></span>
<span><span class="co"># simulation environment</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/init_env.html" class="external-link">init_env</a></span><span class="op">(</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># set up parallelization across all CPUs on the current machine</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># bind data frames with empirical summary statistics into a named list</span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  diversity  <span class="op">=</span> <span class="va">observed_diversity</span>,</span>
<span>  divergence <span class="op">=</span> <span class="va">observed_divergence</span>,</span>
<span>  f4         <span class="op">=</span> <span class="va">observed_f4</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># define a model generating function using the slendr interface</span></span>
<span><span class="co"># (each of the function parameters correspond to a parameter we want to infer)</span></span>
<span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ne_A</span>, <span class="va">Ne_B</span>, <span class="va">Ne_C</span>, <span class="va">Ne_D</span>, <span class="va">T_AB</span>, <span class="va">T_BC</span>, <span class="va">T_CD</span>, <span class="va">gf_BC</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"A"</span>, time <span class="op">=</span> <span class="fl">1</span>,    N <span class="op">=</span> <span class="va">Ne_A</span><span class="op">)</span></span>
<span>  <span class="va">B</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"B"</span>, time <span class="op">=</span> <span class="va">T_AB</span>, N <span class="op">=</span> <span class="va">Ne_B</span>, parent <span class="op">=</span> <span class="va">A</span><span class="op">)</span></span>
<span>  <span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"C"</span>, time <span class="op">=</span> <span class="va">T_BC</span>, N <span class="op">=</span> <span class="va">Ne_C</span>, parent <span class="op">=</span> <span class="va">B</span><span class="op">)</span></span>
<span>  <span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"D"</span>, time <span class="op">=</span> <span class="va">T_CD</span>, N <span class="op">=</span> <span class="va">Ne_D</span>, parent <span class="op">=</span> <span class="va">C</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">B</span>, to <span class="op">=</span> <span class="va">C</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="va">gf_BC</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A</span>, <span class="va">B</span>, <span class="va">C</span>, <span class="va">D</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>    generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    direction <span class="op">=</span> <span class="st">"forward"</span>, serialize <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/schedule_sampling.html" class="external-link">schedule_sampling</a></span><span class="op">(</span></span>
<span>    <span class="va">model</span>, times <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A</span>, <span class="fl">25</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">B</span>, <span class="fl">25</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">C</span>, <span class="fl">25</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">D</span>, <span class="fl">25</span><span class="op">)</span>,</span>
<span>    strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="co"># when a specific sampling schedule is to be used, both model and samples</span></span>
<span>  <span class="co"># must be returned by the function</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># setup priors for model parameters</span></span>
<span></span>
<span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">Ne_A</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">3000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_B</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>,  <span class="fl">1500</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_C</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_D</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">7000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">T_AB</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>,    <span class="fl">4000</span><span class="op">)</span>,</span>
<span>  <span class="va">T_BC</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">9000</span><span class="op">)</span>,</span>
<span>  <span class="va">T_CD</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">gf_BC</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># define summary functions to be computed on simulated data (must be of the</span></span>
<span><span class="co"># same format as the summary statistics computed on empirical data)</span></span>
<span></span>
<span><span class="va">compute_diversity</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_diversity.html" class="external-link">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">compute_divergence</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_divergence.html" class="external-link">ts_divergence</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">compute_f4</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="va">A</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"A"</span><span class="op">]</span>; <span class="va">B</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"B"</span><span class="op">]</span></span>
<span>  <span class="va">C</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"C"</span><span class="op">]</span>; <span class="va">D</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"D"</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_f4ratio.html" class="external-link">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">A</span>, <span class="va">B</span>, <span class="va">C</span>, <span class="va">D</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># the summary functions must be also bound to an R list named in the same</span></span>
<span><span class="co"># way as the empirical summary statistics</span></span>
<span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  diversity  <span class="op">=</span> <span class="va">compute_diversity</span>,</span>
<span>  divergence <span class="op">=</span> <span class="va">compute_divergence</span>,</span>
<span>  f4         <span class="op">=</span> <span class="va">compute_f4</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># validate the individual ABC components for correctness and consistency</span></span>
<span><span class="fu"><a href="reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>,</span>
<span>             sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># run ABC simulations</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span></span>
<span>  <span class="va">model</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  sequence_length <span class="op">=</span> <span class="fl">50e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># infer posterior distributions of parameters using the abc R package</span></span>
<span><span class="va">abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_abc.html">run_abc</a></span><span class="op">(</span><span class="va">data</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="analysing-posterior-distributions-of-parameters">Analysing posterior distributions of parameters<a class="anchor" aria-label="anchor" href="#analysing-posterior-distributions-of-parameters"></a>
</h2>
<p>After we run this R script, we end up with an object called <code>abc</code> here. This object contains the complete information about the results of our inference. In particular, it carries the posterior samples for our parameters of interest (<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math> of populations and their split times).</p>
<p>For instance, we can get a summary table of all parameter posteriors with the function <code><a href="reference/extract_summary.html">extract_summary()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span>
<span><span class="co">#&gt;                            Ne_A      Ne_B      Ne_C     Ne_D      T_AB     T_BC</span></span>
<span><span class="co">#&gt; Min.:                  1492.557  526.6231  6373.061 2254.591  859.5749 5131.473</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  1774.758  672.5853  7344.040 2895.795 1318.3159 5595.637</span></span>
<span><span class="co">#&gt; Weighted Median:       2032.148  848.1467  8553.144 3814.660 1934.0008 6136.188</span></span>
<span><span class="co">#&gt; Weighted Mean:         2021.722  838.5594  8678.066 3804.614 1954.3343 6112.305</span></span>
<span><span class="co">#&gt; Weighted Mode:         2054.933  861.3408  8428.777 3721.917 1933.3694 6230.341</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 2270.334 1003.8421 10162.320 4530.273 2522.5919 6611.315</span></span>
<span><span class="co">#&gt; Max.:                  2438.657 1047.0935 12703.479 5764.139 2650.7248 6851.800</span></span>
<span><span class="co">#&gt;                            T_CD       gf_BC</span></span>
<span><span class="co">#&gt; Min.:                  6752.669 -0.04681798</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  7125.349  0.02469992</span></span>
<span><span class="co">#&gt; Weighted Median:       7835.810  0.09867650</span></span>
<span><span class="co">#&gt; Weighted Mean:         7824.532  0.10282631</span></span>
<span><span class="co">#&gt; Weighted Mode:         7851.617  0.09804401</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 8426.377  0.18467366</span></span>
<span><span class="co">#&gt; Max.:                  8476.157  0.21801675</span></span></code></pre></div>
<p>We can also specify a subset of model parameters to select, or provide a regular expression for this subsetting:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"gf_BC"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                              gf_BC</span></span>
<span><span class="co">#&gt; Min.:                  -0.04681798</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:   0.02469992</span></span>
<span><span class="co">#&gt; Weighted Median:        0.09867650</span></span>
<span><span class="co">#&gt; Weighted Mean:          0.10282631</span></span>
<span><span class="co">#&gt; Weighted Mode:          0.09804401</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.:  0.18467366</span></span>
<span><span class="co">#&gt; Max.:                   0.21801675</span></span></code></pre></div>
<p>Of course, we can also visualize the posterior distributions. Rather than plotting many different distributions at once, let’s first check out the posterior distributions of inferred <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math> values:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Ne"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-posterior_Ne-1.png"></p>
<p>Similarly, we can take a look at the inferred posteriors of the split times:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-posterior_Tsplit-1.png"></p>
<p>And, finally, the rate of gene flow:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"gf"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-posterior_gf-1.png"></p>
<p>Additionally, we have the full diagnostic functionality of the <a href="https://cran.r-project.org/package=abc" class="external-link"><em>abc</em></a> R package at our disposal:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Ne_C"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-diagnostic_Ne-1.png"></p>
<p>Many diagnostic and model selection functions implemented by <em>abc</em> are also supported by <em>demografr</em>. For more information, see <a href="https://bodkan.net/demografr/articles/vignette-06-diagnostics.html">this vignette</a>.</p>
</div>
<div class="section level2">
<h2 id="additional-functionality">Additional functionality<a class="anchor" aria-label="anchor" href="#additional-functionality"></a>
</h2>
<p><em>demografr</em> also provides a couple of its own functions designed to make troubleshooting a little easier.</p>
<p>For instance, assuming we have <code>priors</code> set up as above, we can visualize the prior distribution(s) like this:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_prior.html">plot_prior</a></span><span class="op">(</span><span class="va">priors</span>, <span class="st">"Ne"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-prior_Ne-1.png"></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/bodkan/demografr/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing demografr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Martin Petr <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-4879-8421" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/bodkan/demografr/actions" class="external-link"><img src="https://github.com/bodkan/demografr/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="http://beta.mybinder.org/v2/gh/bodkan/demografr/main?urlpath=rstudio" class="external-link"><img src="http://mybinder.org/badge.svg" alt="Binder"></a></li>
<li><a href="https://app.codecov.io/github/bodkan/demografr?branch=main" class="external-link"><img src="https://codecov.io/gh/bodkan/demografr/branch/main/graph/badge.svg" alt="Coverage status"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
