<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fitting population genetic demographic models using Approximate Bayesian Computation • demografr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Fitting population genetic demographic models using Approximate Bayesian Computation">
<meta property="og:description" content="An Approximate Bayesian Computation (ABC) framework for fitting parameters
    of demographic models in population genetics.">
<meta property="og:image" content="https://bodkan.github.io/demografr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">demografr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/vignette-01-basic-abc.html">A basic ABC analysis</a>
    </li>
    <li>
      <a href="articles/vignette-02-scaffold-models.html">Defining scaffold models</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/dr_bodkan" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="demografr-a-simple-and-efficient-abc-toolkit-for-r">
<em>demografr</em>: A simple and efficient ABC toolkit for R<a class="anchor" aria-label="anchor" href="#demografr-a-simple-and-efficient-abc-toolkit-for-r"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p><img src="reference/figures/logo.png" align="right"></p>
<p>⚠️ <strong>This package is under active development and things often change on short notice.</strong> ⚠️</p>
<p>The goal of <em>demografr</em> is to simplify and streamline <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation" class="external-link">Approximate Bayesian Computation</a> (ABC) in population genetics and make it more reproducible. Additionally, <em>demografr</em> aims to make ABC orders of magnitude faster and more efficient by leveraging the <a href="https://tskit.dev/learn/" class="external-link">tree sequences</a> as an internal data structure and computation engine.</p>
<p>Unlike traditional ABC approaches, which generally involve custom-built pipelines and scripts for population genetic simulation and computation of summary statistics, <em>demografr</em> makes it possible to perform simulation, data analysis, and ABC inference itself entirely in R within a single reproducible analysis script. By eliminating the need to write custom simulation code and scripting for integration of various population genetic tools for computing summary statistics, it lowers the barrier to entry for new users and facilitates reproducibility for all users regardless of their level of experience by eliminating many common sources of bugs.</p>
<div class="section level3">
<h3 id="what-are-the-issues-with-standard-abc">What are the issues with standard ABC?<a class="anchor" aria-label="anchor" href="#what-are-the-issues-with-standard-abc"></a>
</h3>
<p>A traditional ABC analysis in population genetics generally involves:</p>
<ol style="list-style-type: decimal">
<li>Writing a simulation script tailor-made for the demographic model of species in question using tools such as <a href="http://home.uchicago.edu/rhudson1/source/mksamples.html" class="external-link">ms</a>, <a href="https://github.com/scrm/scrm/" class="external-link">scrm</a>, <a href="https://github.com/tskit-dev/msprime/" class="external-link">msprime</a>, or <a href="https://github.com/MesserLab/SLiM/" class="external-link">SLiM</a>. For most realistic demographic models this presents a significant obstacle unless one is well versed in software development.</li>
<li>Developing a pipeline which will draw model parameters from a set of priors, and simulate output data in an appropriate format. Saving output files to disk can be very slow, especially for large genome-scale simulations across thousands of simulation replicates.</li>
<li>Computing summary statistics on the simulated data using appropriate software (often a combination of ADMIXTOOLS, PLINK, vcftools, or various custom scripts). This usually requires conversion of simulated outputs to an appropriate file format. In addition to the cost of disk-access, computing summary statistics can be quite slow depending on the statistic or program in question.</li>
<li>Feeding (and often re-formatting) the outputs of summary statistics in the right format to use with the appropriate ABC inference engine, such as the R package <a href="https://cran.r-project.org/package=abc" class="external-link"><em>abc</em></a>.</li>
</ol>
<p>Given the range of disparate tools required for the steps above, traditional ABC pipelines end up being quite complex, require a large amount of programming work and, because of the specifics of each demographic model, the whole procedure is usually re-invented from scratch for each study.</p>
<p>All in all, <strong>an unnecessary amount of time spent on ABC analyses is dedicated to software development—programming, debugging, and data munging—work that would be better spent doing research</strong>.</p>
</div>
<div class="section level3">
<h3 id="how-does-demografr-help">How does <em>demografr</em> help?<a class="anchor" aria-label="anchor" href="#how-does-demografr-help"></a>
</h3>
<p><em>demografr</em> streamlines every step of a typical ABC pipeline by leveraging the <a href="https://github.com/bodkan/slendr/" class="external-link"><em>slendr</em></a> framework as a building block for simulation and data analysis, making it possible to perform an entire ABC analysis in R. Taking the steps above one by one, here is how <em>demografr</em> makes life easier:</p>
<ol style="list-style-type: decimal">
<li>
<em>slendr</em>’s intuitive, interactive <a href="https://www.slendr.net/articles/vignette-04-nonspatial-models.html" class="external-link">interface for definning population genetic models</a> makes it easy to encode even complex demographic models with only bare minimum of R knowledge needed.</li>
<li>
<em>demografr</em> makes it possible to encode prior distributions of parameters using familiar R interface resembling standard probabilistic statements, and provides an automated function which simulates ABC replicates drawing parameters from priors in a trivial, one-step manner.</li>
<li>Because <em>slendr</em>’s simulation output is a <a href="https://tskit.dev/learn/" class="external-link">tree sequence</a>, most population genetic statistics can be computed directly on such tree sequences using R functions which are part of <em>slendr</em>’s statistical library. A tree sequence is never saved to disk and no conversion between file formats is required.</li>
<li>
<em>demografr</em> facilitates tight integration with the powerful R package <a href="https://cran.r-project.org/package=abc" class="external-link"><em>abc</em></a> by automatically feeding its simulation data to the <em>abc</em> package for inference and analysis.</li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of <em>demografr</em> from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"bodkan/demografr"</span><span class="op">)</span></span></code></pre></div>
<p>Note that this requires an R package <em>devtools</em>, which you can obtain simply by running <code>install.packages("devtools")</code>.</p>
<div class="section level3">
<h3 id="note-on-stability">Note on stability<a class="anchor" aria-label="anchor" href="#note-on-stability"></a>
</h3>
<p><em>demografr</em> is very much in an early experimental stage at this point. Although ABC fitting of “standard” demographic models (i.e. estimating <span class="math inline"><em>N</em><sub><em>e</em></sub></span>, split times and gene-flow parameters for non-spatial models) already works very nicely, our long-term ambitions for the project are much higher. As such, please be aware that the interface might change significantly on a short notice to accomodate features for estimating parameters of more complex custom models such as spatial models etc.</p>
</div>
</div>
<div class="section level2">
<h2 id="an-example-abc-analysis">An example ABC analysis<a class="anchor" aria-label="anchor" href="#an-example-abc-analysis"></a>
</h2>
<p>Imagine that we sequenced genomes of individuals from populations “popA”, “popB”, “popC”, and “popD”.</p>
<p>Let’s also assume that we know that the three populations are phylogenetically related in the following way but we don’t know anything else (i.e., we have no idea about their <span class="math inline"><em>N</em><sub><em>e</em></sub></span> or split times):</p>
<p><img src="reference/figures/README-ape_tree-1.png" style="display: block; margin: auto;"></p>
<p>After sequencing the genomes of individuals from these populations, we computed the nucleotide diversity in these populations as well as their pairwise genetic divergence, and observed the following values which we saved in two standard R data frames:</p>
<ol style="list-style-type: decimal">
<li>Nucleotide diversity in each population:</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diversity_df</span></span>
<span><span class="co">#&gt;      stat        value</span></span>
<span><span class="co">#&gt; 1 pi_popA 3.962604e-05</span></span>
<span><span class="co">#&gt; 2 pi_popB 9.827821e-05</span></span>
<span><span class="co">#&gt; 3 pi_popC 1.501280e-04</span></span>
<span><span class="co">#&gt; 4 pi_popD 1.286567e-04</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Pairwise divergence d_X_Y between populations X and Y:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">divergence_df</span></span>
<span><span class="co">#&gt;          stat        value</span></span>
<span><span class="co">#&gt; 1 d_popA_popB 0.0001984640</span></span>
<span><span class="co">#&gt; 2 d_popA_popC 0.0001991969</span></span>
<span><span class="co">#&gt; 3 d_popA_popD 0.0001999640</span></span>
<span><span class="co">#&gt; 4 d_popB_popC 0.0001825852</span></span>
<span><span class="co">#&gt; 5 d_popB_popD 0.0001832798</span></span>
<span><span class="co">#&gt; 6 d_popC_popD 0.0001754839</span></span></code></pre></div>
<div class="section level3">
<h3 id="a-complete-abc-analysis-in-a-single-r-script">A complete ABC analysis in a single R script<a class="anchor" aria-label="anchor" href="#a-complete-abc-analysis-in-a-single-r-script"></a>
</h3>
<p>This is how we would use <em>demografr</em> to estimate the <span class="math inline"><em>N</em><sub><em>e</em></sub></span> and split times for all populations with Approximate Bayesian Computation in a single R script. You will find a more detailed explanation of each step below.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">demografr</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># bind data frames with empirical summary statistics into a named list</span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">diversity_df</span>, divergence <span class="op">=</span> <span class="va">divergence_df</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># create a "scaffold model" to be used for fitting parameters</span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/tree_model.html">tree_model</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="st">"(popA,(popB,(popC,popD)));"</span>, time_span <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># setup priors for model parameters</span></span>
<span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">Ne_popA</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_popB</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_popC</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_popD</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">Tsplit_popA_popB</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3000</span><span class="op">)</span>,</span>
<span>  <span class="va">Tsplit_popB_popC</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">6000</span><span class="op">)</span>,</span>
<span>  <span class="va">Tsplit_popC_popD</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">6000</span>, <span class="fl">9000</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># define summary functions to be computed on simulated data (must be of the</span></span>
<span><span class="co"># same format as the summary statistics computed on empirical data)</span></span>
<span><span class="va">compute_diversity</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_samples.html" class="external-link">ts_samples</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_diversity.html" class="external-link">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"pi_"</span>, <span class="va">set</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">stat</span>, value <span class="op">=</span> <span class="va">diversity</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">compute_divergence</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_samples.html" class="external-link">ts_samples</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_divergence.html" class="external-link">ts_divergence</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"d_%s_%s"</span>, <span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">stat</span>, value <span class="op">=</span> <span class="va">divergence</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># the summary functions must be also bound to an R list named in the same</span></span>
<span><span class="co"># way as the empirical summary statistics</span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">compute_diversity</span>, divergence <span class="op">=</span> <span class="va">compute_divergence</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># run ABC simulations</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span></span>
<span>  <span class="va">model</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  sequence_length <span class="op">=</span> <span class="fl">10e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#--------------------------------------------------------------------------------</span></span>
<span><span class="co"># infer posterior distributions of parameters</span></span>
<span><span class="va">abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/perform_abc.html">perform_abc</a></span><span class="op">(</span><span class="va">data</span>, tolerance <span class="op">=</span> <span class="fl">0.05</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="analysing-posterior-distributions-of-parameters">Analysing posterior distributions of parameters<a class="anchor" aria-label="anchor" href="#analysing-posterior-distributions-of-parameters"></a>
</h2>
<p>After we run this R script, we end up with an object called <code>abc</code> here. This object contains the complete information about the results of our inference. In particular, it carries the posterior samples for our parameters of interest (<span class="math inline"><em>N</em><sub><em>e</em></sub></span> of populations and their split times).</p>
<p>For instance, we can get a table of all posterior values with the function <code><a href="reference/extract_summary.html">extract_summary()</a></code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span>
<span><span class="co">#&gt;                          Ne_popA  Ne_popB   Ne_popC  Ne_popD Tsplit_popA_popB</span></span>
<span><span class="co">#&gt; Min.:                   523.7406 2083.255  7664.463 2166.191        -81.05097</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:   801.0286 2312.572  8208.358 2677.950        202.02356</span></span>
<span><span class="co">#&gt; Weighted Median:       1260.2080 2685.938  9086.437 3124.858       1650.63009</span></span>
<span><span class="co">#&gt; Weighted Mean:         1274.2332 2707.789  9068.257 3141.304       1622.06456</span></span>
<span><span class="co">#&gt; Weighted Mode:         1178.8326 2639.446  8804.845 3090.615       2400.59639</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 1766.0479 3137.912  9949.667 3618.092       2973.30404</span></span>
<span><span class="co">#&gt; Max.:                  2005.8523 3395.432 10640.415 3811.722       3340.87736</span></span>
<span><span class="co">#&gt;                        Tsplit_popB_popC Tsplit_popC_popD</span></span>
<span><span class="co">#&gt; Min.:                          2734.523         5789.999</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:          3090.184         6297.085</span></span>
<span><span class="co">#&gt; Weighted Median:               4453.029         7524.670</span></span>
<span><span class="co">#&gt; Weighted Mean:                 4471.333         7495.847</span></span>
<span><span class="co">#&gt; Weighted Mode:                 4213.312         7739.591</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.:         5906.066         8613.613</span></span>
<span><span class="co">#&gt; Max.:                          6249.404         9510.527</span></span></code></pre></div>
<p>We can also visualize the posterior distributions. Rather than plotting many different distributions at once, let’s first check out the posterior distributions of inferred <span class="math inline"><em>N</em><sub><em>e</em></sub></span> values:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span>, type <span class="op">=</span> <span class="st">"Ne"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-posterior_Ne-1.png"><!-- --></p>
<p>Similarly, we can take a look at the inferred posteriors of the split times:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span>, type <span class="op">=</span> <span class="st">"Tsplit"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 2 rows containing non-finite values (`stat_density()`).</span></span></code></pre></div>
<p><img src="reference/figures/README-posterior_Tsplit-1.png"><!-- --></p>
<p>Finally, we have the diagnostic functionality of the <a href="https://cran.r-project.org/web/packages/abc/vignettes/abcvignette.pdf" class="external-link"><em>abc</em></a> R package at our disposal:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Ne_popC"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-diagnostic_Ne-1.png"><!-- --></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing demografr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Martin Petr <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-4879-8421" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
