<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A basic ABC analysis • demografr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="A basic ABC analysis">
<meta property="og:description" content="demografr">
<meta property="og:image" content="https://bodkan.net/demografr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">demografr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette-01-basics.html">A basic ABC analysis</a>
    </li>
    <li>
      <a href="../articles/vignette-02-priors.html">Specifying prior distributions</a>
    </li>
    <li>
      <a href="../articles/vignette-03-grids.html">Support for grid simulations</a>
    </li>
    <li>
      <a href="../articles/vignette-04-parallelization.html">Parallelization options</a>
    </li>
    <li>
      <a href="../articles/vignette-05-custom-engines.html">Custom SLiM or Python simulations</a>
    </li>
    <li>
      <a href="../articles/vignette-06-diagnostics.html">ABC diagnostics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/demografr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://fosstodon.org/bodkan" class="external-link">
    <span class="fa fa-mastodon"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A basic ABC analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/demografr/blob/HEAD/vignettes/vignette-01-basics.Rmd" class="external-link"><code>vignettes/vignette-01-basics.Rmd</code></a></small>
      <div class="hidden name"><code>vignette-01-basics.Rmd</code></div>

    </div>

    
    
<p>⚠️⚠️⚠️</p>
<p><strong>Note:</strong> The <em>demografr</em> R package is still under active development. As a result, its documentation is in a draft stage at best. Apologies for typos, inconsistencies, and other issues.</p>
<p>⚠️⚠️⚠️</p>
<p>This vignette contains an expanded version of the basic ABC inference example from the <a href="https://bodkan.net/demografr/">homepage</a> of <em>demografr</em>. It walks through the process of setting up a <em>demografr</em> ABC pipeline step by step.</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Imagine that we sequenced genomes of individuals from populations “popA”, “popB”, “popC”, and “popD”.</p>
<p>Let’s also assume that we know that the three populations are phylogenetically related in the following way but we don’t know anything else (i.e., we have no idea about the values of <span class="math inline">\(N_e\)</span>, split times, or gene-flow rates):</p>
<p><img src="vignette-01-basics_files/figure-html/ape_tree-1.png" width="400" style="display: block; margin: auto;"></p>
<p>After sequencing the genomes of individuals from these populations, we computed nucleotide diversity in these populations as well as their pairwise genetic divergence and a one <span class="math inline">\(f_4\)</span> statistic, observing the following values of these summary statistics (which we saved in standard R data frames—perhaps saved by software we used for computing these from empirical sequence data):</p>
<ol style="list-style-type: decimal">
<li>Nucleotide diversity in each population:</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_diversity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/observed_diversity.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_diversity</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;    set    diversity</span></span>
<span><span class="co">#&gt; 1 popA 8.037847e-05</span></span>
<span><span class="co">#&gt; 2 popB 3.242467e-05</span></span>
<span><span class="co">#&gt; 3 popC 1.021123e-04</span></span>
<span><span class="co">#&gt; 4 popD 8.968777e-05</span></span></code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Pairwise divergence d_X_Y between populations X and Y:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_divergence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/observed_divergence.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_divergence</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;      x    y   divergence</span></span>
<span><span class="co">#&gt; 1 popA popB 0.0002387594</span></span>
<span><span class="co">#&gt; 2 popA popC 0.0002391843</span></span>
<span><span class="co">#&gt; 3 popA popD 0.0002389617</span></span>
<span><span class="co">#&gt; 4 popB popC 0.0001089125</span></span>
<span><span class="co">#&gt; 5 popB popD 0.0001155571</span></span>
<span><span class="co">#&gt; 6 popC popD 0.0001105323</span></span></code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Value of the following <span class="math inline">\(f_4\)</span>-statistic:</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_f4</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/observed_f4.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_f4</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;      W    X    Y    Z            f4</span></span>
<span><span class="co">#&gt; 1 popA popB popC popD -3.433205e-06</span></span>
<span><span class="co">#&gt; 2 popA popC popB popD -7.125812e-07</span></span>
<span><span class="co">#&gt; 3 popA popD popB popC  2.720624e-06</span></span></code></pre>
<p><strong>Now let’s develop a simple ABC pipeline which will infer the posterior distributions of two sets of parameters we are interested in: <span class="math inline">\(N_e\)</span> of each population lineage, as well as split times between our populations of interest.</strong></p>
</div>
<div class="section level2">
<h2 id="developing-an-abc-pipeline">Developing an ABC pipeline<a class="anchor" aria-label="anchor" href="#developing-an-abc-pipeline"></a>
</h2>
<p>Let’s begin by loading <em>demografr</em> together with the R package <a href="https://www.slendr.net/" class="external-link"><em>slendr</em></a> on which <em>demografr</em> relies on for building and simulating demographic models.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/demografr" class="external-link">demografr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we also have to activate slendr's internal environment for tree sequences</span></span>
<span><span class="co"># simulation and analysis</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/init_env.html" class="external-link">init_env</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># setup parallelization across all CPUs</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multicore</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For the purpose of the ABC analysis below, we will bind all statistics in an R list, naming them appropriately. The names of each statistic (here “diversity” and “divergence”) have meaning and are quite important for later steps:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  diversity  <span class="op">=</span> <span class="va">observed_diversity</span>,</span>
<span>  divergence <span class="op">=</span> <span class="va">observed_divergence</span>,</span>
<span>  f4         <span class="op">=</span> <span class="va">observed_f4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="setting-up-a-scaffold-model">1. Setting up a “scaffold” model<a class="anchor" aria-label="anchor" href="#setting-up-a-scaffold-model"></a>
</h3>
<p>The first step in a <em>demografr</em> ABC analysis is setting up a “scaffold” model—a <em>slendr</em> function which will return a compiled <em>slendr</em> model as its output, and which will accept the model parameters in form of normal R function arguments. In our simple examples, we will define the following function:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ne_A</span>, <span class="va">Ne_B</span>, <span class="va">Ne_C</span>, <span class="va">Ne_D</span>, <span class="va">T_AB</span>, <span class="va">T_BC</span>, <span class="va">T_CD</span>, <span class="va">gf_BC</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">popA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popA"</span>, time <span class="op">=</span> <span class="fl">1</span>,    N <span class="op">=</span> <span class="va">Ne_A</span><span class="op">)</span></span>
<span>  <span class="va">popB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popB"</span>, time <span class="op">=</span> <span class="va">T_AB</span>, N <span class="op">=</span> <span class="va">Ne_B</span>, parent <span class="op">=</span> <span class="va">popA</span><span class="op">)</span></span>
<span>  <span class="va">popC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popC"</span>, time <span class="op">=</span> <span class="va">T_BC</span>, N <span class="op">=</span> <span class="va">Ne_C</span>, parent <span class="op">=</span> <span class="va">popB</span><span class="op">)</span></span>
<span>  <span class="va">popD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popD"</span>, time <span class="op">=</span> <span class="va">T_CD</span>, N <span class="op">=</span> <span class="va">Ne_D</span>, parent <span class="op">=</span> <span class="va">popC</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">popB</span>, to <span class="op">=</span> <span class="va">popC</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="va">gf_BC</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popA</span>, <span class="va">popB</span>, <span class="va">popC</span>, <span class="va">popD</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>    generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    direction <span class="op">=</span> <span class="st">"forward"</span>, serialize <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/schedule_sampling.html" class="external-link">schedule_sampling</a></span><span class="op">(</span></span>
<span>    <span class="va">model</span>, times <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popA</span>, <span class="fl">25</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popB</span>, <span class="fl">25</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popC</span>, <span class="fl">25</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popD</span>, <span class="fl">25</span><span class="op">)</span>,</span>
<span>    strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="co"># when sampling schedule is used, both model and samples must be</span></span>
<span>  <span class="co"># returned by the model function</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="setting-up-priors">2. Setting up priors<a class="anchor" aria-label="anchor" href="#setting-up-priors"></a>
</h3>
<p>We are interested in estimating the <span class="math inline">\(N_e\)</span> of all populations and their split times. <em>demografr</em> makes this very easy using a familiar symbolic formula syntax in R:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">Ne_A</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_B</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_C</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_D</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">T_AB</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>,    <span class="fl">4000</span><span class="op">)</span>,</span>
<span>  <span class="va">T_BC</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">9000</span><span class="op">)</span>,</span>
<span>  <span class="va">T_CD</span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">gf_BC</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>In an ABC simulation step below, the formulas are used to draw the values of each parameter from specified distributions (in this case, all uniform distributions across a wide range of parameter values).</p>
<p>**For more detail into how <em>demografr</em>’s prior sampling formulas work (and why), take a look <a href="vignette-02-priors.html">here</a>.</p>
</div>
<div class="section level3">
<h3 id="defining-summary-statistics">3. Defining summary statistics<a class="anchor" aria-label="anchor" href="#defining-summary-statistics"></a>
</h3>
<p>Each run of a <em>demografr</em> ABC simulation internally produces a tree sequence as an output. Because tree sequence represents an efficient, succint representation of the complete genealogical history of a set of samples, it is possible to compute population genetic statistics directly on the tree sequence without having to first save each simulation output to disk for computation in different software. Thanks to <em>slendr</em>’s library of <a href="https://www.slendr.net/reference/index.html#tree-sequence-statistics" class="external-link">tree-sequence functions</a> serving as an R interface to the <a href="https://tskit.dev/tskit/docs/stable/stats.html" class="external-link"><em>tskit</em> module</a>, you can specify summary statistics to be computed for ABC using plain and simple R code.</p>
<p>In our example, because we computed nucleotide diversity and pairwise divergence in the individuals sequenced from populations “p1”, “p2”, and “p3”, we will define the following functions. Crucially, when run on a tree-sequence object, they will produce an output data frame in the format analogous to the empirical statistics shown in data frames <code>diversity</code> and <code>divergence</code> above:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compute_diversity</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_diversity.html" class="external-link">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">compute_divergence</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_divergence.html" class="external-link">ts_divergence</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">compute_f4</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="va">A</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"popA"</span><span class="op">]</span>; <span class="va">B</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"popB"</span><span class="op">]</span></span>
<span>  <span class="va">C</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"popC"</span><span class="op">]</span>; <span class="va">D</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"popD"</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_f4ratio.html" class="external-link">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">A</span>, <span class="va">B</span>, <span class="va">C</span>, <span class="va">D</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_f4ratio.html" class="external-link">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">A</span>, <span class="va">C</span>, <span class="va">B</span>, <span class="va">D</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_f4ratio.html" class="external-link">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">A</span>, <span class="va">D</span>, <span class="va">B</span>, <span class="va">C</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  diversity  <span class="op">=</span> <span class="va">compute_diversity</span>,</span>
<span>  divergence <span class="op">=</span> <span class="va">compute_divergence</span>,</span>
<span>  f4         <span class="op">=</span> <span class="va">compute_f4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Crucially, the outputs of these summary functions <em>must</em> match the format of the observed summary statistics (i.e., the data frames produced must have the same format). This minor inconvenience during ABC setup saves us the headache of having to match values of statistics between observed and simulated data during ABC inference itself.</p>
</div>
<div class="section level3">
<h3 id="abc-simulations">4. ABC simulations<a class="anchor" aria-label="anchor" href="#abc-simulations"></a>
</h3>
<p>Having defined the scaffold model, a set of priors for our parameters of interest (<span class="math inline">\(N_e\)</span> and split times), as well as two summary statistic functions, we can plug all this information into the function <code><a href="../reference/simulate_abc.html">simulate_abc()</a></code>.</p>
<p>Before we run a potentially computationally costly simulations, it is a good idea to validate the ABC components we have so far assembled using the function <code><a href="../reference/validate_abc.html">validate_abc()</a></code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; Testing sampling of each prior parameter:</span></span>
<span><span class="co">#&gt;   * Ne_A ✅</span></span>
<span><span class="co">#&gt;   * Ne_B ✅</span></span>
<span><span class="co">#&gt;   * Ne_C ✅</span></span>
<span><span class="co">#&gt;   * Ne_D ✅</span></span>
<span><span class="co">#&gt;   * T_AB ✅</span></span>
<span><span class="co">#&gt;   * T_BC ✅</span></span>
<span><span class="co">#&gt;   * T_CD ✅</span></span>
<span><span class="co">#&gt;   * gf_BC ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; The model is a slendr function</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the return statement of the model function... ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the presence of required function arguments... ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Simulating tree sequence from the given model... ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Computing user-defined summary functions:</span></span>
<span><span class="co">#&gt;   * diversity ✅</span></span>
<span><span class="co">#&gt;   * divergence ✅</span></span>
<span><span class="co">#&gt;   * f4 ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the format of simulated summary statistics:</span></span>
<span><span class="co">#&gt;   * diversity ✅  </span></span>
<span><span class="co">#&gt;   * divergence ✅  </span></span>
<span><span class="co">#&gt;   * f4 ✅  </span></span>
<span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; No issues have been found in the ABC setup!</span></span></code></pre>
<p>Having verified that all model components are set up correctly, we can proceed to the ABC simulations themselves, using <em>demografr</em>’s function <code><a href="../reference/simulate_abc.html">simulate_abc()</a></code>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span></span>
<span>  <span class="va">model</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  sequence_length <span class="op">=</span> <span class="fl">20e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>The total runtime for the ABC simulations was 0 hours 24 minutes 49 seconds parallelized across 96 CPUs.</strong></p>
<p>At this point we have generated summary statistics for simulations of models using parameters drawn from our priors. In the next step, we can finally do inference of our parameters.</p>
</div>
<div class="section level3">
<h3 id="abc-inference">5. ABC inference<a class="anchor" aria-label="anchor" href="#abc-inference"></a>
</h3>
<p>Having all the information about observed and simulated data bound in a single R object <code>data</code>, we can finally perform the ABC inference. <em>demografr</em> includes a convenient function <code><a href="../reference/perform_abc.html">perform_abc()</a></code> which reformats the simulated and observed data in a format required by the R package <a href="https://cran.r-project.org/package=abc" class="external-link"><em>abc</em></a> and internally calls its function <code>abc()</code>.</p>
<p>Note that <code>perform_abc</code> is just convenience wrapper around the <code>abc()</code> function in the package <em>abc</em><code>, saving us a little work juggling the necessary matrices manually. As such, all parameters of the function</code>abc()<code>can be provided to</code>perform_abc()`, which will then pass them on appropriately.</p>
</div>
<div class="section level3">
<h3 id="posterior-predictive-check">6. Posterior predictive check<a class="anchor" aria-label="anchor" href="#posterior-predictive-check"></a>
</h3>
<p>Before we proceed with inferring values of the model parameters from their posterior distributions, we should check whether our model can actually produce summary statistics which match the statistics observed in the empirical data.</p>
<p>In order to do this, we can use <em>demografr</em>’s <code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code> method which accepts two mandatory function arguments: the first is the abc object generated by the <code><a href="../reference/perform_abc.html">perform_abc()</a></code> function (and contains, among other things, the information about the inferred posteriors), and the number of draws to take from the multidimensional posterior distribution of the parameters:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># because we set up a parallelization plan() above, the predictions are</span></span>
<span><span class="co"># computed in parallel across all available CPUs</span></span>
<span><span class="va">predictions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">abc</span>, samples <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning: Please note that sampling adjusted parameters from the posterior</span></span>
<span><span class="co">#&gt; can lead to invalid models (i.e. inconsistent orders of split times or</span></span>
<span><span class="co">#&gt; nonsensical values of parameters in general). If you get an error, you</span></span>
<span><span class="co">#&gt; can either sample unadjusted parameters  by setting `posterior = "unadj"`</span></span>
<span><span class="co">#&gt; or instruct the prediction function to skip invalid models by setting</span></span>
<span><span class="co">#&gt; `strict = FALSE`.</span></span></code></pre>
<pre><code><span><span class="co">#&gt; Out of the total 7 simulations, 1000 runs resulted in an error.</span></span>
<span><span class="co">#&gt; The most likely explanation for this is that some parameter combinations</span></span>
<span><span class="co">#&gt; lead to an invalid model (such as inconsistent order of split times).</span></span></code></pre>
<p>If we take a closer look at the produced result, we see that it’s a data frame object with several special columns (so-called <a href="https://dcl-prog.stanford.edu/list-columns.html" class="external-link">“list columns”</a>). Those columns contain nested data frames with the values of tree-sequence summary statistics computed for each combination of the parameter values. Because this data has the same format as the output of the function <code><a href="../reference/simulate_grid.html">simulate_grid()</a></code> (in fact, internally it <em>is</em> produced by this function), see the vignette on <a href="vignette-03-grids.html">grid simulations</a> for more detail how to analyse this.</p>
<p>In order to perform a detailed analysis of the posterior predictive check results for individual (or all) summary statistics, you can use the function <code><a href="../reference/extract_prediction.html">extract_prediction()</a></code>. A convenient alternative to check the results visually is provided by the function <code><a href="../reference/plot_prediction.html">plot_prediction()</a></code>. For instance, we can take a look at the distributions of divergences produced from the parameter posterior like this:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predictions</span>, <span class="st">"diversity"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basics_files/figure-html/postpred_diversity-1.png" width="640" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predictions</span>, <span class="st">"divergence"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basics_files/figure-html/postpred_divergence-1.png" width="640" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predictions</span>, <span class="st">"f4"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basics_files/figure-html/postpred_f4-1.png" width="640" style="display: block; margin: auto;"></p>
<p><strong>Nice! It appears that the posterior from our simplistic model captures the observed summary statistics (dashed vertical lines) quite well.</strong></p>
<p>(To make evaluation a little bit easier, there’s an option <code>file =</code> which instructs <code>plot_prediction</code> to save a figure to disk instead of plotting it using a graphical device—useful for work on remote servers!).</p>
<p><strong>For more details on possible downstream validation and troubleshooting analyses, please see <a href="vignette-06-diagnostics.html">this vignette</a>.</strong></p>
</div>
<div class="section level3">
<h3 id="posterior-analysis">7. Posterior analysis<a class="anchor" aria-label="anchor" href="#posterior-analysis"></a>
</h3>
<div class="section level4">
<h4 id="extracting-posterior-summary-tables">Extracting posterior summary tables<a class="anchor" aria-label="anchor" href="#extracting-posterior-summary-tables"></a>
</h4>
<p>Now that we have the ABC output object ready and have more or less convinced ourselves that our model can capture the summary statistics we’ve chosen via posterior predictive checks, we can proceed with parameter inference.</p>
<p>First, we can get a data frame with summary statistics of the posterior distributions of our parameters. For instance, we can easily read the maximum a posteriori probability (MAP) of the parameters in the row labelled “Mode:”:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                             Ne_A     Ne_B      Ne_C     Ne_D     T_AB     T_BC</span></span>
<span><span class="co">#&gt; Min.:                   929.0679 670.3812  2966.659 1790.060 1192.360 4988.493</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  1201.0043 711.6460  4075.275 2181.297 1561.180 5259.732</span></span>
<span><span class="co">#&gt; Weighted Median:       1990.5578 799.0462  7366.514 3714.350 2003.515 5678.190</span></span>
<span><span class="co">#&gt; Weighted Mean:         1967.7919 802.3221  7075.016 3578.215 2003.738 5655.562</span></span>
<span><span class="co">#&gt; Weighted Mode:         1954.3619 757.0564  7620.487 3809.785 1939.879 5752.531</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 2837.7028 924.5835 10858.144 4842.347 2326.478 5941.146</span></span>
<span><span class="co">#&gt; Max.:                  3144.9674 927.3249 11696.130 5228.341 2569.966 6036.115</span></span>
<span><span class="co">#&gt;                            T_CD       gf_BC</span></span>
<span><span class="co">#&gt; Min.:                  6863.233 -0.02455945</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  7294.305  0.02674433</span></span>
<span><span class="co">#&gt; Weighted Median:       8127.699  0.16684306</span></span>
<span><span class="co">#&gt; Weighted Mean:         8196.626  0.18184086</span></span>
<span><span class="co">#&gt; Weighted Mode:         8052.760  0.12030457</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 9387.034  0.33763567</span></span>
<span><span class="co">#&gt; Max.:                  9387.034  0.35898225</span></span></code></pre>
<p>Because large tables can get a little hard to read, it is possible to subset to only a specific <em>type</em> of parameter:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Ne"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                             Ne_A     Ne_B      Ne_C     Ne_D</span></span>
<span><span class="co">#&gt; Min.:                   929.0679 670.3812  2966.659 1790.060</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  1201.0043 711.6460  4075.275 2181.297</span></span>
<span><span class="co">#&gt; Weighted Median:       1990.5578 799.0462  7366.514 3714.350</span></span>
<span><span class="co">#&gt; Weighted Mean:         1967.7919 802.3221  7075.016 3578.215</span></span>
<span><span class="co">#&gt; Weighted Mode:         1954.3619 757.0564  7620.487 3809.785</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 2837.7028 924.5835 10858.144 4842.347</span></span>
<span><span class="co">#&gt; Max.:                  3144.9674 927.3249 11696.130 5228.341</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                            T_AB     T_BC     T_CD</span></span>
<span><span class="co">#&gt; Min.:                  1192.360 4988.493 6863.233</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  1561.180 5259.732 7294.305</span></span>
<span><span class="co">#&gt; Weighted Median:       2003.515 5678.190 8127.699</span></span>
<span><span class="co">#&gt; Weighted Mean:         2003.738 5655.562 8196.626</span></span>
<span><span class="co">#&gt; Weighted Mode:         1939.879 5752.531 8052.760</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 2326.478 5941.146 9387.034</span></span>
<span><span class="co">#&gt; Max.:                  2569.966 6036.115 9387.034</span></span></code></pre>
<p>Alternatively, we can also extract the posterior summary for a single parameter like this:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Ne_D"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                            Ne_D</span></span>
<span><span class="co">#&gt; Min.:                  1790.060</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  2181.297</span></span>
<span><span class="co">#&gt; Weighted Median:       3714.350</span></span>
<span><span class="co">#&gt; Weighted Mean:         3578.215</span></span>
<span><span class="co">#&gt; Weighted Mode:         3809.785</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 4842.347</span></span>
<span><span class="co">#&gt; Max.:                  5228.341</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="visualizing-posterior-distributions-of-parameters">Visualizing posterior distributions of parameters<a class="anchor" aria-label="anchor" href="#visualizing-posterior-distributions-of-parameters"></a>
</h4>
<p>Because a chart is always more informative than a table, we can easily get a visualization of our posteriors using the function <code><a href="../reference/plot_posterior.html">plot_posterior()</a></code>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Ne"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basics_files/figure-html/posterior_Ne-1.png" width="640" style="display: block; margin: auto;"></p>
<p>Excellent! It looks like we got really nice and informative posterior distributions of <span class="math inline">\(N_e\)</span> values!</p>
<p>In contrast, it looks like the posterior distributions for split times are not as informative compared to the priors that we specified:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basics_files/figure-html/posterior_Tsplit-1.png" width="640" style="display: block; margin: auto;"></p>
<p>Because the internals of <em>demografr</em> ABC objects are represented by standard objects created by the <em>abc</em> package, we have many of the standard diagnostics functions of the <em>abc</em> R package at our disposal. For instance, we can use the standard function <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> to verify that the posterior (red line) for one of the split times matches the prior (dashed line), suggesting that the data we provided (nucleotide diversity and pairwise divergence) are not sufficient statistics to capture enough information about population divergences.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"T_BC"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basics_files/figure-html/diagnostic_Tsplit-1.png" width="800" style="display: block; margin: auto;"></p>
<p>In contrast, we can see that there most definitely is sufficient information encoded in the summary statistics to tell us quite a bit about the <span class="math inline">\(N_e\)</span> of our populations:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Ne_B"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basics_files/figure-html/diagnostic_Ne-1.png" width="800" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"gf"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basics_files/figure-html/diagnostic_hist_Ne-1.png" width="800" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a232fffe35f6bc68735023bb34d1fee2',
    indexName: 'bodkan',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
