<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A basic ABC analysis • demografr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A basic ABC analysis">
<meta property="og:description" content="demografr">
<meta property="og:image" content="https://bodkan.github.io/demografr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">demografr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette-01-basic-abc.html">A basic ABC analysis</a>
    </li>
    <li>
      <a href="../articles/vignette-02-scaffold-models.html">Defining scaffold models</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/dr_bodkan" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A basic ABC analysis</h1>
            
      
      
      <div class="hidden name"><code>vignette-01-basic-abc.Rmd</code></div>

    </div>

    
    
<p>This vignette contains an expanded version of the basic ABC inference
example from the homepage of <em>demografr</em>.</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Imagine that we sequenced genomes of individuals from populations
“popA”, “popB”, “popC”, and “popD”.</p>
<p>Let’s also assume that we know that the three populations are
phylogenetically related in the following way but we don’t know anything
else (i.e., we have no idea about their <span class="math inline">\(N_e\)</span> or split times):</p>
<p><img src="vignette-01-basic-abc_files/figure-html/ape_tree-1.png" width="480" style="display: block; margin: auto;"></p>
<p>After sequencing the genomes of individuals from these populations,
we computed nucleotide diversity in these populations as well as their
pairwise genetic divergence, and observed the following values which we
saved in two standard R data frames:</p>
<ol style="list-style-type: decimal">
<li>Nucleotide diversity in each population:</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diversity_df</span></span>
<span><span class="co">#&gt;      stat        value</span></span>
<span><span class="co">#&gt; 1 pi_popA 3.962604e-05</span></span>
<span><span class="co">#&gt; 2 pi_popB 9.827821e-05</span></span>
<span><span class="co">#&gt; 3 pi_popC 1.501280e-04</span></span>
<span><span class="co">#&gt; 4 pi_popD 1.286567e-04</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Pairwise divergence d_X_Y between populations X and Y:</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">divergence_df</span></span>
<span><span class="co">#&gt;          stat        value</span></span>
<span><span class="co">#&gt; 1 d_popA_popB 0.0001984640</span></span>
<span><span class="co">#&gt; 2 d_popA_popC 0.0001991969</span></span>
<span><span class="co">#&gt; 3 d_popA_popD 0.0001999640</span></span>
<span><span class="co">#&gt; 4 d_popB_popC 0.0001825852</span></span>
<span><span class="co">#&gt; 5 d_popB_popD 0.0001832798</span></span>
<span><span class="co">#&gt; 6 d_popC_popD 0.0001754839</span></span></code></pre></div>
<p><strong>Now let’s develop a simple ABC pipeline which will infer the
posterior distributions of two sets of parameters we are interested in:
<span class="math inline">\(N_e\)</span> of each population lineage, as
well as split times between our populations of interest.</strong></p>
</div>
<div class="section level2">
<h2 id="developing-the-abc-pipeline">Developing the ABC pipeline<a class="anchor" aria-label="anchor" href="#developing-the-abc-pipeline"></a>
</h2>
<p>Let’s begin by loading <em>demografr</em> together with two helper R
packages. First is <em>dplyr</em>, which we use for convenient table
munging, the other is <a href="https://www.slendr.net/" class="external-link"><em>slendr</em></a> which is required for
building and simulating demographic models.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">demografr</span><span class="op">)</span></span></code></pre></div>
<p>For the purpose of the ABC analysis below, we will bind all
statistics in an R list, naming them appropriately. The names of each
statistic (here “diversity” and “divergence”) have meaning and are quite
important for later steps:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">diversity_df</span>, divergence <span class="op">=</span> <span class="va">divergence_df</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="setting-up-a-scaffold-model">Setting up a “scaffold” model<a class="anchor" aria-label="anchor" href="#setting-up-a-scaffold-model"></a>
</h3>
<p>The first step in a <em>demografr</em> ABC analysis is setting up a
“scaffold” model for which we will estimate the posterior parameters of
interest.</p>
<p>One way to do this is by building a normal <em>slendr</em> model (for
instance, a model such as <a href="https://www.slendr.net/articles/vignette-04-nonspatial-models.html" class="external-link">this</a>).
However, for simpler models like ours, it can be easier to input the
scaffold as a standard phylogenetic tree (here we use the function
<code><a href="../reference/tree_model.html">tree_model()</a></code> to input a tree in the <a href="https://en.wikipedia.org/wiki/Newick_format" class="external-link">Newick
format</a>).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tree_model.html">tree_model</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="st">"(popA,(popB,(popC,popD)));"</span>, time_span <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
<p>(Note that the parameter <code>time_span</code> indicates how much
evolutionary time does of our model cover, in units of generations.
Support for arbitrary units such as years will be supported soon.)</p>
<p>Because the <code>model</code> object contains a standard
<em>slendr</em> demographic model, we can inspect it as such, to make
sure everything is set up correctly:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span></span>
<span><span class="co">#&gt; slendr 'model' object </span></span>
<span><span class="co">#&gt; --------------------- </span></span>
<span><span class="co">#&gt; populations: popA, popB, popC, popD </span></span>
<span><span class="co">#&gt; geneflow events: [no geneflow]</span></span>
<span><span class="co">#&gt; generation time: 1 </span></span>
<span><span class="co">#&gt; time direction: forward </span></span>
<span><span class="co">#&gt; total running length: 10000 model time units</span></span>
<span><span class="co">#&gt; model type: non-spatial</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; non-serialized slendr model</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/plot_model.html" class="external-link">plot_model</a></span><span class="op">(</span><span class="va">model</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basic-abc_files/figure-html/slendr_model-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Note that we don’t pay attention to the split times or population
sizes of this model because we will be fitting those parameters with the
ABC procedure below—the model we just constructed is really just a
<em>scaffold</em> capturing some prior information we have about the
phylogenetic relationship between populations, the values of other model
parameters (<span class="math inline">\(N_e\)</span>, split times) are
arbitrary.</p>
<p>That said, we have the option to fix some aspects of our model by
building and fine-tuning the model using standard <em>slendr</em>
features for defining models rather than importing the model as a
phylogenetic tree (see <a href="https://www.slendr.net/articles/vignette-04-nonspatial-models.html" class="external-link">this</a>
<em>slendr</em> vignette).</p>
</div>
<div class="section level3">
<h3 id="setting-up-priors">Setting up priors<a class="anchor" aria-label="anchor" href="#setting-up-priors"></a>
</h3>
<p>We are interested in estimating the <span class="math inline">\(N_e\)</span> of all populations and their split
times. <em>demografr</em> makes this very easy using a familiar symbolic
formula syntax in R:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">Ne_popA</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_popB</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_popC</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span>  <span class="va">Ne_popD</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">Tsplit_popA_popB</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3000</span><span class="op">)</span>,</span>
<span>  <span class="va">Tsplit_popB_popC</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">6000</span><span class="op">)</span>,</span>
<span>  <span class="va">Tsplit_popC_popD</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">6000</span>, <span class="fl">9000</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>In an ABC simulation step below, the formulas are used to draw the
values of each parameter from specified distributions (in this case, all
uniform distributions across a wide range of parameter values).</p>
</div>
<div class="section level3">
<h3 id="defining-summary-functions">Defining summary functions<a class="anchor" aria-label="anchor" href="#defining-summary-functions"></a>
</h3>
<p>Each run of a <em>demografr</em> ABC simulation internally produces a
tree sequence as an output. Because tree sequence represents an
efficient, succint representation of the complete genealogical history
of a set of samples, it is possible to compute population genetic
statistics directly on the tree sequence without having to first save
each simulation output to disk for computation in different software.
Thanks to <em>slendr</em>’s library of <a href="https://www.slendr.net/reference/index.html#tree-sequence-statistics" class="external-link">tree-sequence
functions</a> serving as an R interface to the <a href="https://tskit.dev/tskit/docs/stable/stats.html" class="external-link"><em>tskit</em>
module</a>, you can specify summary statistics to be computed for ABC
using plain and simple R code.</p>
<p>In our example, because we computed nucleotide diversity and pairwise
divergence in the individuals sequenced from populations “p1”, “p2”, and
“p3”, we will define the following functions. Crucially, when run on a
tree-sequence object, they will produce an output data frame in the
format analogous to the empirical statistics shown in data frames
<code>diversity</code> and <code>divergence</code> above:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compute_diversity</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_samples.html" class="external-link">ts_samples</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_diversity.html" class="external-link">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"pi_"</span>, <span class="va">set</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">stat</span>, value <span class="op">=</span> <span class="va">diversity</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">compute_divergence</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_samples.html" class="external-link">ts_samples</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_divergence.html" class="external-link">ts_divergence</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"d_%s_%s"</span>, <span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">stat</span>, value <span class="op">=</span> <span class="va">divergence</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">compute_diversity</span>, divergence <span class="op">=</span> <span class="va">compute_divergence</span><span class="op">)</span></span></code></pre></div>
<p>Crucially, the outputs of these summary functions <em>must</em> match
the format of the observed summary statistics (i.e., the data frames
produced must have the same format). This minor inconvenience during ABC
setup saves us the headache of having to match values of statistics
between observed and simulated data during ABC inference itself.</p>
</div>
<div class="section level3">
<h3 id="abc-simulations">ABC simulations<a class="anchor" aria-label="anchor" href="#abc-simulations"></a>
</h3>
<p>Having defined the scaffold model, a set of priors for our parameters
of interest (<span class="math inline">\(N_e\)</span> and split times),
as well as two summary statistic functions, we can plug all this
information into the function <code><a href="../reference/simulate_abc.html">simulate_abc()</a></code>.</p>
<p>Before we run a potentially computationally costly simulations, it is
a good idea to validate the ABC components we have so far assembled
using the function <code><a href="../reference/validate_abc.html">validate_abc()</a></code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span><span class="op">)</span></span>
<span><span class="co">#&gt; ============================================================</span></span>
<span><span class="co">#&gt; Standard slendr model provided as a scaffold</span></span>
<span><span class="co">#&gt; ============================================================</span></span>
<span><span class="co">#&gt; Checking the correct syntax of population names...  ✓</span></span>
<span><span class="co">#&gt; Checking the correctness of prior parameter names...  ✓</span></span>
<span><span class="co">#&gt; ------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Testing sampling of each prior parameter:</span></span>
<span><span class="co">#&gt;   * Ne_popA ✓</span></span>
<span><span class="co">#&gt;   * Ne_popB ✓</span></span>
<span><span class="co">#&gt;   * Ne_popC ✓</span></span>
<span><span class="co">#&gt;   * Ne_popD ✓</span></span>
<span><span class="co">#&gt;   * Tsplit_popA_popB ✓</span></span>
<span><span class="co">#&gt;   * Tsplit_popB_popC ✓</span></span>
<span><span class="co">#&gt;   * Tsplit_popC_popD ✓</span></span>
<span><span class="co">#&gt; ------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Modifying the scaffold model with sampled prior values... ✓</span></span>
<span><span class="co">#&gt; ------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Simulating a tree sequence from the constructed model... ✓</span></span>
<span><span class="co">#&gt; ------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Computing user-defined summary functions:</span></span>
<span><span class="co">#&gt;   * diversity ✓</span></span>
<span><span class="co">#&gt;   * divergence ✓</span></span>
<span><span class="co">#&gt; ------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the format of simulated summary statistics:</span></span>
<span><span class="co">#&gt;   * diversity ✓</span></span>
<span><span class="co">#&gt;   * divergence ✓</span></span>
<span><span class="co">#&gt; ============================================================</span></span>
<span><span class="co">#&gt; No issues have been found in the ABC setup!</span></span></code></pre></div>
<p>Having verified that all model components are set up correctly, we
can proceed to the ABC simulations themselves, using
<em>demografr</em>’s function <code><a href="../reference/simulate_abc.html">simulate_abc()</a></code>:</p>
<p><!-- #eval=!file.exists("inst/extdata/data.rds")} --></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span></span>
<span>  <span class="va">model</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  sequence_length <span class="op">=</span> <span class="fl">10e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Because running the full ABC on thousands of simulation replicates
would take too long, we will cheat a little bit here and load the ABC
simulation data directly from a file distributed with the package:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/01_data.rds"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>At this point we have generated summary statistics for simulations of
models using parameters drawn from our priors. In the next step, we can
finally do inference of our parameters.</p>
</div>
<div class="section level3">
<h3 id="abc-inference">ABC inference<a class="anchor" aria-label="anchor" href="#abc-inference"></a>
</h3>
<p>Having all the information about observed and simulated data bound in
a single R object <code>abc_data</code>, we can finally perform the ABC
inference. <em>demografr</em> includes a convenient function
<code><a href="../reference/perform_abc.html">perform_abc()</a></code> which reformats the simulated and observed
data in a format required by the R package <a href="https://cran.r-project.org/package=abc" class="external-link"><em>abc</em></a> and
internally calls <a href="https://cran.r-project.org/web/packages/abc/abc.pdf" class="external-link">the function
<code>abc()</code></a> of that package.</p>
<p>Note that <code>perform_abc</code> is just convenience wrapper around
the <code>abc()</code> function in the package
<em>abc</em><code>, saving us a little work juggling the necessary matrices manually. As such, all parameters of the function</code>abc()<code>can be provided to</code>perform_abc()`,
which will then pass them on appropriately.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform_abc.html">perform_abc</a></span><span class="op">(</span><span class="va">data</span>, tolerance <span class="op">=</span> <span class="fl">0.05</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: All parameters are "none" transformed.</span></span>
<span><span class="co">#&gt; 12345678910</span></span>
<span><span class="co">#&gt; 12345678910</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="inspecting-the-posteriors">Inspecting the posteriors<a class="anchor" aria-label="anchor" href="#inspecting-the-posteriors"></a>
</h3>
<div class="section level4">
<h4 id="extracting-posterior-summary-tables">Extracting posterior summary tables<a class="anchor" aria-label="anchor" href="#extracting-posterior-summary-tables"></a>
</h4>
<p>Now that we have the ABC output object ready, we can get a data frame
with summary statistics of the posterior distributions of our
parameters. For instance, we can easily read the maximum a posteriori
probability (MAP) of the parameters in the row labelled “Weighted
Mode:”:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span>
<span><span class="co">#&gt;                          Ne_popA  Ne_popB   Ne_popC  Ne_popD Tsplit_popA_popB</span></span>
<span><span class="co">#&gt; Min.:                   878.8014 1780.394  7630.407 1918.176        -30.42729</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  1051.9572 2228.788  8094.783 2558.055        122.77767</span></span>
<span><span class="co">#&gt; Weighted Median:       1440.8680 2660.364  8810.756 3153.811       1549.07714</span></span>
<span><span class="co">#&gt; Weighted Mean:         1501.4152 2681.990  8816.794 3178.400       1522.50911</span></span>
<span><span class="co">#&gt; Weighted Mode:         1181.3398 2641.655  8794.326 3034.724       2322.57187</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 2176.6469 3236.581  9657.190 3875.698       2881.71222</span></span>
<span><span class="co">#&gt; Max.:                  2556.7822 3777.090 10147.025 4247.706       3291.55469</span></span>
<span><span class="co">#&gt;                        Tsplit_popB_popC Tsplit_popC_popD</span></span>
<span><span class="co">#&gt; Min.:                          2940.514         5872.130</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:          3171.989         6459.630</span></span>
<span><span class="co">#&gt; Weighted Median:               4497.437         7824.452</span></span>
<span><span class="co">#&gt; Weighted Mean:                 4529.729         7779.635</span></span>
<span><span class="co">#&gt; Weighted Mode:                 4346.200         8022.213</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.:         5922.436         9096.519</span></span>
<span><span class="co">#&gt; Max.:                          6188.116         9953.798</span></span></code></pre></div>
<p>Because large tables can get a little hard to read, it is possible to
subset to only a specific <em>type</em> of parameter:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span>, type <span class="op">=</span> <span class="st">"Ne"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                          Ne_popA  Ne_popB   Ne_popC  Ne_popD</span></span>
<span><span class="co">#&gt; Min.:                   878.8014 1780.394  7630.407 1918.176</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  1051.9572 2228.788  8094.783 2558.055</span></span>
<span><span class="co">#&gt; Weighted Median:       1440.8680 2660.364  8810.756 3153.811</span></span>
<span><span class="co">#&gt; Weighted Mean:         1501.4152 2681.990  8816.794 3178.400</span></span>
<span><span class="co">#&gt; Weighted Mode:         1181.3398 2641.655  8794.326 3034.724</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 2176.6469 3236.581  9657.190 3875.698</span></span>
<span><span class="co">#&gt; Max.:                  2556.7822 3777.090 10147.025 4247.706</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span>, type <span class="op">=</span> <span class="st">"Tsplit"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                        Tsplit_popA_popB Tsplit_popB_popC Tsplit_popC_popD</span></span>
<span><span class="co">#&gt; Min.:                         -30.42729         2940.514         5872.130</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:         122.77767         3171.989         6459.630</span></span>
<span><span class="co">#&gt; Weighted Median:             1549.07714         4497.437         7824.452</span></span>
<span><span class="co">#&gt; Weighted Mean:               1522.50911         4529.729         7779.635</span></span>
<span><span class="co">#&gt; Weighted Mode:               2322.57187         4346.200         8022.213</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.:       2881.71222         5922.436         9096.519</span></span>
<span><span class="co">#&gt; Max.:                        3291.55469         6188.116         9953.798</span></span></code></pre></div>
<p>Alternatively, we can also extract the posterior summary for a single
parameter like this:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Ne_popD"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         Ne_popD</span></span>
<span><span class="co">#&gt; Min.:                  1918.176</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  2558.055</span></span>
<span><span class="co">#&gt; Weighted Median:       3153.811</span></span>
<span><span class="co">#&gt; Weighted Mean:         3178.400</span></span>
<span><span class="co">#&gt; Weighted Mode:         3034.724</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 3875.698</span></span>
<span><span class="co">#&gt; Max.:                  4247.706</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualizing-posterior-distributions-of-parameters">Visualizing posterior distributions of parameters<a class="anchor" aria-label="anchor" href="#visualizing-posterior-distributions-of-parameters"></a>
</h4>
<p>Because a chart is always more informative than a table, we can
easily get a visualization of our posteriors using the function
<code><a href="../reference/plot_posterior.html">plot_posterior()</a></code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span>, type <span class="op">=</span> <span class="st">"Ne"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basic-abc_files/figure-html/posterior_Ne-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Excellent! It looks like we got really nice and informative posterior
distributions of <span class="math inline">\(N_e\)</span> values!</p>
<p>In contrast, it looks like the posterior distributions for split
times are not as informative compared to the priors that we
specified:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span>, type <span class="op">=</span> <span class="st">"Tsplit"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 4 rows containing non-finite values (`stat_density()`).</span></span></code></pre></div>
<p><img src="vignette-01-basic-abc_files/figure-html/posterior_Tsplit-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Because the internals of <em>demografr</em> ABC objects are
represented by standard objects created by the <em>abc</em> package, we
have many of the standard diagnostics functions of the <em>abc</em> R
package at our disposal. For instance, we can use the standard function
<code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> to verify that the posterior (red line) for one of
the split times matches the prior (dashed line), suggesting that the
data we provided (nucleotide diversity and pairwise divergence) are not
sufficient statistics to capture enough information about population
divergences.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Tsplit_popB_popC"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basic-abc_files/figure-html/diagnostic_Tsplit-1.png" width="960" style="display: block; margin: auto;"></p>
<p>In contrast, we can see that there most definitely is sufficient
information encoded in the summary statistics to tell us quite a bit
about the <span class="math inline">\(N_e\)</span> of our
populations:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">abc</span>, param <span class="op">=</span> <span class="st">"Ne_popB"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-01-basic-abc_files/figure-html/diagnostic_Ne-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
