<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Custom inference using SLiM or Python • demografr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Custom inference using SLiM or Python">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">demografr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>

  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>

    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>

    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette-01-basics.html">A basic ABC workflow</a>
    </li>
    <li>
      <a href="../articles/vignette-02-priors.html">Specifying prior distributions</a>
    </li>
    <li>
      <a href="../articles/vignette-03-grids.html">Support for grid simulations</a>
    </li>
    <li>
      <a href="../articles/vignette-04-parallelization.html">Parallelization options</a>
    </li>
    <li>
      <a href="../articles/vignette-05-custom.html">Custom inference using SLiM or Python</a>
    </li>
    <li>
      <a href="../articles/vignette-06-diagnostics.html">Model selection and other diagnostics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/demografr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>

    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Custom inference using SLiM or Python</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/demografr/blob/HEAD/vignettes/vignette-05-custom.Rmd" class="external-link"><code>vignettes/vignette-05-custom.Rmd</code></a></small>
      <div class="hidden name"><code>vignette-05-custom.Rmd</code></div>

    </div>

    
    
<p>⚠️⚠️⚠️</p>
<p><strong>The <em>demografr</em> R package is still under active
development!</strong></p>
<p>⚠️️⚠️⚠️</p>
<p>By default, <em>demografr</em> uses the <a href="https://github.com/bodkan/slendr" class="external-link"><em>slendr</em></a> package for
defining models and simulating data from them. But what if you need to
do inference using your own scripts? Perhaps <em>slendr</em>’s
opinionated interface doesn’t allow you to do all that you need to do,
such as using some of the powerful simulation features and options of
“raw” <a href="https://messerlab.org/slim/" class="external-link">SLiM</a> or <a href="https://tskit.dev/msprime/docs/stable/intro.html" class="external-link"><em>msprime</em></a>?
Alternatively, perhaps you don’t want to compute summary statistics
(just) with <em>slendr</em>’s tree-sequence interface to <em>tskit</em>
(you might want to compute statistics using other software), or you want
to compute statistics with fully customized Python functions. This
vignette explains how you can use any standard SLiM or <em>msprime</em>
script as a simulation engine in a standard <em>demografr</em> pipeline
without <em>slendr</em>, as well as how to utilize customized,
non-<em>slendr</em> summary statistic functions.</p>
<p>First, let’s load <em>demografr</em> itself and also <em>slendr</em>
(which, in this vignette, will serve only for working with simulated
tree sequences, but not simulations themselves).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/demografr" class="external-link">demografr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/init_env.html" class="external-link">init_env</a></span><span class="op">(</span>quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="toy-inference-problem">Toy inference problem<a class="anchor" aria-label="anchor" href="#toy-inference-problem"></a>
</h2>
<p>Suppose that we intent to use ABC to infer the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>
of a constant-sized population given that we observed the following
value of nucleotide diversity:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_diversity</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   set   diversity</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop   0.000<span style="text-decoration: underline;">039</span>4</span></span></code></pre>
<p>We want to infer the posterior distribution of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>.
In this vignette, we’re going to show how to accomplish first not via
the normal <em>slendr</em> interface (as shown <a href="vignette-01-basics.html">here</a>) but using a simple SLiM or
Python script. In a later section, we’re also going to use this example
to demonstrate how to compute population genetic summary statistic (like
nucleotide diversity) using external software.</p>
<p><strong>We acknowledge that this is a completely trivial example, not
really worth spending so much effort on doing an ABC for.</strong> The
example was chosen because it runs fast and demonstrates all the
features of <em>demografr</em> that you would use regardless of the
complexity of your model.</p>
</div>
<div class="section level2">
<h2 id="using-custom-scripts-as-simulation-engines">Using custom scripts as simulation engines<a class="anchor" aria-label="anchor" href="#using-custom-scripts-as-simulation-engines"></a>
</h2>
<p>In order to be able to use a custom SLiM or msprime script with
<em>demografr</em>, the script must conform to a couple of rules:</p>
<div class="section level3">
<h3 id="it-must-be-runnable-on-the-command-line-as-any-other-command-line-script">1. It must be runnable on the command-line as any other command-line
script<a class="anchor" aria-label="anchor" href="#it-must-be-runnable-on-the-command-line-as-any-other-command-line-script"></a>
</h3>
<ul>
<li><p><strong>For an <em>msprime</em> script</strong>, this means
something like
<code>python &lt;your script&gt;.py &lt;command-line arguments&gt;</code>.</p></li>
<li><p><strong>For a SLiM script</strong>, this means something like
<code>slim &lt;command-line arguments&gt; &lt;your script&gt;.slim</code>.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="it-must-accept---path-argument-pointing-to-a-directory-where-it-will-save-output-files">2. It must accept <code>--path</code> argument pointing to a
directory where it will save output files<a class="anchor" aria-label="anchor" href="#it-must-accept---path-argument-pointing-to-a-directory-where-it-will-save-output-files"></a>
</h3>
<p>This is analogous to the <code>path =</code> argument of
<em>slendr</em> functions <code><a href="https://rdrr.io/pkg/slendr/man/slim.html" class="external-link">slim()</a></code> and
<code><a href="https://rdrr.io/pkg/slendr/man/msprime.html" class="external-link">msprime()</a></code>.</p>
<p><strong>For an <em>msprime</em> script</strong>, this parameter can
be specified via the Python built-in module <code>argparse</code>, and
provided on the command-line as
<code>--path &lt;path to directory&gt;</code>. In the script itself, if
you use the <code>argparse</code> module and have thus the values of
provided arguments (for instance) in an <code>args</code> object, you
can then refer to these arguments as <code>args.path</code>, etc. Given
that you’re reading this, I assume you know what the above information
means, but here’s a <a href="https://docs.python.org/3/library/argparse.html" class="external-link">link to the
relevant section</a> of the Python documentation for completeness.</p>
<p><strong>For a SLiM script</strong>, this parameter can be specified
via SLiM’s standard way of supplying command-line arguments as
<code>-d "path='&lt;path to a directory&gt;'"</code>. Importantly, note
SLiM’s format of specifying string arguments for the <code>path</code>
argument. If you need further detail, see Section 20.2 of the SLiM
manual. In the script itself, you can then refer to this argument via
the constant <code>path</code>.</p>
</div>
<div class="section level3">
<h3 id="all-model-parameters-must-be-provided-as-additional-command-line-arguments">3. All model parameters must be provided as additional command-line
arguments<a class="anchor" aria-label="anchor" href="#all-model-parameters-must-be-provided-as-additional-command-line-arguments"></a>
</h3>
<p>You can refer to them in your script as you would to the mandatory
arguments as described in the previous section.</p>
<p><strong>A useful check to see whether you script would make a valid
engine for a <em>demografr</em> inference pipeline is to run it manually
on the command line like this:</strong></p>
<pre><code>python \
  &lt;path to your Python script&gt;                   \
  --path &lt;path to a directory&gt; \
  &lt;... your model parameters ...&gt;</code></pre>
<p>Or, if you want to run a SLiM-powered ABC, like this:</p>
<pre><code>slim \
  -d "path='&lt;path to a directory&gt;'" \
  &lt;... your model parameters ...&gt;</code></pre>
<p>Then, if you look into the directory and see the desired files
produced by your simulation script, you’re good to go!</p>
<p><strong>Below you are going to see how to actually read back
simulation results from disk and use them for computing summary
statistics. For now, just note that you can name the files produced by
your simulation script however you’d like and can have as many as you’d
like, as long as they are all in that one directory given by
<code>path</code>.</strong></p>
</div>
<div class="section level3">
<h3 id="example-pure-slim-and-msprime-script">Example pure SLiM and msprime script<a class="anchor" aria-label="anchor" href="#example-pure-slim-and-msprime-script"></a>
</h3>
<p>To demonstrate the requirements 1-3 above in practice, let’s run an
ABC inference using an example <em>msprime</em> script and a SLiM script
as simulation engines.</p>
<p>Imagine we have the following two scripts which we want to use as
simulation engines instead of <em>slendr</em>’s own functions
<code><a href="https://rdrr.io/pkg/slendr/man/slim.html" class="external-link">slim()</a></code> and <code><a href="https://rdrr.io/pkg/slendr/man/msprime.html" class="external-link">msprime()</a></code>. The scripts have only
<strong>one <em>model parameter</em>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>
of the single modeled population</strong>. The <strong>other parameters
are mandatory</strong>, just discussed in the previous section.</p>
<div class="section level4">
<h4 id="slim-script">SLiM script<a class="anchor" aria-label="anchor" href="#slim-script"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">slim_script</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/custom.slim"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span></span></code></pre></div>
<pre><code>initialize() {
  initializeTreeSeq();
  initializeMutationRate(1e-8);
  initializeMutationType("m1", 0.5, "f", 0.0);
  initializeGenomicElementType("g1", m1, 1.0);
  initializeGenomicElement(g1, 0, 1e6 - 1);
  initializeRecombinationRate(1e-8);
}

1 early() {
  sim.addSubpop("p0", asInteger(Ne));
}

10000 late() {
  ts_path = path + "/" + "result.trees";
  sim.treeSeqOutput(ts_path);
}</code></pre>
</div>
<div class="section level4">
<h4 id="python-script">Python script<a class="anchor" aria-label="anchor" href="#python-script"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">python_script</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/custom.py"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span></span></code></pre></div>
<pre><code>import argparse

import msprime

parser = argparse.ArgumentParser()

# mandatory command-line argument
parser.add_argument("--path", type=str, required=True)

# model parameters
parser.add_argument("--Ne", type=float, required=True)

args = parser.parse_args()

ts = msprime.sim_ancestry(
  samples=round(args.Ne),
  population_size=args.Ne,
  sequence_length=1e6,
  recombination_rate=1e-8,
)

ts = msprime.sim_mutations(ts, 1e-8)

ts.dump(args.path + "/" + "result.trees")</code></pre>
<p><strong>Note that the Python script specifies all model parameters
(here just
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>)
and mandatory arguments via a command-line interface provided by the
Python module <code>argparse</code></strong>.</p>
<p><strong>The SLiM script, on the other hand, uses SLiM’s features for
command-line specification of parameters and simply refers to each
parameter by its symbolic name.</strong> No other setup is
necessary.</p>
<p>Please also note that given that there are can be discrepancies
between values of arguments of some SLiM or Python methods (such as
<code>addSubPop</code> of SLiM which expects an integer value for a
population size, or the <code>samples</code> argument of
<code>msprime.sim_ancestry</code>), and values of parameters sampled
from priors by <em>demografr</em> (i.e.,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>
often being a floating-point value after sampling from a continuous
prior), you might have to perform explicit type conversion in your
custom scripts (such as <code>sim.addSubPop("p0", asInteger(Ne)</code>
as above).</p>
</div>
</div>
<div class="section level3">
<h3 id="abc-inference-using-custom-msprime-script">ABC inference using custom <em>msprime</em> script<a class="anchor" aria-label="anchor" href="#abc-inference-using-custom-msprime-script"></a>
</h3>
<p>Apart from the user-defined simulation SLiM and <em>msprime</em>
scripts, the components of our toy inference remains the same—we need to
define the observed statistics, tree-sequence summary functions, and
priors. We don’t need a model function—that will be served by our custom
script.</p>
<p>Here are the <em>demografr</em> pipeline components (we won’t be
discussing them here because that’s extensively taken care of <a href="vignette-01-basics.html">elsewhere</a> in <em>demografr</em>’s
documentation vignettes):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a single prior parameter</span></span>
<span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">Ne</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># a single observed statistic</span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">observed_diversity</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute diversity using 100 chromosomes</span></span>
<span><span class="va">compute_diversity_ts</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="va">ts</span><span class="op">$</span><span class="fu">samples</span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_diversity.html" class="external-link">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">samples</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">compute_diversity_ts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># data-generating functions for computing summary statistics</span></span>
<span><span class="va">gens</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  ts <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">path</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_read.html" class="external-link">ts_read</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"result.trees"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>(Note that because our simulated tree sequences won’t be coming with
<em>slendr</em> metadata, we have to refer to individuals’ chromosomes
using numerical indices rather than <em>slendr</em> symbolic names like
we do in all of our other vignettes.)</p>
<div class="section level4">
<h4 id="simulating-a-testing-tree-sequence">Simulating a testing tree sequence<a class="anchor" aria-label="anchor" href="#simulating-a-testing-tree-sequence"></a>
</h4>
<p>As we explained elsewhere, a useful function for developing inference
pipelines using <em>demografr</em> is a function
<code><a href="../reference/simulate_model.html">simulate_model()</a></code>, which normally accepts a <em>slendr</em>
model generating function and the model parameters (either given as
priors or as a list of named values), and simulates a tree-sequence
object. This function (as any other <em>demografr</em> function
operating with models) accepts our custom-defined simulation scripts in
place of standard <em>slendr</em> models.</p>
<p>For instance, we can simulate a couple of Mb of testing sequence from
our <em>msprime</em> script like this:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_run</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_model.html">simulate_model</a></span><span class="op">(</span><span class="va">python_script</span>, parameters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Ne <span class="op">=</span> <span class="fl">123</span><span class="op">)</span>, format <span class="op">=</span> <span class="st">"files"</span>, data <span class="op">=</span> <span class="va">gens</span><span class="op">)</span></span></code></pre></div>
<p>Now we can finally test our toy tree-sequence summary function,
verifying that we can indeed compute the summary statistic we want
to:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">functions</span><span class="op">$</span><span class="fu">diversity</span><span class="op">(</span><span class="va">model_run</span><span class="op">$</span><span class="va">ts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop   0.000<span style="text-decoration: underline;">008</span>40</span></span></code></pre>
<p>The function works as expected—we have a single population and want
to compute nucleotide diversity in the whole population, and this is
exactly what we get.</p>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop   0.000<span style="text-decoration: underline;">006</span>23</span></span></code></pre>
<pre><code><span><span class="co">#&gt; $diversity</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop   0.000<span style="text-decoration: underline;">006</span>62</span></span></code></pre>
<p>Note that we are also able to use a dedicated function for this,
called <code><a href="../reference/summarise_data.html">summarise_data()</a></code>, which takes in a product of a
simulation (tree sequence and/or a path to a directory with simulation
results), and apply all summary functions to this data:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarise_data.html">summarise_data</a></span><span class="op">(</span><span class="va">model_run</span>, <span class="va">functions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop   0.000<span style="text-decoration: underline;">008</span>32</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="abc-inference">ABC inference<a class="anchor" aria-label="anchor" href="#abc-inference"></a>
</h4>
<p>Having all components of our pipeline set up we should, again,
validate everything before we proceed to (potentially very costly)
simulations. In the remainder of this vignette we’ll only continue with
the Python msprime custom script in order to save some computational
time. That said, it’s important to realize that you could use this sort
of workflow for any kind of SLiM script, including very elaborate
spatial simulations, non-WF models, and all kinds of phenotypic
simulations too! <strong>Although primarily designed to work with
<em>slendr</em>, the <em>demografr</em> package intends to fully support
any kind of SLiM or <em>msprime</em> simulation. If something doesn’t
work, consider it a <em>demografr</em> bug!</strong></p>
<p>Let’s first validate all components of our pipeline:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">python_script</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, data <span class="op">=</span> <span class="va">gens</span>, format <span class="op">=</span> <span class="st">"files"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; Testing sampling of each prior parameter:</span></span>
<span><span class="co">#&gt;   - Ne ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; The model is a custom user-defined msprime script</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Simulating tree sequence from the given model... ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Validating custom data-generating functions...  ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Generating data from simulation results:</span></span>
<span><span class="co">#&gt;   - ts (type slendr_ts) ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Computing user-defined summary functions:</span></span>
<span><span class="co">#&gt;   - diversity ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the format of simulated summary statistics:</span></span>
<span><span class="co">#&gt;   - diversity [data frame] ✅</span></span>
<span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; No issues have been found in the ABC setup!</span></span></code></pre>
<p>Looking good! Now let’s first run ABC simulations. Again, note the
use of the Python script where we would normally provide a
<em>slendr</em> model function in place of the <code>model</code>
argument:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="co"># parallelize across all CPUs</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span></span>
<span>  model <span class="op">=</span> <span class="va">python_script</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  data <span class="op">=</span> <span class="va">gens</span>, format <span class="op">=</span> <span class="st">"files"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>The total runtime for the ABC simulations was 0 hours 29
minutes 36 seconds parallelized across 96 CPUs.</strong></p>
<p>Once the simulations are finished, we can perform inference of the
posterior distribution of the single parameter of our model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_abc.html">run_abc</a></span><span class="op">(</span><span class="va">data</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; 12345678910</span></span>
<span><span class="co">#&gt; 12345678910</span></span></code></pre>
<p>Having done so, we can again look at the summary statistics of the
posterior distribution and also plot the results (we’re skipping
diagnostics such as posterior predictive checks but you can read more
about those <a href="vignette-01-basics.html">here</a> and <a href="vignette-06-diagnostics.html">here</a>). Because this
<code>observed</code> diversity is based on simulated data from a known
model, we’ll indicate the true “hidden”
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>
value with a vertical line.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                               Ne</span></span>
<span><span class="co">#&gt; Min.:                   648.0043</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:   683.6240</span></span>
<span><span class="co">#&gt; Weighted Median:       1005.4589</span></span>
<span><span class="co">#&gt; Weighted Mean:         1044.8909</span></span>
<span><span class="co">#&gt; Weighted Mode:          997.3463</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 1521.7129</span></span>
<span><span class="co">#&gt; Max.:                  1555.8246</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">1000</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-05-custom_files/figure-html/custom_Ne_posterior1-1.png" width="640" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="using-external-programs-to-compute-summary-statistics">Using external programs to compute summary statistics<a class="anchor" aria-label="anchor" href="#using-external-programs-to-compute-summary-statistics"></a>
</h3>
<p>Now let’s move things one step forward. What if we either don’t find
<em>slendr</em>’s tree-sequence interface to <em>tskit</em> sufficient
and want to compute summary statistics from the simulated data using
some external software, like PLINK or EIGENSTRAT?</p>
<p>To keep things as easy as possible (and avoid dragging in multiple
software dependencies), let’s say we have a program <code>vcfpi</code>,
which is run in the following way:</p>
<pre><code>./vcfpi --vcf &lt;path to a VCF&gt; --tsv &lt;path to a TSV table&gt;</code></pre>
<p>You can find it at
/root/renv/project-20bde257/linux-ubuntu-noble/R-4.5/x86_64-pc-linux-gnu/demografr/examples/vcfpi)
and run it yourself in the terminal.</p>
<p>This program takes in a VCF file and computes the nucleotide
diversity across all individuals in that VCF file. For simplicity, it
produces a table of results in exactly the same format as the
<code>observed_diversity</code> data frame above.</p>
<p>In this particular example, we want to replace the computation of
statistics performed by the R function <code>compute_diversity</code> in
the previous section (which also produces a data frame) with the product
of the external software <code>vcfpi</code>.</p>
<p>(We provide this <code>vcfpi</code> program as an example script, but
it could of course be any other program you could imagine.)</p>
<p>First, the definitions of <code>priors</code> and
<code>observed</code> statistics do not change:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a single prior parameter</span></span>
<span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">Ne</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># a single observed statistic</span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">observed_diversity</span><span class="op">)</span></span></code></pre></div>
<p>What does change is the simulated summary statistics definition.
There are a number of ways we could go about this (and we will show
others below), but given that our <code>vcfpi</code> is a command-line
program, we will use the function</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compute_pi_vcf</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># path to the output summary statistic file</span></span>
<span>  <span class="va">tsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># path to the example command-line utility (in this example we use a built-in toy program</span></span>
<span>  <span class="co"># called `vcfpi`, but this could be a path to any other program of your choosing, of course)</span></span>
<span>  <span class="va">program</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/vcfpi"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># execute the program on the command line using appropriate arguments</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system2.html" class="external-link">system2</a></span><span class="op">(</span><span class="va">program</span>, args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"--vcf"</span>, <span class="va">vcf</span>, <span class="st">"--tsv"</span>, <span class="va">tsv</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># read the computed table with computed nucleotide diversity</span></span>
<span>  <span class="co"># (for simplicity, our example vcfpi program creates an output TSV</span></span>
<span>  <span class="co"># file already in the necessary format but any required reformatting</span></span>
<span>  <span class="co"># changes could easily happen at this stage)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">tsv</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">compute_pi_vcf</span><span class="op">)</span></span></code></pre></div>
<p>There’s one last thing we need to do – our summary function
<code>compute_diversity</code> does not operate on a tree-sequence file
this time. Instead, <code>vcfpi</code> requires a VCF file as its input.
But, because of our simulation model <code>python_script</code> still
creates a tree-sequence file, we need to create a VCF to be able to
compute nucleotide diversity.</p>
<p>(Of course, we could modify the simulation model script to produce
such VCF easily, but for the sake of making this example a bit more
educational, let’s make our job a little bit harder).</p>
<p>Whenever we need to run a simulation summary statistic on a different
form of data than a tree-sequence, we can further customize the
data-generating function(s). In the first example, we had to provide a
function which creates a <code>ts</code> object from <code>path</code>
to a directory with all result files. In this example, let’s push this
one step further and create a VCF file.</p>
<p>(Again, we could’ve just as easily create a VCF file from our
simulation script, but bear with me here!)</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gens</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  vcf <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">path</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># read the simulated tree sequence from a given path and subset it</span></span>
<span>    <span class="co"># (same code as previously, where we worked exclusively on tree sequences)</span></span>
<span>    <span class="va">ts_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"result.trees"</span><span class="op">)</span></span>
<span>    <span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_read.html" class="external-link">ts_read</a></span><span class="op">(</span><span class="va">ts_path</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_simplify.html" class="external-link">ts_simplify</a></span><span class="op">(</span>simplify_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">99</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># export genotypes from the tree sequence to a VCF (in the same directory)</span></span>
<span>    <span class="va">vcf_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"genotypes.vcf.gz"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_vcf.html" class="external-link">ts_vcf</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">vcf_path</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># return the path to the VCF for use in computing summary statistics</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">vcf_path</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_run</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_model.html">simulate_model</a></span><span class="op">(</span><span class="va">python_script</span>, parameters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Ne <span class="op">=</span> <span class="fl">123</span><span class="op">)</span>, format <span class="op">=</span> <span class="st">"files"</span>, data <span class="op">=</span> <span class="va">gens</span><span class="op">)</span></span></code></pre></div>
<p>When we look at the result, we can see that it no longer contains a
<code>ts</code> tree-sequence object but a path to a VCF file!</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_run</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $vcf</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpHVT3ENdemografr_323504404/genotypes.vcf.gz"</span></span></code></pre>
<p>Let’s test our summary function on that VCF:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">functions</span><span class="op">$</span><span class="fu">diversity</span><span class="op">(</span><span class="va">model_run</span><span class="op">$</span><span class="va">vcf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt; 1 pop 1.005325e-05</span></span></code></pre>
<p>Or, for convenience, we can do this using the more general function
<code><a href="../reference/summarise_data.html">summarise_data()</a></code>:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarise_data.html">summarise_data</a></span><span class="op">(</span><span class="va">model_run</span>, <span class="va">functions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span>
<span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt; 1 pop 1.005325e-05</span></span></code></pre>
<p>Now let’s validate our customized setup before we unleash the full
scale of ABC simulations on the problem!</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">python_script</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, data <span class="op">=</span> <span class="va">gens</span>, format <span class="op">=</span> <span class="st">"files"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; Testing sampling of each prior parameter:</span></span>
<span><span class="co">#&gt;   - Ne ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; The model is a custom user-defined msprime script</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Simulating tree sequence from the given model... ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Validating custom data-generating functions...  ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Generating data from simulation results:</span></span>
<span><span class="co">#&gt;   - vcf (type character) ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Computing user-defined summary functions:</span></span>
<span><span class="co">#&gt;   - diversity ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the format of simulated summary statistics:</span></span>
<span><span class="co">#&gt;   - diversity [data frame] ✅</span></span>
<span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; No issues have been found in the ABC setup!</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="co"># parallelize across all CPUs</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span></span>
<span>  model <span class="op">=</span> <span class="va">python_script</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  data <span class="op">=</span> <span class="va">gens</span>, format <span class="op">=</span> <span class="st">"files"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>The total runtime for the ABC simulations was 0 hours 33
minutes 25 seconds parallelized across 96 CPUs.</strong></p>
<p>Once the simulations are finished, we can perform inference of the
posterior distribution of the single parameter of our model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_abc.html">run_abc</a></span><span class="op">(</span><span class="va">data</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; 12345678910</span></span>
<span><span class="co">#&gt; 12345678910</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                               Ne</span></span>
<span><span class="co">#&gt; Min.:                   641.4341</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:   701.8172</span></span>
<span><span class="co">#&gt; Weighted Median:        989.0880</span></span>
<span><span class="co">#&gt; Weighted Mean:         1026.1556</span></span>
<span><span class="co">#&gt; Weighted Mode:          919.2565</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 1495.9363</span></span>
<span><span class="co">#&gt; Max.:                  1566.3955</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">1000</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-05-custom_files/figure-html/custom_Ne_posterior2-1.png" width="640" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="another-means-of-computing-simulated-summary-statistics">Another means of computing simulated summary statistics<a class="anchor" aria-label="anchor" href="#another-means-of-computing-simulated-summary-statistics"></a>
</h3>
<p>Let’s look at a yet another variation of the same. What if we want to
compute a summary statistics in R on a different source of genotype data
(not a tree-sequence object in memory but also not a VCF file stored on
disk)?</p>
<p>For instance, what if we want to compute a summary statistics on a
simple data frame of genotypes (<code>0</code> – ancestral state,
<code>1</code> – derived state).</p>
<p>We can modify our pipeline in the following way.</p>
<p>First, we need to define a summary statistic function which will
operate on a data frame, say called <code>gt</code>.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compute_diversity</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">gt</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># generate a list of all pairs of chromosomes</span></span>
<span>  <span class="va">pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gt</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>, <span class="fl">2</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># count the number of nucleotide differences for each pair</span></span>
<span>  <span class="va">diffs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">pairs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">pair</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">gt</span><span class="op">[[</span><span class="va">pair</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">]</span> <span class="op">!=</span> <span class="va">gt</span><span class="op">[[</span><span class="va">pair</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># the mean is the nucleotide diversity of the sample of chromosomes</span></span>
<span>  <span class="va">pi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">diffs</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1e6</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>set <span class="op">=</span> <span class="st">"pop"</span>, diversity <span class="op">=</span> <span class="va">pi</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">compute_diversity</span><span class="op">)</span></span></code></pre></div>
<p>We also need a data-generating function, which will create the table
of genotypes from standard simulated output:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gens</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  gt <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">path</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># read the simulated tree sequence from a given path and subset it</span></span>
<span>    <span class="va">ts_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"result.trees"</span><span class="op">)</span></span>
<span>    <span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_read.html" class="external-link">ts_read</a></span><span class="op">(</span><span class="va">ts_path</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_simplify.html" class="external-link">ts_simplify</a></span><span class="op">(</span>simplify_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">99</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># suppress warnings due to multiallelic sites</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_genotypes.html" class="external-link">ts_genotypes</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And that’s it! Let’s check again that the whole setup works
correctly.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_run</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_model.html">simulate_model</a></span><span class="op">(</span><span class="va">python_script</span>, parameters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Ne <span class="op">=</span> <span class="fl">123</span><span class="op">)</span>, format <span class="op">=</span> <span class="st">"files"</span>, data <span class="op">=</span> <span class="va">gens</span><span class="op">)</span></span></code></pre></div>
<p>When we look at the result, we can see that it no longer contains a
<code>ts</code> tree-sequence object but a simple table of genotypes
created via the <em>slendr</em> function <code><a href="https://rdrr.io/pkg/slendr/man/ts_genotypes.html" class="external-link">ts_genotypes()</a></code>
(see the data generating function just above).</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_run</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $gt</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 30 × 101</span></span></span>
<span><span class="co">#&gt;       pos pop_0_0_chr1 pop_0_0_chr2 pop_0_1_chr1 pop_0_1_chr2 pop_0_2_chr1</span></span>
<span><span class="co">#&gt;     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>  <span style="text-decoration: underline;">12</span>466            0            0            0            0            0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>  <span style="text-decoration: underline;">96</span>458            0            0            0            0            0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> <span style="text-decoration: underline;">124</span>061            1            1            1            1            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> <span style="text-decoration: underline;">174</span>991            0            0            0            1            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> <span style="text-decoration: underline;">197</span>064            0            0            0            0            0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> <span style="text-decoration: underline;">223</span>645            0            0            0            0            0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> <span style="text-decoration: underline;">237</span>832            0            0            0            0            0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> <span style="text-decoration: underline;">262</span>791            0            0            0            0            0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> <span style="text-decoration: underline;">263</span>576            0            0            0            0            0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> <span style="text-decoration: underline;">265</span>948            1            1            1            1            1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 95 more variables: pop_0_2_chr2 &lt;int&gt;, pop_0_3_chr1 &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   pop_0_3_chr2 &lt;int&gt;, pop_0_4_chr1 &lt;int&gt;, pop_0_4_chr2 &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   pop_0_5_chr1 &lt;int&gt;, pop_0_5_chr2 &lt;int&gt;, pop_0_6_chr1 &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   pop_0_6_chr2 &lt;int&gt;, pop_0_7_chr1 &lt;int&gt;, pop_0_7_chr2 &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   pop_0_8_chr1 &lt;int&gt;, pop_0_8_chr2 &lt;int&gt;, pop_0_9_chr1 &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   pop_0_9_chr2 &lt;int&gt;, pop_0_10_chr1 &lt;int&gt;, pop_0_10_chr2 &lt;int&gt;, …</span></span></span></code></pre>
<p>Let’s test our summary function on the output of the single testing
simulation run:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarise_data.html">summarise_data</a></span><span class="op">(</span><span class="va">model_run</span>, <span class="va">functions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span>
<span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt; 1 pop 6.589899e-06</span></span></code></pre>
<p>Let’s also validate our customized setup before we unleash the full
scale of ABC simulations on the problem!</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">python_script</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, data <span class="op">=</span> <span class="va">gens</span>, format <span class="op">=</span> <span class="st">"files"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; Testing sampling of each prior parameter:</span></span>
<span><span class="co">#&gt;   - Ne ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; The model is a custom user-defined msprime script</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Simulating tree sequence from the given model... ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Validating custom data-generating functions...  ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Generating data from simulation results:</span></span>
<span><span class="co">#&gt;   - gt (type tbl_df) ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Computing user-defined summary functions:</span></span>
<span><span class="co">#&gt;   - diversity ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the format of simulated summary statistics:</span></span>
<span><span class="co">#&gt;   - diversity [data frame] ✅</span></span>
<span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; No issues have been found in the ABC setup!</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span></span>
<span>  model <span class="op">=</span> <span class="va">python_script</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  data <span class="op">=</span> <span class="va">gens</span>, format <span class="op">=</span> <span class="st">"files"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 826 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 993 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.167% out of 600 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.151% out of 663 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 915 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 949 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 827 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 773 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.132% out of 755 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1014 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 995 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 984 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 937 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.175% out of 571 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.265% out of 378 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 923 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 741 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 820 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.219% out of 915 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 808 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 866 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 812 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 770 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 908 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 873 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.181% out of 552 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 877 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 913 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 954 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 918 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.170% out of 589 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.206% out of 486 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1045 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.161% out of 621 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 897 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.304% out of 329 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.244% out of 820 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 916 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.187% out of 535 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.141% out of 709 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.133% out of 752 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.233% out of 857 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 927 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 772 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.205% out of 487 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 802 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 789 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 908 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 965 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.212% out of 942 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.148% out of 677 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.088% out of 1135 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.179% out of 1119 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.088% out of 1142 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 966 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.149% out of 673 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 875 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.145% out of 692 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.137% out of 728 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 768 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.143% out of 700 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 992 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.086% out of 1168 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.161% out of 623 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.197% out of 507 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.231% out of 867 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 995 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 785 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.234% out of 428 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1067 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.139% out of 719 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 991 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1037 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 774 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.242% out of 825 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 969 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.172% out of 581 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1037 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 981 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1007 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1037 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.299% out of 669 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 804 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.167% out of 598 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 804 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1007 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 796 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.150% out of 665 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.232% out of 861 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.178% out of 563 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.253% out of 396 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.151% out of 664 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 897 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 888 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 727 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 887 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.149% out of 673 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 848 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 862 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.244% out of 818 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 798 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 965 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.242% out of 414 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.147% out of 680 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 912 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.158% out of 634 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 743 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.092% out of 1082 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.092% out of 1083 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 818 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.134% out of 748 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 842 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 888 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 953 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 897 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.172% out of 580 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.294% out of 680 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.295% out of 339 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 960 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.164% out of 609 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 768 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 825 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.223% out of 449 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.128% out of 779 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.128% out of 783 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.194% out of 1029 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.280% out of 714 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.305% out of 655 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 950 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 893 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 994 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 811 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 921 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 982 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.174% out of 575 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.177% out of 564 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.184% out of 543 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.139% out of 720 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 892 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.259% out of 386 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.150% out of 667 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.152% out of 657 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 774 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.231% out of 865 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.209% out of 959 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.214% out of 467 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 792 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 3 multiallelic sites (0.274% out of 1095 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.192% out of 522 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1027 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 788 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 987 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.204% out of 982 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 851 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.186% out of 539 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.257% out of 389 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.238% out of 841 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 829 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.181% out of 553 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.295% out of 339 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.216% out of 928 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.189% out of 1059 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.245% out of 408 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.217% out of 922 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.171% out of 585 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 891 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 927 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 808 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 743 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.223% out of 897 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 929 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 977 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 867 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 916 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.226% out of 886 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.143% out of 700 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 853 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.132% out of 758 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.178% out of 562 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.155% out of 645 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 739 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 911 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 992 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 795 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 786 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.219% out of 457 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.313% out of 319 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 743 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 837 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.206% out of 971 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1006 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.216% out of 464 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 846 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 898 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1026 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 3 multiallelic sites (0.326% out of 921 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.180% out of 555 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 951 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.187% out of 535 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.143% out of 698 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.207% out of 484 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 991 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.098% out of 1020 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 846 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 797 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.338% out of 296 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 684 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.267% out of 374 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1057 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.246% out of 812 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 3 multiallelic sites (0.580% out of 517 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.170% out of 588 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.180% out of 1112 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 742 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 868 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1054 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.134% out of 746 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 948 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 851 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 911 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.137% out of 728 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 773 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 919 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.244% out of 409 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 855 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 817 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 953 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.205% out of 977 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.098% out of 1020 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.225% out of 444 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.142% out of 702 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 908 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 901 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1037 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 724 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1064 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 905 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 933 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.139% out of 717 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.098% out of 1025 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.223% out of 896 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1068 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 742 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 898 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 958 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 793 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.086% out of 1169 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.267% out of 375 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 828 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.179% out of 1119 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 790 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 939 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.450% out of 444 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.168% out of 597 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1041 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.662% out of 151 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.166% out of 603 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.147% out of 679 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 895 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.472% out of 212 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 920 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.181% out of 552 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.264% out of 379 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 971 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.236% out of 847 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 896 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 813 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.174% out of 576 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 810 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 954 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.088% out of 1132 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.176% out of 567 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.128% out of 779 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 871 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 687 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 726 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.194% out of 515 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.170% out of 588 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 918 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.200% out of 1001 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.159% out of 630 total) detected and removed</span></span>
<span><span class="co">#&gt; 1 multiallelic sites (0.159% out of 630 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.183% out of 545 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.211% out of 947 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.180% out of 557 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 896 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 851 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.222% out of 451 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.150% out of 666 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.187% out of 1072 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.175% out of 571 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.181% out of 554 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.205% out of 488 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.279% out of 717 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.143% out of 700 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.149% out of 673 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1040 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.212% out of 472 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 817 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 897 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.236% out of 423 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.226% out of 443 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 923 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (1.515% out of 66 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.128% out of 782 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1005 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.218% out of 917 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.145% out of 692 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 773 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1044 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.169% out of 1184 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.134% out of 744 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.229% out of 874 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1035 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.248% out of 808 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 683 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.091% out of 1095 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.261% out of 383 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 787 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 794 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 932 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 777 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 947 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 981 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 892 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 941 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.134% out of 749 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.141% out of 708 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 836 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.205% out of 487 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 887 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 974 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 826 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.220% out of 908 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.092% out of 1090 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.333% out of 300 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.168% out of 596 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.158% out of 632 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 841 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.222% out of 450 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 867 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.225% out of 890 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 882 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.325% out of 308 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.222% out of 902 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 3 multiallelic sites (0.290% out of 1035 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 824 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.196% out of 1023 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.392% out of 255 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 683 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.090% out of 1112 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.193% out of 517 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.208% out of 960 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 913 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.234% out of 428 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 869 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1002 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.205% out of 974 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 905 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 989 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.187% out of 535 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.144% out of 695 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 852 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.128% out of 779 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.227% out of 882 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 969 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.191% out of 523 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 861 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 899 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.144% out of 694 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 794 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.216% out of 928 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.242% out of 414 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1039 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.402% out of 497 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1026 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1029 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 927 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 997 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 911 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.181% out of 554 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.287% out of 697 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.211% out of 949 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 828 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.214% out of 467 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.132% out of 758 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.398% out of 251 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.157% out of 635 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 926 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 873 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.145% out of 692 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.203% out of 493 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.246% out of 812 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.208% out of 963 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 852 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.152% out of 659 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.231% out of 867 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 776 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.176% out of 569 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.463% out of 216 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 725 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.093% out of 1077 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 932 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 934 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.152% out of 659 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.214% out of 468 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 984 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 908 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 974 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.339% out of 295 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.503% out of 199 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 978 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.174% out of 1148 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.167% out of 600 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.155% out of 645 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 775 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 807 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 835 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.279% out of 359 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1000 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.154% out of 651 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 939 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 955 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.098% out of 1024 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.211% out of 948 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 952 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.158% out of 631 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.156% out of 642 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.149% out of 673 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.187% out of 534 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.271% out of 737 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.158% out of 631 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1045 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 810 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 928 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.190% out of 525 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 891 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.156% out of 643 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 860 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.199% out of 1007 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 956 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 821 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 894 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 974 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.151% out of 662 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 943 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 740 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 913 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.160% out of 625 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.216% out of 464 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1054 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1001 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.098% out of 1019 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 947 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 852 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 976 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1015 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.091% out of 1096 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.216% out of 464 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.230% out of 871 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.230% out of 434 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.088% out of 1130 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 918 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.230% out of 434 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.162% out of 616 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1009 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.203% out of 492 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.228% out of 439 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 814 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.219% out of 457 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.250% out of 801 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.148% out of 677 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.332% out of 301 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.198% out of 506 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.128% out of 781 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 874 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 921 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.250% out of 400 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 774 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 899 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 803 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.191% out of 524 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1008 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.199% out of 503 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.141% out of 710 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1043 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 771 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 848 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.265% out of 378 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.141% out of 710 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.093% out of 1073 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 957 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 806 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 954 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 973 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 986 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.163% out of 612 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.198% out of 1011 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.093% out of 1070 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 969 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 815 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 786 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 846 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.086% out of 1160 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 943 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1001 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 804 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.140% out of 714 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.092% out of 1086 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 842 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 772 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 839 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 981 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.186% out of 1076 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.092% out of 1082 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.147% out of 680 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 990 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.137% out of 732 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1001 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 903 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.299% out of 334 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.175% out of 570 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.227% out of 440 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 980 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.161% out of 622 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1067 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1001 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.204% out of 491 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1006 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 972 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.147% out of 679 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.194% out of 1030 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.212% out of 943 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 962 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.173% out of 577 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.198% out of 504 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 978 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.141% out of 707 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 931 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.093% out of 1074 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 847 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 726 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 815 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.243% out of 822 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 794 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1013 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.177% out of 566 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.093% out of 1076 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 817 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.166% out of 601 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.148% out of 676 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.152% out of 660 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 996 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 920 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.155% out of 644 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 943 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.741% out of 135 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.199% out of 502 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.194% out of 515 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.183% out of 546 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.166% out of 601 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.131% out of 761 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 977 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 767 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1051 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.752% out of 133 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.213% out of 469 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 946 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.156% out of 640 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 952 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1060 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.141% out of 709 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.278% out of 360 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 838 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 806 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.234% out of 427 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.175% out of 571 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.250% out of 400 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.437% out of 229 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 863 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.093% out of 1081 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 802 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 836 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.147% out of 678 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 981 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.144% out of 694 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.293% out of 341 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.156% out of 643 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 961 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.174% out of 575 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 867 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 884 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.186% out of 1077 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.137% out of 728 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.192% out of 521 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.196% out of 1021 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 892 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.149% out of 673 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 909 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 810 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 855 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 840 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.181% out of 1103 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 742 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.213% out of 470 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.187% out of 534 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.144% out of 694 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 929 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1011 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 864 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.194% out of 1031 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.220% out of 454 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.103% out of 973 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 725 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.166% out of 604 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.196% out of 510 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 925 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.483% out of 207 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 875 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.134% out of 746 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.158% out of 633 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 985 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.134% out of 749 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 954 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.377% out of 531 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.143% out of 699 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.408% out of 245 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.179% out of 560 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 739 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.163% out of 614 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.088% out of 1139 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 891 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.365% out of 274 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.237% out of 422 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.248% out of 808 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.264% out of 379 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 912 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.218% out of 919 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.196% out of 1021 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 893 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 902 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 945 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 948 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1068 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.168% out of 597 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 798 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1007 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.201% out of 993 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.205% out of 975 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1013 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 864 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 856 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 888 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.189% out of 529 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.152% out of 657 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 905 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1069 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.202% out of 496 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.203% out of 493 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.189% out of 529 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 875 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.248% out of 806 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 683 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 788 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.139% out of 717 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.180% out of 557 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 997 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1005 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 811 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 3 multiallelic sites (0.274% out of 1095 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 926 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.089% out of 1118 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.175% out of 573 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 886 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.248% out of 404 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.168% out of 597 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 854 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 685 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.274% out of 365 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 829 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 984 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.342% out of 292 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.439% out of 228 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.188% out of 1063 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 871 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 814 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.269% out of 372 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 995 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.229% out of 437 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.242% out of 827 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1062 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.168% out of 597 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1001 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.139% out of 721 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 995 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.132% out of 756 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.202% out of 496 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.098% out of 1020 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.133% out of 752 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.186% out of 538 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 880 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.248% out of 403 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 778 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.090% out of 1105 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.198% out of 504 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 791 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1027 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.088% out of 1141 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1064 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.261% out of 766 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 844 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.362% out of 276 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.298% out of 336 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 819 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.172% out of 1165 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.182% out of 548 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1001 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.248% out of 805 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.149% out of 670 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 830 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 920 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.133% out of 754 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.172% out of 582 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 906 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.183% out of 547 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1067 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.196% out of 510 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 936 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 868 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 914 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.092% out of 1083 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1028 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.172% out of 580 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1035 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.158% out of 631 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 867 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.132% out of 758 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 937 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.136% out of 735 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.171% out of 584 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 991 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.197% out of 508 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.209% out of 957 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.181% out of 551 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.179% out of 559 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 937 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 912 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.148% out of 677 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 878 total) detected and removed</span></span>
<span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 878 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.086% out of 1162 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 4 multiallelic sites (0.410% out of 976 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.089% out of 1126 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.098% out of 1018 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.442% out of 226 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 911 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.209% out of 478 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 874 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.134% out of 745 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.155% out of 646 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 798 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 921 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.174% out of 576 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 925 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.250% out of 800 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.167% out of 598 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.157% out of 636 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.168% out of 597 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 924 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 854 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 860 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.253% out of 395 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 830 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.190% out of 526 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.285% out of 701 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.157% out of 637 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.089% out of 1123 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.169% out of 592 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 929 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.262% out of 764 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1014 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 806 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.136% out of 733 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.221% out of 907 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.129% out of 775 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.242% out of 825 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 892 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 909 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 990 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.204% out of 489 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.483% out of 207 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.176% out of 567 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.302% out of 331 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 799 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.229% out of 436 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.153% out of 652 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 804 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.145% out of 689 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 865 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 947 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 956 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (1.562% out of 64 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.092% out of 1090 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.220% out of 911 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 905 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 901 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 931 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 963 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 885 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.224% out of 446 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.232% out of 431 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 870 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1061 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 822 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.092% out of 1086 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.093% out of 1077 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.144% out of 693 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.178% out of 562 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 929 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1069 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 929 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 890 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 726 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 943 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.241% out of 831 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 887 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 845 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 792 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.083% out of 1199 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 877 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 905 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1005 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.207% out of 482 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 984 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1064 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 946 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1015 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 989 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1029 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.156% out of 639 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.141% out of 707 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.089% out of 1123 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 940 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.164% out of 609 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 979 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 906 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 941 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1065 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.263% out of 380 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.151% out of 664 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 977 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 862 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.261% out of 767 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.132% out of 756 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 724 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 955 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.168% out of 596 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 995 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.098% out of 1017 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1055 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.154% out of 651 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 884 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.089% out of 1129 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 822 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 802 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.229% out of 874 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 884 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1032 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.148% out of 674 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 953 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.221% out of 453 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 801 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1057 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.136% out of 734 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 935 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.136% out of 738 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.164% out of 609 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 867 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 933 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 881 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.133% out of 750 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.203% out of 493 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 934 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 815 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.142% out of 704 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.145% out of 690 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.136% out of 737 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 686 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.093% out of 1073 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.189% out of 530 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.133% out of 752 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 787 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.143% out of 701 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.187% out of 1071 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 811 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 922 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.155% out of 645 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 836 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 844 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 809 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.294% out of 680 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.182% out of 1097 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 930 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.188% out of 531 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 801 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.128% out of 783 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.153% out of 654 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.148% out of 676 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 767 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 947 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.143% out of 701 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.163% out of 615 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.163% out of 614 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 855 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.208% out of 480 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.444% out of 225 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.145% out of 689 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.289% out of 692 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.340% out of 294 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 917 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.155% out of 645 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.132% out of 755 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.157% out of 635 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.175% out of 1142 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.155% out of 646 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 798 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1038 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 686 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 836 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.228% out of 879 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.124% out of 809 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.203% out of 985 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 814 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1058 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.115% out of 866 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.201% out of 498 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.164% out of 610 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.088% out of 1135 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.221% out of 904 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 723 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.131% out of 761 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 904 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.153% out of 652 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.209% out of 479 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.222% out of 899 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.149% out of 672 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.153% out of 652 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 812 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 941 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.090% out of 1116 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.166% out of 602 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.090% out of 1113 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1029 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.216% out of 464 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 964 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 845 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 882 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.166% out of 602 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 955 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 941 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 797 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 895 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1037 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.140% out of 713 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 958 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 961 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1035 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 3 multiallelic sites (0.273% out of 1099 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 769 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.139% out of 722 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1032 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.213% out of 470 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.221% out of 453 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.200% out of 501 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.351% out of 570 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.194% out of 1029 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.231% out of 867 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1009 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.155% out of 647 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.217% out of 920 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 837 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 930 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 891 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 944 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 907 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 795 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.261% out of 383 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.148% out of 675 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 962 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.275% out of 363 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.147% out of 681 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.150% out of 667 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 841 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 958 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.285% out of 702 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 874 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 785 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.190% out of 525 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.178% out of 563 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.167% out of 600 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.195% out of 514 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.155% out of 646 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.187% out of 536 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.153% out of 652 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 823 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.222% out of 451 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1040 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 928 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1051 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 964 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.218% out of 916 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.170% out of 588 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1006 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 892 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1039 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.184% out of 1086 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 936 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 824 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.161% out of 620 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 827 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.311% out of 322 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 965 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 941 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1008 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.142% out of 705 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 723 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.119% out of 842 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.116% out of 862 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.266% out of 376 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.353% out of 566 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.157% out of 635 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.172% out of 582 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.132% out of 759 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.228% out of 876 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.259% out of 386 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 831 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 916 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.118% out of 850 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1026 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.091% out of 1094 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.210% out of 953 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 685 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.246% out of 814 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 900 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 947 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 831 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1044 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 940 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1064 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 965 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.154% out of 649 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 942 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.375% out of 267 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 932 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1051 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1056 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 924 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.230% out of 435 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 741 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.224% out of 447 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 883 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.172% out of 583 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 743 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.254% out of 393 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.231% out of 865 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.211% out of 473 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 930 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.160% out of 625 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 890 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.107% out of 931 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 878 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 882 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.128% out of 782 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1007 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.220% out of 455 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.200% out of 499 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.114% out of 874 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.117% out of 856 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 836 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.110% out of 912 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.240% out of 416 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 941 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.197% out of 508 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.211% out of 474 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 801 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 891 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 888 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.139% out of 721 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 994 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.109% out of 914 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.136% out of 734 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.225% out of 444 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.141% out of 710 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.149% out of 670 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.228% out of 879 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.106% out of 943 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.833% out of 120 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.170% out of 588 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.140% out of 716 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 3 multiallelic sites (0.312% out of 963 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.383% out of 261 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 889 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.212% out of 942 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 792 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.271% out of 738 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.104% out of 960 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.143% out of 699 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.105% out of 956 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.264% out of 379 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.211% out of 949 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.113% out of 883 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.135% out of 742 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.485% out of 206 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 813 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 832 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.325% out of 308 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.217% out of 460 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1051 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.090% out of 1105 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.102% out of 979 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.201% out of 996 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.160% out of 624 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.265% out of 377 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.123% out of 816 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.228% out of 879 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 833 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 684 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.125% out of 803 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.160% out of 624 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 894 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.126% out of 793 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.415% out of 241 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.280% out of 715 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.185% out of 1081 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.170% out of 589 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 926 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.403% out of 248 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 830 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.322% out of 311 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.205% out of 975 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1006 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.171% out of 586 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.275% out of 363 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.328% out of 305 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.173% out of 577 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.166% out of 604 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.146% out of 683 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.101% out of 994 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.111% out of 900 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.140% out of 714 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.185% out of 1081 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.195% out of 514 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.127% out of 786 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.121% out of 827 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.097% out of 1032 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.140% out of 713 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.148% out of 677 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 890 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.242% out of 828 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.137% out of 728 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.202% out of 990 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.093% out of 1070 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.299% out of 334 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.367% out of 545 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.091% out of 1098 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.100% out of 1000 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.212% out of 472 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.122% out of 820 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.095% out of 1057 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.099% out of 1009 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.096% out of 1044 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.142% out of 706 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.120% out of 832 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.112% out of 893 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.108% out of 925 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.138% out of 726 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.091% out of 1098 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.152% out of 657 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1065 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.272% out of 368 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.145% out of 689 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.241% out of 831 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 2 multiallelic sites (0.185% out of 1084 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.130% out of 768 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 3 multiallelic sites (0.341% out of 879 total) detected and removed</span></span></code></pre>
<pre><code><span><span class="co">#&gt; 1 multiallelic sites (0.094% out of 1067 total) detected and removed</span></span></code></pre>
<p><strong>The total runtime for the ABC simulations was 0 hours 25
minutes 38 seconds parallelized across 96 CPUs.</strong></p>
<p>Once the simulations are finished, we can perform inference of the
posterior distribution of the single parameter of our model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>:</p>
<div class="sourceCode" id="cb1241"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_abc.html">run_abc</a></span><span class="op">(</span><span class="va">data</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; 12345678910</span></span>
<span><span class="co">#&gt; 12345678910</span></span></code></pre>
<div class="sourceCode" id="cb1243"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                               Ne</span></span>
<span><span class="co">#&gt; Min.:                   593.7036</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:   620.6549</span></span>
<span><span class="co">#&gt; Weighted Median:       1013.2286</span></span>
<span><span class="co">#&gt; Weighted Mean:         1008.2233</span></span>
<span><span class="co">#&gt; Weighted Mode:         1029.4001</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 1436.4719</span></span>
<span><span class="co">#&gt; Max.:                  1584.7489</span></span></code></pre>
<div class="sourceCode" id="cb1246"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">1000</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-05-custom_files/figure-html/custom_Ne_posterior3-1.png" width="640" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="computing-summary-statistics-with-python">Computing summary statistics with Python<a class="anchor" aria-label="anchor" href="#computing-summary-statistics-with-python"></a>
</h3>
<p>As a last example for customization of <em>demografr</em> inference
pipeline, let’s consider the scenario in which the tree-sequence
functionality available through <em>slendr</em> isn’t sufficient, and
you need to compute summary statistics on the simulated tree-sequence
using pure Python and the <em>tskit</em> module directly. As with the
other customization examples above, this is again very easy to do.</p>
<p>First, just to change things up, let’s say that we want to use a
normal <em>slendr</em> model function as a scaffold model for inference,
simulate a tree sequence with the standard <em>slendr</em> /
<em>demografr</em> functionality (i.e, not via the external Python
script as in the examples above), and only customize the summary
statistic computation of nucleotide diversity.</p>
<div class="sourceCode" id="cb1247"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ne</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># this will be a single-population coalescent model, so the time of</span></span>
<span>  <span class="co"># the appearance of the population is arbitrary -- let's say we</span></span>
<span>  <span class="co"># want the model to start at 1000 generations ago, just to pick an</span></span>
<span>  <span class="co"># arbitrary point in time</span></span>
<span>  <span class="va">pop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"pop"</span>, N <span class="op">=</span> <span class="va">Ne</span>, time <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>    <span class="va">pop</span>, generation_time <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    direction <span class="op">=</span> <span class="st">"backward"</span>,</span>
<span>    serialize <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="va">schedule</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/schedule_sampling.html" class="external-link">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">pop</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">schedule</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We will use the same prior and observed statistic as above:</p>
<div class="sourceCode" id="cb1248"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_diversity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"inst/examples/custom_diversity.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">Ne</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">observed_diversity</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb1249"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compute_diversity</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/py_run.html" class="external-link">py_run_string</a></span><span class="op">(</span><span class="st">r"(</span></span>
<span><span class="st">def compute_diversity(ts):</span></span>
<span><span class="st">    diffs = []</span></span>
<span><span class="st"></span></span>
<span><span class="st">    for i in range(ts.num_samples - 1):</span></span>
<span><span class="st">        for j in range(i + 1, ts.num_samples):</span></span>
<span><span class="st">            diffs.append(ts.divergence(sample_sets=[[i], [j]], mode="site"))</span></span>
<span><span class="st"></span></span>
<span><span class="st">    pi = sum(diffs) / len(diffs)</span></span>
<span><span class="st"></span></span>
<span><span class="st">    return pd.DataFrame({"set": ["pop"], "diversity": [pi]})</span></span>
<span><span class="st">)"</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="fu">reticulate</span><span class="fu">::</span><span class="va"><a href="https://rstudio.github.io/reticulate/reference/py.html" class="external-link">py</a></span><span class="op">$</span><span class="fu">compute_diversity</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">compute_diversity</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb1250"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_run</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_model.html">simulate_model</a></span><span class="op">(</span><span class="va">model</span>, parameters <span class="op">=</span> <span class="va">priors</span>, sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<p>When we look at the result, we can see that it no longer contains a
<code>ts</code> tree-sequence object but a simple table of genotypes
created via the <em>slendr</em> function <code><a href="https://rdrr.io/pkg/slendr/man/ts_genotypes.html" class="external-link">ts_genotypes()</a></code>
(see the data generating function just above).</p>
<div class="sourceCode" id="cb1251"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_run</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ╔═══════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence               ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤═══════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │        505║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│  1,000,000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │generations║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │        100║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │   96.9 KiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧═══════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤═════╤═════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows │Size     │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪═════╪═════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │1,989│ 62.2 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼─────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│   50│  1.4 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼─────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │    0│  8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼─────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │    0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼─────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │  564│ 15.4 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼─────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│    1│222 Bytes│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼─────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│    1│  1.4 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼─────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │    0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧═════╧═════════╧════════════╝</span></span></code></pre>
<p>Let’s test our summary function on the output of the single testing
simulation run:</p>
<div class="sourceCode" id="cb1253"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarise_data.html">summarise_data</a></span><span class="op">(</span><span class="va">model_run</span>, <span class="va">functions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span>
<span><span class="co">#&gt;   set diversity</span></span>
<span><span class="co">#&gt; 1 pop         0</span></span></code></pre>
<p>Let’s also validate our customized setup before we unleash the full
scale of ABC simulations on the problem!</p>
<div class="sourceCode" id="cb1255"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>,</span>
<span>             sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; Testing sampling of each prior parameter:</span></span>
<span><span class="co">#&gt;   - Ne ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; The model is a slendr function</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the return statement of the model function... ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the presence of required model function arguments...---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Simulating tree sequence from the given model... ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Computing user-defined summary functions:</span></span>
<span><span class="co">#&gt;   - diversity ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the format of simulated summary statistics:</span></span>
<span><span class="co">#&gt;   - diversity [data frame] ✅</span></span>
<span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; No issues have been found in the ABC setup!</span></span></code></pre>
<div class="sourceCode" id="cb1257"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span></span>
<span>  <span class="va">model</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>The total runtime for the ABC simulations was 0 hours 17
minutes 22 seconds parallelized across 96 CPUs.</strong></p>
<p>Once the simulations are finished, we can perform inference of the
posterior distribution of the single parameter of our model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>e</mi></msub><annotation encoding="application/x-tex">N_e</annotation></semantics></math>:</p>
<div class="sourceCode" id="cb1258"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_abc.html">run_abc</a></span><span class="op">(</span><span class="va">data</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; 12345678910</span></span>
<span><span class="co">#&gt; 12345678910</span></span></code></pre>
<div class="sourceCode" id="cb1260"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                               Ne</span></span>
<span><span class="co">#&gt; Min.:                   565.2685</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:   679.8289</span></span>
<span><span class="co">#&gt; Weighted Median:        979.1790</span></span>
<span><span class="co">#&gt; Weighted Mean:         1014.0002</span></span>
<span><span class="co">#&gt; Weighted Mode:          869.9794</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 1517.4734</span></span>
<span><span class="co">#&gt; Max.:                  1705.5757</span></span></code></pre>
<div class="sourceCode" id="cb1263"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">1000</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-05-custom_files/figure-html/custom_Ne_posterior4-1.png" width="640" style="display: block; margin: auto;"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>


<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({


    apiKey: 'a232fffe35f6bc68735023bb34d1fee2',
    indexName: 'bodkan',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
