<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Custom SLiM or Python simulations • demografr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Custom SLiM or Python simulations">
<meta property="og:description" content="demografr">
<meta property="og:image" content="https://bodkan.net/demografr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">demografr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette-01-basics.html">A basic ABC analysis</a>
    </li>
    <li>
      <a href="../articles/vignette-02-priors.html">Specifying prior distributions</a>
    </li>
    <li>
      <a href="../articles/vignette-03-grids.html">Support for grid simulations</a>
    </li>
    <li>
      <a href="../articles/vignette-04-parallelization.html">Parallelization options</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Custom SLiM or Python simulations</h1>
            
      
      
      <div class="hidden name"><code>vignette-05-custom-engines.Rmd</code></div>

    </div>

    
    
<p>By default, <em>demografr</em> uses the <a href="https://github.com/bodkan/slendr" class="external-link"><em>slendr</em></a> simulation framework for defining models and simulating data from them. But what if you need to do inference using your own scripts? Perhaps <em>slendr</em>’s opinionated interface doesn’t allow you to do all that you need to do, such as using all powerful simulation features and options of raw SLiM or msprime? This vignette explains how you can use any standard SLiM or msprime script which produces a tree sequence file as its output for standard <em>demografr</em> analysis.</p>
<p>First, let’s load <em>demografr</em> itself and also <em>slendr</em> (which, in this vignette, will serve only for analyzing simulated tree-sequence data, but not simulations themselves).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">demografr</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/init_env.html" class="external-link">init_env</a></span><span class="op">(</span>quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="restrictions-on-user-defined-simulation-script">Restrictions on user-defined simulation script<a class="anchor" aria-label="anchor" href="#restrictions-on-user-defined-simulation-script"></a>
</h2>
<p>In order to be able to use a custom SLiM or msprime script with <em>demografr</em>, the script must conform to a couple of rules:</p>
<div class="section level3">
<h3 id="it-must-be-runnable-on-the-command-line-as-any-other-command-line-script">1. It must be runnable on the command-line as any other command-line script<a class="anchor" aria-label="anchor" href="#it-must-be-runnable-on-the-command-line-as-any-other-command-line-script"></a>
</h3>
<ul>
<li><p><strong>For an msprime script</strong>, this means something like <code>python &lt;your script&gt;.py &lt;command-line arguments&gt;</code>.</p></li>
<li><p><strong>For a SLiM script</strong>, this means something like <code>slim &lt;command-line arguments&gt; &lt;your script&gt;.slim</code>.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="it-must-accept-three-mandatory-arguments-with-these-exact-names">2. It must accept three mandatory arguments with these exact names<a class="anchor" aria-label="anchor" href="#it-must-accept-three-mandatory-arguments-with-these-exact-names"></a>
</h3>
<ul>
<li>
<code>sequence_length</code>: the amount of sequence to simulated</li>
<li>
<code>recombination_rate</code>: the recombination rate along the simulated sequence (in units of crossovers per basepair pergeneration)</li>
<li>
<code>output_path</code>: the path where the simulated tree-sequence object will be saved</li>
</ul>
<p><strong>For an msprime script</strong>, these parameters can be specified via the Python built-in module <code>argparse</code>, and provided on the command-line as <code>--sequence_length &lt;value&gt;</code>, <code>--recombination_rate &lt;value&gt;</code>, and <code>--output_path &lt;path&gt;</code>. In the script itself, if you use the <code>argparse</code> module and have thus the values of provided arguments (for instance) in an <code>args</code> object, you can then refer to these arguments as <code>args.sequence_length</code>, etc.</p>
<p><strong>For a SLiM script</strong>, these parameters can be specified via SLiM’s standard way of supplying command-line arguments as <code>-d sequence_length=&lt;value&gt;</code>, <code>-d recombination_rate=&lt;value&gt;</code>, and <code>-d "output_path='&lt;path&gt;'"</code>. Importantly, note SLiM’s format of specifying string arguments for the <code>output_path</code> argument. If you need further detail, see Section 20.2 of the SLiM manual. In the script itself, you can then refer to these arguments via constants <code>sequence_length</code>, <code>recombination_rate</code>, and <code>output_path</code>.</p>
</div>
<div class="section level3">
<h3 id="all-model-parameters-must-be-provided-as-additional-command-line-arguments">3. All model parameters must be provided as additional command-line arguments<a class="anchor" aria-label="anchor" href="#all-model-parameters-must-be-provided-as-additional-command-line-arguments"></a>
</h3>
<p>You can refer to them in your script as you would to the mandatory arguments as described in the previous section.</p>
</div>
</div>
<div class="section level2">
<h2 id="abc-analysis-using-custom-defined-simulation-scripts">ABC analysis using custom-defined simulation scripts<a class="anchor" aria-label="anchor" href="#abc-analysis-using-custom-defined-simulation-scripts"></a>
</h2>
<div class="section level3">
<h3 id="toy-problem">Toy problem<a class="anchor" aria-label="anchor" href="#toy-problem"></a>
</h3>
<p>Suppose that we intent to use ABC to infer the <span class="math inline">\(N_e\)</span> of a constant-sized population given that we observed nucleotide diversity of:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_diversity</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   set   diversity</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop   0.000<span style="text-decoration: underline;">021</span>0</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="example-pure-slim-and-msprime-script">Example pure SLiM and msprime script<a class="anchor" aria-label="anchor" href="#example-pure-slim-and-msprime-script"></a>
</h2>
<p>To demonstrate the use of requirements 1-3 above, let’s perform ABC using each of those scripts as tree-sequence simulation engines.</p>
<p>Imagine we have the following two scripts which we want to use as simulation engines instead of <em>slendr</em>’s own functions <code><a href="https://rdrr.io/pkg/slendr/man/slim.html" class="external-link">slim()</a></code> and <code><a href="https://rdrr.io/pkg/slendr/man/msprime.html" class="external-link">msprime()</a></code>. The scripts have only <strong>one <em>model parameter</em>, which is the <span class="math inline">\(N_e\)</span> of the simulated populations</strong>. The *other parameters are mandatory, just discussed in the previous section.</p>
<div class="section level3">
<h3 id="slim-script">SLiM script<a class="anchor" aria-label="anchor" href="#slim-script"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">slim_script</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/custom.slim"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">slim_script</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code>initialize() {
  initializeTreeSeq();
  initializeMutationRate(0);
  initializeMutationType("m1", 0.5, "f", 0.0);
  initializeGenomicElementType("g1", m1, 1.0);
  initializeGenomicElement(g1, 0, sequence_length - 1);
  initializeRecombinationRate(recombination_rate);
}

1 early() {
  sim.addSubpop("p0", asInteger(Ne));
}

10000 late() {
  sim.treeSeqOutput(output_path);
}</code></pre>
</div>
<div class="section level3">
<h3 id="python-script">Python script<a class="anchor" aria-label="anchor" href="#python-script"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">py_script</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/custom.py"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">py_script</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code>import argparse

import msprime

parser = argparse.ArgumentParser()

# mandatory command-line arguments
parser.add_argument("--sequence_length", type=float, required=True)
parser.add_argument("--recombination_rate", type=float, required=True)
parser.add_argument("--output_path", type=str, required=True)

# model parameters
parser.add_argument("--Ne", type=float, required=True)

args = parser.parse_args()

ts = msprime.sim_ancestry(
  samples=round(args.Ne),
  population_size=args.Ne,
  sequence_length=args.sequence_length,
  recombination_rate=args.recombination_rate,
)

ts.dump(args.output_path)</code></pre>
</div>
<div class="section level3">
<h3 id="individual-abc-components">Individual ABC components<a class="anchor" aria-label="anchor" href="#individual-abc-components"></a>
</h3>
<p>Apart from the user-defined simulation SLiM and msprime scripts, the components of our toy inference remains the same—we need to define the observed statistics, tree-sequence summary functions, and priors. We don’t need a model function—that’s served by out custom scripts (which also serve as simulation engines).</p>
<p>Here are the components (we won’t be discussing them here because that’s extensively taken care of elsewhere in <em>demografr</em>’s documentation vignettes):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">Ne</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>diversity <span class="op">=</span> <span class="va">observed_diversity</span><span class="op">)</span></span>
<span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  diversity <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># compute diversity on 20 sampled chromosomes</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_diversity.html" class="external-link">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">19</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>(Note that because our simulated tree sequences won’t be coming with <em>slendr</em> metadata, we have to refer to individuals’ chromosomes using numerical indices rather than <em>slendr</em> symbolic names like we do in all of our other vignettes. This is something we hope to solve on <em>slendr</em>’s side at some point soon.)</p>
</div>
<div class="section level3">
<h3 id="simulating-a-testing-tree-sequence">Simulating a testing tree sequence<a class="anchor" aria-label="anchor" href="#simulating-a-testing-tree-sequence"></a>
</h3>
<p>As we explained elsewhere, a useful function for developing inference pipelines using <em>demografr</em> is a function <code><a href="../reference/simulate_ts.html">simulate_ts()</a></code>, which normally accepts a <em>slendr</em> model generating function and some parameters (either given as priors or as a list of named values), and simulates a tree-sequence object. This function (as any other <em>demografr</em> function operating with models) accepts our custom-defined simulation scripts in place of standard <em>slendr</em> models.</p>
<p>For instance, we can simulate a couple of Mb of sequence from our SLiM script like this:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts_slim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_ts.html">simulate_ts</a></span><span class="op">(</span><span class="va">slim_script</span>, parameters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Ne <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span>,</span>
<span>                       sequence_length <span class="op">=</span> <span class="fl">10e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<pre><code>&gt; Setting 'engine' to "custom" is only allowed with a user-defined 'model' script.
&gt; Setting slendr function as a 'model' is only allowed with "msprime" or "slim" as 'engine'.</code></pre>
<p>Oh no! We got an error! No reason to panic though—this is <em>demografr</em> being cautious in an effort to prevent the user from making a silly mistake.</p>
<p>Because <em>demografr</em> assumes that most users will be using <em>slendr</em>-defined models as inputs, if you want to use a custom-built SLiM or msprime model, you need to explicitly tell every simulation function (<code><a href="../reference/simulate_ts.html">simulate_ts()</a></code>, <code><a href="../reference/simulate_abc.html">simulate_abc()</a></code>, and <code><a href="../reference/simulate_grid.html">simulate_grid()</a></code>) that you intend to do this by setting <code>engine = "custom"</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_ts.html">simulate_ts</a></span><span class="op">(</span><span class="va">slim_script</span>, parameters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Ne <span class="op">=</span> <span class="fl">123</span><span class="op">)</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span>
<span><span class="va">ts_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_ts.html">simulate_ts</a></span><span class="op">(</span><span class="va">py_script</span>, parameters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Ne <span class="op">=</span> <span class="fl">123</span><span class="op">)</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<p>We can compare that we get a very similar tree sequence with both the SLiM and msprime engines:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts_s</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ╔════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence            ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │       1║</span></span>
<span><span class="co">#&gt; ╟───────────────┼────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│ 1000000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │   ticks║</span></span>
<span><span class="co">#&gt; ╟───────────────┼────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │     246║</span></span>
<span><span class="co">#&gt; ╟───────────────┼────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │65.0 KiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤════╤════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows│Size    │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪════╪════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │ 447│14.0 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│ 123│13.8 KiB│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │   0│ 8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │ 128│ 5.8 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │ 448│17.3 KiB│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│   1│ 2.3 KiB│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│   2│ 3.1 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │ 128│ 3.1 KiB│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧════╧════════╧════════════╝</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts_m</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ╔═══════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence               ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤═══════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │          1║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│    1000000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │generations║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │        246║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │   39.8 KiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧═══════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤════╤═════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows│Size     │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪════╪═════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │ 490│ 15.3 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│ 123│  3.4 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │   0│  8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │  32│  1.2 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │ 491│ 13.4 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│   1│224 Bytes│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│   2│  1.6 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │  32│816 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧════╧═════════╧════════════╝</span></span></code></pre>
<p>Now we can finally test our little toy tree-sequence summary functions, verifying that we can indeed compute the summary statistic we want to:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">functions</span><span class="op">$</span><span class="fu">diversity</span><span class="op">(</span><span class="va">ts_s</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop   0.000<span style="text-decoration: underline;">003</span>42</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">functions</span><span class="op">$</span><span class="fu">diversity</span><span class="op">(</span><span class="va">ts_m</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> pop   0.000<span style="text-decoration: underline;">006</span>94</span></span></code></pre>
<p>The functions works as expected—we have a single population and want to compute nucleotide diversity in the whole population, and this is exactly what we get.</p>
</div>
<div class="section level3">
<h3 id="running-an-abc-analysis">Running an ABC analysis<a class="anchor" aria-label="anchor" href="#running-an-abc-analysis"></a>
</h3>
<p>Having all components of our pipeline set up we should, again, validate everything before we proceed to (potentially very costly) simulations. In the remainder of this vignette we’ll only continue with the Python msprime custom script, in order to save some computational time. That said, it’s important to realize that you could use this sort of workflow for any kind of SLiM script, including very elaborate spatial simulations, non-WF models, and all kinds of phenotypic simulations too!</p>
<p>Let’s first validate all components of our pipeline:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">py_script</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; A model generating function was provided as a scaffold</span></span>
<span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; Testing sampling of each prior parameter:</span></span>
<span><span class="co">#&gt;   * Ne ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; The model is a custom user-defined msprime script</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Simulating tree sequence from the given model... ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Computing user-defined summary functions:</span></span>
<span><span class="co">#&gt;   * diversity ✅</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Checking the format of simulated summary statistics:</span></span>
<span><span class="co">#&gt;   * diversity ✅  </span></span>
<span><span class="co">#&gt; ======================================================================</span></span>
<span><span class="co">#&gt; No issues have been found in the ABC setup!</span></span></code></pre>
<p>Looking good! Now let’s first run ABC simulations. Note the use of the script where we would normally plug in a slendr model function in place of the <code>model</code> argument:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multicore</span>, workers <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="co"># parallelize across 50 CPUs</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span></span>
<span>  model <span class="op">=</span> <span class="va">py_script</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Once the simulations are finished, we can perform inference of the posterior distribution of the single parameter of our model, <span class="math inline">\(N_e\)</span>:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform_abc.html">perform_abc</a></span><span class="op">(</span><span class="va">data</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.03</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; 12345678910</span></span>
<span><span class="co">#&gt; 12345678910</span></span></code></pre>
<p>Having done so, we can again look at the summary statistics of the posterior distribution and also plot the results:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;                              Ne</span></span>
<span><span class="co">#&gt; Min.:                  258.3854</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  312.4582</span></span>
<span><span class="co">#&gt; Weighted Median:       546.4880</span></span>
<span><span class="co">#&gt; Weighted Mean:         560.5439</span></span>
<span><span class="co">#&gt; Weighted Mode:         543.5098</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 869.9848</span></span>
<span><span class="co">#&gt; Max.:                  980.6594</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-05-custom-engines_files/figure-html/custom_Ne_posterior-1.png" width="640" style="display: block; margin: auto;"></p>
<p>Again, in addition to <em>demografr</em>’s visualization functions (like <code><a href="../reference/plot_posterior.html">plot_posterior()</a></code> above) we have the full range of diagnostic plotting functions bundled with the <em>abc</em> R package at our disposal:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-05-custom-engines_files/figure-html/custom_Ne_diagnostic-1.png" width="640" style="display: block; margin: auto;"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
