<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Diagnostics, model selection, troubleshooting • demografr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Diagnostics, model selection, troubleshooting">
<meta property="og:description" content="demografr">
<meta property="og:image" content="https://bodkan.net/demografr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">demografr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette-01-basics.html">A basic ABC workflow</a>
    </li>
    <li>
      <a href="../articles/vignette-02-priors.html">Specifying prior distributions</a>
    </li>
    <li>
      <a href="../articles/vignette-03-grids.html">Support for grid simulations</a>
    </li>
    <li>
      <a href="../articles/vignette-04-parallelization.html">Parallelization options</a>
    </li>
    <li>
      <a href="../articles/vignette-05-custom.html">Custom inference using SLiM or Python</a>
    </li>
    <li>
      <a href="../articles/vignette-06-diagnostics.html">Model selection and other diagnostics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/demografr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Diagnostics, model selection, troubleshooting</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/demografr/blob/HEAD/vignettes/vignette-06-diagnostics.Rmd" class="external-link"><code>vignettes/vignette-06-diagnostics.Rmd</code></a></small>
      <div class="hidden name"><code>vignette-06-diagnostics.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/demografr" class="external-link">demografr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/init_env.html" class="external-link">init_env</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; The interface to all required Python modules has been activated.</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SEED</span> <span class="op">&lt;-</span> <span class="fl">42</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span></span></code></pre></div>
<p>⚠️⚠️⚠️</p>
<p><strong>Note:</strong> The <em>demografr</em> R package is still under active development. As a result, its documentation is in a draft stage at best. Typos, inconsistencies, and other issues are unfortunately expected.</p>
<p>⚠️⚠️⚠</p>
<p>Let’s return to our <a href="vignette-01-basics.html">first</a> example. However, this time, imagine that we don’t really know which of the three following phylogenetic relationships is the one that captures the features of our data best, perhaps with different sources of evidence being consistent with each one of them so we want to consider all three in our modeling (as always, this is supposed to be a hypothetical example for demonstration purposes). These models differ both in terms of tree topology of the population relationships as well as in the gene flow, and we want to perform <em>model selection</em>.</p>
<p><img src="vignette-06-diagnostics_files/figure-html/ape_tree_modelX-1.png" width="400" style="display: block; margin: auto;"></p>
<p>For completeness, here is again our observed data which was, in reality, generated by a hidden evolutionary process whose parameters we’re trying to infer by doing ABC:</p>
<ol style="list-style-type: decimal">
<li>Nucleotide diversity in each population:</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_diversity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/basics_diversity.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_diversity</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt; 1   A 8.030512e-05</span></span>
<span><span class="co">#&gt; 2   B 3.288576e-05</span></span>
<span><span class="co">#&gt; 3   C 1.013804e-04</span></span>
<span><span class="co">#&gt; 4   D 8.910909e-05</span></span></code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Pairwise divergence d_X_Y between populations X and Y:</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_divergence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/basics_divergence.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_divergence</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   x y   divergence</span></span>
<span><span class="co">#&gt; 1 A B 0.0002378613</span></span>
<span><span class="co">#&gt; 2 A C 0.0002375761</span></span>
<span><span class="co">#&gt; 3 A D 0.0002379385</span></span>
<span><span class="co">#&gt; 4 B C 0.0001088217</span></span>
<span><span class="co">#&gt; 5 B D 0.0001157056</span></span>
<span><span class="co">#&gt; 6 C D 0.0001100633</span></span></code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Value of the following <span class="math inline">\(f_4\)</span>-statistic:</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_f4</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/basics_f4.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_f4</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   W X Y Z            f4</span></span>
<span><span class="co">#&gt; 1 A B C D -3.262146e-06</span></span></code></pre>
<p>We will again bind them into a list:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  diversity  <span class="op">=</span> <span class="va">observed_diversity</span>,</span>
<span>  divergence <span class="op">=</span> <span class="va">observed_divergence</span>,</span>
<span>  f4 <span class="op">=</span> <span class="va">observed_f4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="three-competing-models">Three competing models<a class="anchor" aria-label="anchor" href="#three-competing-models"></a>
</h3>
<p>First, in order to perform model selection, we need to specify models themselves. We do this by defining three separate <em>slendr</em> functions, each of them encoding the three population relationships from the diagrams above. Note that we’re not trying to infer the time of gene flow, for simplicity, but we do fix the gene flow event to different times (again, reflected in the alternative model sketches above), just to make the models stand out from one another:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">modelX</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ne_A</span>, <span class="va">Ne_B</span>, <span class="va">Ne_C</span>, <span class="va">Ne_D</span>, <span class="va">T_1</span>, <span class="va">T_2</span>, <span class="va">T_3</span>, <span class="va">gf</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">popA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popA"</span>, time <span class="op">=</span> <span class="fl">1</span>,   N <span class="op">=</span> <span class="va">Ne_A</span><span class="op">)</span></span>
<span>  <span class="va">popB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popB"</span>, time <span class="op">=</span> <span class="va">T_1</span>, N <span class="op">=</span> <span class="va">Ne_B</span>, parent <span class="op">=</span> <span class="va">popA</span><span class="op">)</span></span>
<span>  <span class="va">popC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popC"</span>, time <span class="op">=</span> <span class="va">T_2</span>, N <span class="op">=</span> <span class="va">Ne_C</span>, parent <span class="op">=</span> <span class="va">popB</span><span class="op">)</span></span>
<span>  <span class="va">popD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popD"</span>, time <span class="op">=</span> <span class="va">T_3</span>, N <span class="op">=</span> <span class="va">Ne_D</span>, parent <span class="op">=</span> <span class="va">popC</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">popB</span>, to <span class="op">=</span> <span class="va">popC</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="va">gf</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popA</span>, <span class="va">popB</span>, <span class="va">popC</span>, <span class="va">popD</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>    generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    direction <span class="op">=</span> <span class="st">"forward"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/schedule_sampling.html" class="external-link">schedule_sampling</a></span><span class="op">(</span></span>
<span>    <span class="va">model</span>, times <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popA</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popB</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popC</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popD</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">modelY</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ne_A</span>, <span class="va">Ne_B</span>, <span class="va">Ne_C</span>, <span class="va">Ne_D</span>, <span class="va">T_1</span>, <span class="va">T_2</span>, <span class="va">T_3</span>, <span class="va">gf</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">popA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popA"</span>, time <span class="op">=</span> <span class="fl">1</span>,   N <span class="op">=</span> <span class="va">Ne_A</span><span class="op">)</span></span>
<span>  <span class="va">popB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popB"</span>, time <span class="op">=</span> <span class="va">T_1</span>, N <span class="op">=</span> <span class="va">Ne_B</span>, parent <span class="op">=</span> <span class="va">popA</span><span class="op">)</span></span>
<span>  <span class="va">popC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popC"</span>, time <span class="op">=</span> <span class="va">T_2</span>, N <span class="op">=</span> <span class="va">Ne_C</span>, parent <span class="op">=</span> <span class="va">popA</span><span class="op">)</span></span>
<span>  <span class="va">popD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popD"</span>, time <span class="op">=</span> <span class="va">T_3</span>, N <span class="op">=</span> <span class="va">Ne_D</span>, parent <span class="op">=</span> <span class="va">popA</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">popA</span>, to <span class="op">=</span> <span class="va">popB</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="va">gf</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popA</span>, <span class="va">popB</span>, <span class="va">popC</span>, <span class="va">popD</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>    generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    direction <span class="op">=</span> <span class="st">"forward"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/schedule_sampling.html" class="external-link">schedule_sampling</a></span><span class="op">(</span></span>
<span>    <span class="va">model</span>, times <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popA</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popB</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popC</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popD</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">modelZ</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ne_A</span>, <span class="va">Ne_B</span>, <span class="va">Ne_C</span>, <span class="va">Ne_D</span>, <span class="va">T_1</span>, <span class="va">T_2</span>, <span class="va">T_3</span>, <span class="va">gf</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">popA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popA"</span>, time <span class="op">=</span> <span class="fl">1</span>,   N <span class="op">=</span> <span class="va">Ne_A</span><span class="op">)</span></span>
<span>  <span class="va">popB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popB"</span>, time <span class="op">=</span> <span class="va">T_1</span>, N <span class="op">=</span> <span class="va">Ne_B</span>, parent <span class="op">=</span> <span class="va">popA</span><span class="op">)</span></span>
<span>  <span class="va">popC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popC"</span>, time <span class="op">=</span> <span class="va">T_2</span>, N <span class="op">=</span> <span class="va">Ne_C</span>, parent <span class="op">=</span> <span class="va">popA</span><span class="op">)</span></span>
<span>  <span class="va">popD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popD"</span>, time <span class="op">=</span> <span class="va">T_3</span>, N <span class="op">=</span> <span class="va">Ne_D</span>, parent <span class="op">=</span> <span class="va">popC</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">popC</span>, to <span class="op">=</span> <span class="va">popD</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="va">gf</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popA</span>, <span class="va">popB</span>, <span class="va">popC</span>, <span class="va">popD</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>    generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    direction <span class="op">=</span> <span class="st">"forward"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/schedule_sampling.html" class="external-link">schedule_sampling</a></span><span class="op">(</span></span>
<span>    <span class="va">model</span>, times <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popA</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popB</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popC</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popD</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now, let’s <a href="vignette-02-priors.html">specify priors</a> using <em>demografr</em>’s <a href="vignette-02-priors.html#prior-parameter-templates">templating syntax</a>. This saves us a bit of typing, making the prior definition code a bit more consise and easier to read:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">Ne...</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">T_1</span>   <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>,    <span class="fl">4000</span><span class="op">)</span>,</span>
<span>  <span class="va">T_2</span>   <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">9000</span><span class="op">)</span>,</span>
<span>  <span class="va">T_3</span>   <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">gf</span>    <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Let’s also put together a list of tree-sequence summary functions and observed summary statistics:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compute_diversity</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_diversity.html" class="external-link">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">compute_divergence</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_divergence.html" class="external-link">ts_divergence</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">compute_f4</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="va">A</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"popA"</span><span class="op">]</span>; <span class="va">B</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"popB"</span><span class="op">]</span></span>
<span>  <span class="va">C</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"popC"</span><span class="op">]</span>; <span class="va">D</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"popD"</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_f4ratio.html" class="external-link">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">A</span>, <span class="va">B</span>, <span class="va">C</span>, <span class="va">D</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  diversity <span class="op">=</span> <span class="va">compute_diversity</span>,</span>
<span>  divergence <span class="op">=</span> <span class="va">compute_divergence</span>,</span>
<span>  f4 <span class="op">=</span> <span class="va">compute_f4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Let’s validate the ABC setup of all three models – this is an important check that the <em>slendr</em> model functions are defined correctly (we set <code>quiet = TRUE</code> to surpress writing out the full log output):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/validate_abc.html" class="external-link">validate_abc</a></span><span class="op">(</span><span class="va">modelX</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/validate_abc.html" class="external-link">validate_abc</a></span><span class="op">(</span><span class="va">modelY</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/validate_abc.html" class="external-link">validate_abc</a></span><span class="op">(</span><span class="va">modelZ</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>With that out of the way, we can proceed with generating simulated data for inference using all three models. What we’ll do is perform three runs and save them into appropriately named variables <code>dataX</code>, <code>dataY</code>, and <code>dataZ</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/simulate_abc.html" class="external-link">simulate_abc</a></span><span class="op">(</span><span class="va">modelX</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                      sequence_length <span class="op">=</span> <span class="fl">10e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/simulate_abc.html" class="external-link">simulate_abc</a></span><span class="op">(</span><span class="va">modelY</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                      sequence_length <span class="op">=</span> <span class="fl">10e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/simulate_abc.html" class="external-link">simulate_abc</a></span><span class="op">(</span><span class="va">modelZ</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                      sequence_length <span class="op">=</span> <span class="fl">10e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<p><strong>The total runtime for the ABC simulations was 1 hours 11 minutes 25 seconds parallelized across 96 CPUs.</strong></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abcX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/run_abc.html" class="external-link">run_abc</a></span><span class="op">(</span><span class="va">dataX</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span>
<span><span class="va">abcY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/run_abc.html" class="external-link">run_abc</a></span><span class="op">(</span><span class="va">dataY</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span>
<span><span class="va">abcZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/run_abc.html" class="external-link">run_abc</a></span><span class="op">(</span><span class="va">dataZ</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cross-validation">Cross-validation<a class="anchor" aria-label="anchor" href="#cross-validation"></a>
</h2>
<p>Before doing model selection, it’s important to perform cross-validation to answer the question whether our ABC setup can even distinguish between the competing models. Without having the power to do this, trying to select the best model wouldn’t make much sense.</p>
<p>This can be done using <em>demografr</em>’s <code><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate()</a></code> function which is built around <em>abc</em>’s own function <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code>. We will not go into too much detail, as this function simply calls <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code> under the hood, passing to it all specified function arguments on behalf of a user to avoid unnecessary manual data munging (creating the <code>index</code> vector, merging the summary statistics into <code>sumstat</code> parameter, etc.). For more details, read section “Model selection” in the <a href="https://cran.r-project.org/package=abc/vignettes/abcvignette.pdf" class="external-link">vignette</a> of the <em>abc</em> R package.</p>
<p>The one difference between the two functions is that <code><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate()</a></code> removes the need to prepare character indices and bind together summary statistic matrices from different models—given that <em>demografr</em>’s ABC output objects track all this information along in their internals, this is redundant, and you can perform cross-validation of different ABC models simply by calling this:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">abcX</span>, <span class="va">abcY</span>, <span class="va">abcZ</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cv_selection</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate</a></span><span class="op">(</span><span class="va">models</span>, nval <span class="op">=</span> <span class="fl">100</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<p>If we print out the result, we get a quick summary of the resulting confusion matrices and other diagnostic information:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv_selection</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Confusion matrix based on 100 samples for each model.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX     95      0      5</span></span>
<span><span class="co">#&gt; modelY      0     98      2</span></span>
<span><span class="co">#&gt; modelZ      0      1     99</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Mean model posterior probabilities (neuralnet)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX 0.8877 0.0597 0.0526</span></span>
<span><span class="co">#&gt; modelY 0.0467 0.9336 0.0197</span></span>
<span><span class="co">#&gt; modelZ 0.0097 0.0237 0.9667</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $conf.matrix</span></span>
<span><span class="co">#&gt; $conf.matrix$tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX     95      0      5</span></span>
<span><span class="co">#&gt; modelY      0     98      2</span></span>
<span><span class="co">#&gt; modelZ      0      1     99</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $probs</span></span>
<span><span class="co">#&gt; $probs$tol0.01</span></span>
<span><span class="co">#&gt;             modelX     modelY     modelZ</span></span>
<span><span class="co">#&gt; modelX 0.887732255 0.05966634 0.05260140</span></span>
<span><span class="co">#&gt; modelY 0.046703463 0.93359347 0.01970306</span></span>
<span><span class="co">#&gt; modelZ 0.009655459 0.02368555 0.96665899</span></span></code></pre>
<p>Similarly, you can use the <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> function to visualize the result. This function, yet again, internally calls <em>abc</em>’s own plotting method internall, with a bonus option to save a figure to a PDF right from the <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> call (useful when working on a remote server):</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cv_selection</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/plot_cv-1.png" width="640" style="display: block; margin: auto;"></p>
<p>Because we have three models, each of the three barplots shows how often were summary statistics sampled from each model classified as likely coming from one of the three models. In other words, with absolutely perfect classification, each barplot would show just one of the three colors. If a barplot (results for one model) shows multiple colors, this means that some fraction of simulated statistics from that model was incorrectly classified as another model. Again, for more detail on interpretation, caveats, and best practices, please consult the <em>abc</em> R package <a href="https://cran.r-project.org/package=abc/vignettes/abcvignette.pdf" class="external-link">vignette</a> and a relevant statistical textbook.</p>
<p>The confusion matrices and the visualization all suggest that ABC can distinguish between the three models very well. For instance, <em>modelX</em> has been classified correctly in 95 simulations out of the total of 100 cross-validation simulations, with an overall misclassification rate for all models of only 2.7%.</p>
</div>
<div class="section level2">
<h2 id="model-selection">Model selection<a class="anchor" aria-label="anchor" href="#model-selection"></a>
</h2>
<p>Armed with confidence in the ability of ABC to correctly identify the correct model based on simulated data, we can proceed to selection of the best model for our empirical data set. This can be done with the function <code><a href="https://rdrr.io/pkg/demografr/man/select_model.html" class="external-link">select_model()</a></code> which is <em>demografr</em>’s convenience wrapper around <em>abc</em>’s own function <code>postpr</code>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">abcX</span>, <span class="va">abcY</span>, <span class="va">abcZ</span><span class="op">)</span></span>
<span></span>
<span><span class="va">modsel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/select_model.html" class="external-link">select_model</a></span><span class="op">(</span><span class="va">models</span>, tol <span class="op">=</span> <span class="fl">0.03</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<p>We can make a decision on the model selection by inspecting the <code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code> of the produced result:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">modsel</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Call: </span></span>
<span><span class="co">#&gt; abc::postpr(target = parts$target, index = parts$index, sumstat = parts$sumstat, </span></span>
<span><span class="co">#&gt;     tol = 0.03, method = "neuralnet")</span></span>
<span><span class="co">#&gt; Data:</span></span>
<span><span class="co">#&gt;  postpr.out$values (900 posterior samples)</span></span>
<span><span class="co">#&gt; Models a priori:</span></span>
<span><span class="co">#&gt;  modelX, modelY, modelZ</span></span>
<span><span class="co">#&gt; Models a posteriori:</span></span>
<span><span class="co">#&gt;  modelX, modelY, modelZ</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Proportion of accepted simulations (rejection):</span></span>
<span><span class="co">#&gt; modelX modelY modelZ </span></span>
<span><span class="co">#&gt; 0.8711 0.0056 0.1233 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Bayes factors:</span></span>
<span><span class="co">#&gt;          modelX   modelY   modelZ</span></span>
<span><span class="co">#&gt; modelX   1.0000 156.8000   7.0631</span></span>
<span><span class="co">#&gt; modelY   0.0064   1.0000   0.0450</span></span>
<span><span class="co">#&gt; modelZ   0.1416  22.2000   1.0000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Posterior model probabilities (neuralnet):</span></span>
<span><span class="co">#&gt; modelX modelY modelZ </span></span>
<span><span class="co">#&gt;      1      0      0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Bayes factors:</span></span>
<span><span class="co">#&gt;             modelX      modelY      modelZ</span></span>
<span><span class="co">#&gt; modelX      1.0000 333035.2444 523080.6411</span></span>
<span><span class="co">#&gt; modelY      0.0000      1.0000      1.5706</span></span>
<span><span class="co">#&gt; modelZ      0.0000      0.6367      1.0000</span></span></code></pre>
<p>As we can see, <code>modelX</code> shows the highest rate of acceptance among all simulations. Specifically, we see that the proportion of acceptance is 87.1%.</p>
<p>Similarly, looking at the Bayes factors, we see that the “modelX” is 156.8 times more likely than “modelY”, and 7.1 more likely than “modelZ”.</p>
<p>When the posterior probabilities are computed using a more elaborate neural network method (again, see more details in the <em>abc</em> <a href="https://cran.r-project.org/package=abc/vignettes/abcvignette.pdf" class="external-link">vignette</a>), we find that the probability of “modelX” is even more overwhelming. In fact, the analysis shows that it has a 100% probabiliy of being the correct model to explain our data.</p>
<p><strong>But how accurate are these conclusions?</strong> Well, if we take a peek at the <em>slendr</em> model which was <a href="https://github.com/bodkan/demografr/blob/main/vignettes/vignette-01-basics.Rmd#L36-L4" class="external-link">internally used</a> to generate the “observed” summary statistics, we see that the data was indeed simulated by code nearly identical to the one shown above as <code>modelX</code>. When plotted, the true model looks like this:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">popA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popA"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">popB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popB"</span>, time <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="fl">800</span>, parent <span class="op">=</span> <span class="va">popA</span><span class="op">)</span></span>
<span><span class="va">popC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popC"</span>, time <span class="op">=</span> <span class="fl">6000</span>, N <span class="op">=</span> <span class="fl">9000</span>, parent <span class="op">=</span> <span class="va">popB</span><span class="op">)</span></span>
<span><span class="va">popD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"popD"</span>, time <span class="op">=</span> <span class="fl">8000</span>, N <span class="op">=</span> <span class="fl">4000</span>, parent <span class="op">=</span> <span class="va">popC</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">popB</span>, to <span class="op">=</span> <span class="va">popC</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">example_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">popA</span>, <span class="va">popB</span>, <span class="va">popC</span>, <span class="va">popD</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>  generation_time <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  simulation_length <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/plot_model.html" class="external-link">plot_model</a></span><span class="op">(</span><span class="va">example_model</span>, proportions <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/true_model-1.png" width="640" style="display: block; margin: auto;"></p>
<p>It appears that we can be quite confident in which of the three models is most compatible with the data. However, before we move on to inferring parameters of the model, we should check whether the best selected model can indeed capture the most important features of the data, which is in case of ABC represented by our summary statistics.</p>
<p>In order to do this, <em>demografr</em> provides a function call <code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code> which performs posterior predictive check. Below, we we also look at how to use functions <code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code> implemented by the <em>abc</em> R package itself.</p>
<p>Let’s start with the posterior predictive checks.</p>
</div>
<div class="section level2">
<h2 id="posterior-predictive-checks">Posterior predictive checks<a class="anchor" aria-label="anchor" href="#posterior-predictive-checks"></a>
</h2>
<p>Link: <a href="http://www.stat.columbia.edu/~gelman/research/published/A6n41.pdf" class="external-link uri">http://www.stat.columbia.edu/~gelman/research/published/A6n41.pdf</a></p>
<p>It is one thing to Simply speaking, when fitting a model and estimating its parameters using ABC, we rely on a set of summary statistics, which we assume capture the most important features of the true model represented by the data we gathered. Performing posterior predictive checks involves estimating the posterior distributions of the parameters of interest from the data (i.e. doing ABC inference) – something we’ve done above on our <code>modelX</code>, <code>modelY</code>, and <code>modelZ</code> – and then sampling parameters from these distributions and generating distributions of simulated summary statistics from those estimated models. Then, we plot the distributions of each of our <em>simulated</em> summary statistics together with a vertical line representing the values of the <em>observed</em> statistics. Naturally, only if the model can generate statistics compatible with the values observed in real data can we regard the model as sensible and usable for inference.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">predX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">abcX</span>, samples <span class="op">=</span> <span class="fl">1000</span>, posterior <span class="op">=</span> <span class="st">"unadj"</span><span class="op">)</span></span>
<span><span class="va">predY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">abcY</span>, samples <span class="op">=</span> <span class="fl">1000</span>, posterior <span class="op">=</span> <span class="st">"unadj"</span><span class="op">)</span></span>
<span><span class="va">predZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">abcZ</span>, samples <span class="op">=</span> <span class="fl">1000</span>, posterior <span class="op">=</span> <span class="st">"unadj"</span><span class="op">)</span></span></code></pre></div>
<p>Looking at the posterior predictive plots from the <em>model X</em>, we can see that the observed statistics (vertical dashed lines) fall very well within the distributions of statistics simulated from the posterior parameters of the model. This suggests that are model (simplified that it must be given the more complex reality), captures the empoirical statistics quite well:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predX-1.png" width="640" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $divergence</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predX-2.png" width="640" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $f4</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predX-3.png" width="640" style="display: block; margin: auto;"></p>
<p>On the other hand, looking at the <em>model Z</em>, for instance, we see that although the <span class="math inline">\(f_4\)</span> gene-flow detection statistic fits rather nicely, the pairwise divergences between populations are completely from the statistics simulated from posterior distributions. This strongly suggests that the model wasn’t able to capture the features of the data.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predY</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predY-1.png" width="640" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $divergence</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predY-2.png" width="640" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $f4</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predY-3.png" width="640" style="display: block; margin: auto;"></p>
<p>We can compare the divergences between models directly to see which model gives the most reasonable posterior predictive check results more clearly:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predX</span>, <span class="st">"divergence"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predY</span>, <span class="st">"divergence"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predZ</span>, <span class="st">"divergence"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/pred_divergence-1.png" width="640" style="display: block; margin: auto;"></p>
<p>We can see that posterior predictive checks for nucleotide diversity are a little less clear as there are no extremely dramatic outliers as it’s the case for divergence above:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predX</span>, <span class="st">"diversity"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predY</span>, <span class="st">"diversity"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predZ</span>, <span class="st">"diversity"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/pred_diversity-1.png" width="640" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predX</span>, <span class="st">"f4"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predY</span>, <span class="st">"f4"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_prediction.html" class="external-link">plot_prediction</a></span><span class="op">(</span><span class="va">predZ</span>, <span class="st">"f4"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/pred_f4-1.png" width="640" style="display: block; margin: auto;"></p>
<p>If more customization or a more detailed analyses is needed, you can also extract data from the posterior simulations themselves:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/extract_prediction.html" class="external-link">extract_prediction</a></span><span class="op">(</span><span class="va">predX</span>, <span class="st">"diversity"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4,000 × 12</span></span></span>
<span><span class="co">#&gt;      rep   T_1   T_2   T_3    gf  Ne_A  Ne_B  Ne_C  Ne_D summary   stat    value</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1  107. <span style="text-decoration: underline;">4</span>433. <span style="text-decoration: underline;">5</span>238. 0.171  474.  236. <span style="text-decoration: underline;">9</span>856. <span style="text-decoration: underline;">2</span>020. diversity dive… 2.11<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     1  107. <span style="text-decoration: underline;">4</span>433. <span style="text-decoration: underline;">5</span>238. 0.171  474.  236. <span style="text-decoration: underline;">9</span>856. <span style="text-decoration: underline;">2</span>020. diversity dive… 9.61<span style="color: #949494;">e</span><span style="color: #BB0000;">-6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     1  107. <span style="text-decoration: underline;">4</span>433. <span style="text-decoration: underline;">5</span>238. 0.171  474.  236. <span style="text-decoration: underline;">9</span>856. <span style="text-decoration: underline;">2</span>020. diversity dive… 1.04<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     1  107. <span style="text-decoration: underline;">4</span>433. <span style="text-decoration: underline;">5</span>238. 0.171  474.  236. <span style="text-decoration: underline;">9</span>856. <span style="text-decoration: underline;">2</span>020. diversity dive… 6.69<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     2 <span style="text-decoration: underline;">1</span>172. <span style="text-decoration: underline;">8</span>409. <span style="text-decoration: underline;">9</span>205. 0.975 <span style="text-decoration: underline;">1</span>872. <span style="text-decoration: underline;">1</span>480. <span style="text-decoration: underline;">3</span>211. <span style="text-decoration: underline;">4</span>048. diversity dive… 7.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     2 <span style="text-decoration: underline;">1</span>172. <span style="text-decoration: underline;">8</span>409. <span style="text-decoration: underline;">9</span>205. 0.975 <span style="text-decoration: underline;">1</span>872. <span style="text-decoration: underline;">1</span>480. <span style="text-decoration: underline;">3</span>211. <span style="text-decoration: underline;">4</span>048. diversity dive… 5.55<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     2 <span style="text-decoration: underline;">1</span>172. <span style="text-decoration: underline;">8</span>409. <span style="text-decoration: underline;">9</span>205. 0.975 <span style="text-decoration: underline;">1</span>872. <span style="text-decoration: underline;">1</span>480. <span style="text-decoration: underline;">3</span>211. <span style="text-decoration: underline;">4</span>048. diversity dive… 7.49<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     2 <span style="text-decoration: underline;">1</span>172. <span style="text-decoration: underline;">8</span>409. <span style="text-decoration: underline;">9</span>205. 0.975 <span style="text-decoration: underline;">1</span>872. <span style="text-decoration: underline;">1</span>480. <span style="text-decoration: underline;">3</span>211. <span style="text-decoration: underline;">4</span>048. diversity dive… 7.64<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     3 <span style="text-decoration: underline;">2</span>833. <span style="text-decoration: underline;">5</span>046. <span style="text-decoration: underline;">9</span>849. 0.234 <span style="text-decoration: underline;">1</span>785.  752. <span style="text-decoration: underline;">7</span>788. <span style="text-decoration: underline;">3</span>338. diversity dive… 7.78<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>     3 <span style="text-decoration: underline;">2</span>833. <span style="text-decoration: underline;">5</span>046. <span style="text-decoration: underline;">9</span>849. 0.234 <span style="text-decoration: underline;">1</span>785.  752. <span style="text-decoration: underline;">7</span>788. <span style="text-decoration: underline;">3</span>338. diversity dive… 2.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3,990 more rows</span></span></span></code></pre>
<p>However, note that <code>extract_prediction</code> is a convenience function which only unnests, reformats, and renames the list-columns with summary statistics stored in the <code>predictions</code> data frame above. You can do all the processing based on the <code>predictions</code> object above.</p>
</div>
<div class="section level2">
<h2 id="additional-diagnostics-and-backwards-compatibility-with-the-abc-package">Additional diagnostics and backwards compatibility with the <em>abc</em> package<a class="anchor" aria-label="anchor" href="#additional-diagnostics-and-backwards-compatibility-with-the-abc-package"></a>
</h2>
<p>As mentioned above, the functionality behind <code><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate()</a></code> is implemented by reformatting the results of <em>demografr</em> simulation and inference functionality in a form necessary to run the functions of the <em>abc</em> package <code><a href="https://rdrr.io/pkg/abc/man/cv4abc.html" class="external-link">cv4abc()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code> behind the scenes – a task that normally requires a significant amount of potentially error-prone bookkeeping on part of the user.</p>
<p>In addition to <em>demografr</em>’s own function <code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code> which implements posterior predictive checks, the <em>abc</em> package provides additional functionality for evaluating the quality of the fit of an ABC model to the data, namely <code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code>. In this section we demonstrate how this functionality (and potentially other future <em>abc</em> functions) can be applied on <em>demografr</em> data with very little additional work.</p>
<p>First, <em>demografr</em> provides a function <code><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack()</a></code> which – when applied on an object generated either by <code><a href="https://rdrr.io/pkg/demografr/man/simulate_abc.html" class="external-link">simulate_abc()</a></code> or <code><a href="https://rdrr.io/pkg/demografr/man/run_abc.html" class="external-link">run_abc()</a></code> – <em>unpacks</em> the object encapsulating all information describing the simulation and inference which produced it, into individual components with names matching the “low-level” function arguments of all relevant functions of the <em>anc</em> package. This includes <code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code> mentioned above, but also functions <code><a href="https://rdrr.io/pkg/abc/man/cv4abc.html" class="external-link">cv4abc()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code>. Here is how <code><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack()</a></code> works.</p>
<div class="section level3">
<h3 id="unpacking-a-demografr-object-to-low-level-components-used-by-the-abc-package">“Unpacking” a <em>demografr</em> object to low-level components used by the <em>abc</em> package<a class="anchor" aria-label="anchor" href="#unpacking-a-demografr-object-to-low-level-components-used-by-the-abc-package"></a>
</h3>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">partsX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack</a></span><span class="op">(</span><span class="va">abcX</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">partsX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">partsX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "sumstat" "index"   "param"   "target"</span></span></code></pre>
<p>As we can see, the <code><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack()</a></code> function produced a list object with several elements: sumstat, index, param, target. It’s not a coincidence that the names of these list elements correspond to function arguments of various diagnostic functions of the <em>abc</em> package. In this way, <em>demografr</em> pre-generates the information required by those functions (which normally need to be prepared manually), which can then be run without any additional work whatsoever. For instance, we could perform the cross-validation of the model selection using <em>abc</em>’s own functionality like this.</p>
<p>First, because model-selection requires, well, multiple models to select from, we will unpack information from all the three models:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">parts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">abcX</span>, <span class="va">abcY</span>, <span class="va">abcZ</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">parts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">parts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "sumstat" "index"   "param"   "target"</span></span></code></pre>
<p>Then, we simply run <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code> as we normally would, using the information pre-generated by <code><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack()</a></code>:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="co"># cv4postpr() is a function from the abc R package, not demografr!</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Loading required package: abc.data</span></span></code></pre>
<pre><code><span><span class="co">#&gt; Loading required package: nnet</span></span></code></pre>
<pre><code><span><span class="co">#&gt; Loading required package: quantreg</span></span></code></pre>
<pre><code><span><span class="co">#&gt; Loading required package: SparseM</span></span></code></pre>
<pre><code><span><span class="co">#&gt; Loading required package: MASS</span></span></code></pre>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MASS'</span></span></code></pre>
<pre><code><span><span class="co">#&gt; The following object is masked from 'package:slendr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     area</span></span></code></pre>
<pre><code><span><span class="co">#&gt; Loading required package: locfit</span></span></code></pre>
<pre><code><span><span class="co">#&gt; locfit 1.5-9.12   2025-03-05</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv4postpr_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr</a></span><span class="op">(</span>index <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">index</span>, sumstat <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">sumstat</span>, nval <span class="op">=</span> <span class="fl">10</span>, tols <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<p>Unlike the result of <code><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate()</a></code> above, typing out the object <code>cv4postpr_res</code> will not help us, but we can get the overview of the results by applying the function <code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code>:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">cv4postpr_res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Confusion matrix based on 10 samples for each model.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX     10      0      0</span></span>
<span><span class="co">#&gt; modelY      0     10      0</span></span>
<span><span class="co">#&gt; modelZ      0      0     10</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Mean model posterior probabilities (neuralnet)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX 0.9509 0.0114 0.0378</span></span>
<span><span class="co">#&gt; modelY 0.1030 0.8783 0.0187</span></span>
<span><span class="co">#&gt; modelZ 0.0008 0.0271 0.9721</span></span></code></pre>
<p>As you can see, we can replicate the result of <code><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate()</a></code> for model selection by combining the functions <code><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack()</a></code>, <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code> and <code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code>. Of course, using <code><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate()</a></code> allows us to do this all in one go in a straightforward way, but the above hopefully demonstrates that there’s no magic behind <em>demografr</em>’s convenient functionality. It just streamlines the process of ABC diagnostics and avoids unnecessary (and potentially error-prone) data processing.</p>
<p>On a similar note, we can also run <em>abc</em> functions <code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code> (as described in <a href="https://cran.r-project.org/web/packages/abc/vignettes/abcvignette.pdf" class="external-link">this vignette</a>) in exactly the same way. We will focus just on the code, and leave the interpretation to the discussion in the vignette:</p>
<p>First, we will unpack the results of the selected best-fitting model from the appropriate <code>abc</code> object obtained by the <em>demografr</em> function <code><a href="https://rdrr.io/pkg/demografr/man/run_abc.html" class="external-link">run_abc()</a></code> above:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">partsX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack</a></span><span class="op">(</span><span class="va">abcX</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">partsX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">partsX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "sumstat" "index"   "param"   "target"</span></span></code></pre>
<p>Then we use the elements of the parts list that we just created as inputs for <code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code>:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="co"># gfit() is a function from the abc R package, not demografr!</span></span>
<span></span>
<span><span class="va">gfitX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit</a></span><span class="op">(</span>target <span class="op">=</span> <span class="va">partsX</span><span class="op">$</span><span class="va">target</span>, sumstat <span class="op">=</span> <span class="va">partsX</span><span class="op">$</span><span class="va">sumstat</span>, statistic <span class="op">=</span> <span class="va">mean</span>, nb.replicate <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gfitX</span>, main<span class="op">=</span><span class="st">"Histogram under H0 (model X)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/gfitX-1.png" width="640" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gfitX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $pvalue</span></span>
<span><span class="co">#&gt; [1] 0.37</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $s.dist.sim</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   2.936   3.393   3.792   3.925   4.350   7.145 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dist.obs</span></span>
<span><span class="co">#&gt; [1] 4.076438</span></span></code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">partsY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack</a></span><span class="op">(</span><span class="va">abcY</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gfitY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit</a></span><span class="op">(</span>target <span class="op">=</span> <span class="va">partsY</span><span class="op">$</span><span class="va">target</span>, sumstat <span class="op">=</span> <span class="va">partsY</span><span class="op">$</span><span class="va">sumstat</span>, statistic <span class="op">=</span> <span class="va">mean</span>, nb.replicate <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gfitY</span>, main<span class="op">=</span><span class="st">"Histogram under H0 (model Y)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/gfitY-1.png" width="640" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gfitY</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $pvalue</span></span>
<span><span class="co">#&gt; [1] 0.21</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $s.dist.sim</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   2.717   3.099   3.407   3.536   3.818   5.478 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dist.obs</span></span>
<span><span class="co">#&gt; [1] 3.968966</span></span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">partsZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack</a></span><span class="op">(</span><span class="va">abcZ</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gfitZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit</a></span><span class="op">(</span>target <span class="op">=</span> <span class="va">partsZ</span><span class="op">$</span><span class="va">target</span>, sumstat <span class="op">=</span> <span class="va">partsZ</span><span class="op">$</span><span class="va">sumstat</span>, statistic <span class="op">=</span> <span class="va">mean</span>, nb.replicate <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gfitZ</span>, main<span class="op">=</span><span class="st">"Histogram under H0 (model Z)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/gfitZ-1.png" width="640" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gfitZ</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $pvalue</span></span>
<span><span class="co">#&gt; [1] 0.08</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $s.dist.sim</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   2.826   3.331   3.808   3.874   4.353   5.538 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dist.obs</span></span>
<span><span class="co">#&gt; [1] 5.032494</span></span></code></pre>
<p>For completeness, here is how we can run <code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code> on <em>demografr</em> results:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="co"># gfitpca() is a function from the abc R package, not demografr!</span></span>
<span></span>
<span><span class="va">parts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/unpack.html" class="external-link">unpack</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">abcX</span>, <span class="va">abcY</span>, <span class="va">abcZ</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># this needs to be loaded because of a bug in abc dependency handling</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">locfit</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca</a></span><span class="op">(</span>target <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">target</span>, sumstat <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">sumstat</span>, index <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">index</span>, cprob <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>        xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">7</span>, <span class="fl">7</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">6</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/gfitpca-1.png" width="640" style="display: block; margin: auto;"></p>
<p>Based on the PCA, it appears that although model X and model Y are able to generate summary statistics (the areas captured by the colored “envelopes” in 2D space) compatible with the observed values (the cross), modelZ clearly cannot.</p>
</div>
</div>
<div class="section level2">
<h2 id="abc-cross-validation">ABC cross-validation<a class="anchor" aria-label="anchor" href="#abc-cross-validation"></a>
</h2>
<p>Now we are finally at a point at which we can infer the parameters of our model against the data. However, before we do so, we should first check if our ABC setup can even estimate the model parameters at all. As we have seen just above, the <em>abc</em> package provides a ABC cross-validation function <code><a href="https://rdrr.io/pkg/abc/man/cv4abc.html" class="external-link">cv4abc()</a></code> for which <em>demografr</em> implements a convenient interface in the form of the method <code><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate()</a></code>. Above we have seen how to use this method to perform model selection, if given a list of multiple ABC objects as input. Instead, if given a single <em>demografr</em> ABC object (generated by the function <code><a href="https://rdrr.io/pkg/demografr/man/run_abc.html" class="external-link">run_abc()</a></code>), it performs ABC cross-validation which we can use to gauge the accuracy of ABC and the sensitivity of the parameter estimates to the tolerance rate.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv_abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate</a></span><span class="op">(</span><span class="va">abcX</span>, nval <span class="op">=</span> <span class="fl">10</span>, tols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.005</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span>
<span><span class="va">cv_abc_loclinear</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate</a></span><span class="op">(</span><span class="va">abcX</span>, nval <span class="op">=</span> <span class="fl">10</span>, tols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.005</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"loclinear"</span><span class="op">)</span></span>
<span><span class="va">cv_abc_rejection</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/demografr/man/cross_validate.html" class="external-link">cross_validate</a></span><span class="op">(</span><span class="va">abcX</span>, nval <span class="op">=</span> <span class="fl">10</span>, tols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.005</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"rejection"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv_abc</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Prediction error based on a cross-validation sample of 10</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;               T_1         T_2         T_3          gf        Ne_A        Ne_B</span></span>
<span><span class="co">#&gt; 0.005 0.121701891 0.027779218 0.041039449 0.403815575 0.019174000 0.011942228</span></span>
<span><span class="co">#&gt; 0.01  0.093360651 0.099771004 0.068940505 0.112454770 0.011526444 0.021758669</span></span>
<span><span class="co">#&gt; 0.05  0.059804111 0.129336365 0.094746038 0.138724577 0.005246124 0.060144963</span></span>
<span><span class="co">#&gt;              Ne_C        Ne_D</span></span>
<span><span class="co">#&gt; 0.005 0.132595327 0.256991467</span></span>
<span><span class="co">#&gt; 0.01  0.065244745 0.254472759</span></span>
<span><span class="co">#&gt; 0.05  0.094275057 0.343854602</span></span></code></pre>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cv_abc</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/cv_abc-1.png" width="640" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-2.png" width="640" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-3.png" width="640" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-4.png" width="640" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-5.png" width="640" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-6.png" width="640" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-7.png" width="640" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-8.png" width="640" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="parameter-inference">Parameter inference<a class="anchor" aria-label="anchor" href="#parameter-inference"></a>
</h2>
<p>Now we can finally estimate the model parameters from the model which seems to explain data the best – <code>modelX</code>. Because we simulated the data from a (hidden) true, known model, we will also visualize the true parameter values as vertical dashed lines to compare their values against the posterior distributions of all parameters.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abcX</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; abc::abc(target = observed, param = data$parameters, sumstat = simulated, </span></span>
<span><span class="co">#&gt;     tol = 0.01, method = "neuralnet")</span></span>
<span><span class="co">#&gt; Method:</span></span>
<span><span class="co">#&gt; Non-linear regression via neural networks</span></span>
<span><span class="co">#&gt; with correction for heteroscedasticity</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Parameters:</span></span>
<span><span class="co">#&gt; T_1, T_2, T_3, gf, Ne_A, Ne_B, Ne_C, Ne_D</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Statistics:</span></span>
<span><span class="co">#&gt; diversity_popA, diversity_popB, diversity_popC, diversity_popD, divergence_popA_popB, divergence_popA_popC, divergence_popA_popD, divergence_popB_popC, divergence_popB_popD, divergence_popC_popD, f4_popA_popB_popC_popD</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Total number of simulations 10000 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of accepted simulations:  100</span></span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/extract_summary.html" class="external-link">extract_summary</a></span><span class="op">(</span><span class="va">abcX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                             T_1      T_2      T_3          gf      Ne_A</span></span>
<span><span class="co">#&gt; Min.:                  1446.663 2610.313 6272.322 -0.03842706  972.3869</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  1636.117 4526.712 7305.081 -0.02827445 1207.2348</span></span>
<span><span class="co">#&gt; Weighted Median:       2199.323 5776.129 8381.534  0.13226343 2071.7811</span></span>
<span><span class="co">#&gt; Weighted Mean:         2191.807 5697.829 8426.604  0.15050399 2124.0798</span></span>
<span><span class="co">#&gt; Weighted Mode:         2254.975 5890.870 8251.013  0.11367521 1938.0251</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 2706.516 6162.290 9313.692  0.38492659 3109.1082</span></span>
<span><span class="co">#&gt; Max.:                  3113.943 6775.681 9776.249  0.50275726 3759.6823</span></span>
<span><span class="co">#&gt;                             Ne_B       Ne_C      Ne_D</span></span>
<span><span class="co">#&gt; Min.:                   288.5839  -561.0157 -854.5686</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:   381.9132  3436.7953 -241.2636</span></span>
<span><span class="co">#&gt; Weighted Median:        847.6697  7084.6832 2643.6891</span></span>
<span><span class="co">#&gt; Weighted Mean:          880.3263  6782.9282 2756.6931</span></span>
<span><span class="co">#&gt; Weighted Mode:          743.1117  7415.5397 2368.0101</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 1465.8134  9924.0470 5218.7693</span></span>
<span><span class="co">#&gt; Max.:                  1682.5610 11657.3563 7068.6169</span></span></code></pre>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_posterior.html" class="external-link">plot_posterior</a></span><span class="op">(</span><span class="va">abcX</span>, param <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">6000</span>, <span class="fl">8000</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/posterior_Tsplit-1.png" width="640" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_posterior.html" class="external-link">plot_posterior</a></span><span class="op">(</span><span class="va">abcX</span>, param <span class="op">=</span> <span class="st">"Ne"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">800</span>, <span class="fl">9000</span>, <span class="fl">4000</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/posterior_Ne-1.png" width="640" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_posterior.html" class="external-link">plot_posterior</a></span><span class="op">(</span><span class="va">abcX</span>, param <span class="op">=</span> <span class="st">"gf"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.1</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/posterior_gf-1.png" width="640" style="display: block; margin: auto;"></p>
<p>Not bad, considering how few simulations we’ve thrown at the inference in order to save computational resources in this small vignette! We can see that the modes of the inferred posterior distributions of the model parameters (the colorful densities plotted above) line up pretty much perfectly with the true values (dashed vertical lines).</p>
<p>To make this concluding point even clearer (and also to demonstrate the importance of the various validation procedures described above, such as the posterior predictive checks!) let’s say that we weren’t careful enough with model selection and simply assumed that the structure of <code>modelZ</code> is the most appropriate model for our data and used it for inference of the split time parameters instead. <strong>This is complete non-sense and we show this purely for demonstration purposes!</strong></p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/demografr/man/plot_posterior.html" class="external-link">plot_posterior</a></span><span class="op">(</span><span class="va">abcZ</span>, param <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">6000</span>, <span class="fl">8000</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/posteriorZ_Tsplit-1.png" width="640" style="display: block; margin: auto;"></p>
<p>We can clearly see that posterior distributions inferred by ABC for this particular model completely failed to capture the true values.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a232fffe35f6bc68735023bb34d1fee2',
    indexName: 'bodkan',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
