<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Diagnostics, model selection, troubleshooting • demografr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Diagnostics, model selection, troubleshooting">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">demografr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-question-circle"></span> Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/vignette-01-basics.html">A basic ABC workflow</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-02-priors.html">Specifying prior distributions</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-03-grids.html">Support for grid simulations</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-04-parallelization.html">Parallelization options</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-05-custom.html">Custom inference using SLiM or Python</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-06-diagnostics.html">Model selection and other diagnostics</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bodkan/demografr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Diagnostics, model selection, troubleshooting</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/demografr/blob/HEAD/vignettes/vignette-06-diagnostics.Rmd" class="external-link"><code>vignettes/vignette-06-diagnostics.Rmd</code></a></small>
      <div class="d-none name"><code>vignette-06-diagnostics.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/demografr" class="external-link">demografr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/init_env.html" class="external-link">init_env</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; The interface to all required Python modules has been activated.</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SEED</span> <span class="op">&lt;-</span> <span class="fl">42</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span></span></code></pre></div>
<p>⚠️⚠️⚠️</p>
<p><strong>The <em>demografr</em> R package is still under active
development!</strong></p>
<p>⚠️⚠️⚠️</p>
<p>Let’s return to our <a href="vignette-01-basics.html">first</a>
example. However, this time, imagine that we don’t really know which of
the three following phylogenetic relationships is the one that captures
the features of our data best, perhaps with different sources of
evidence being consistent with each one of them so we want to consider
all three in our modeling (as always, this is supposed to be a
hypothetical example for demonstration purposes). These models differ
both in terms of tree topology of the population relationships as well
as in the gene flow, and we want to perform <em>model
selection</em>.</p>
<p><img src="vignette-06-diagnostics_files/figure-html/ape_tree_modelX-1.png" class="r-plt" width="300" style="display: block; margin: auto;"></p>
<p>For completeness, here is again our observed data which was, in
reality, generated by a hidden evolutionary process whose parameters
we’re trying to infer by doing ABC:</p>
<ol style="list-style-type: decimal">
<li>Nucleotide diversity in each population:</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_diversity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/basics_diversity.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_diversity</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   set    diversity</span></span>
<span><span class="co">#&gt; 1   A 8.030512e-05</span></span>
<span><span class="co">#&gt; 2   B 3.288576e-05</span></span>
<span><span class="co">#&gt; 3   C 1.013804e-04</span></span>
<span><span class="co">#&gt; 4   D 8.910909e-05</span></span></code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Pairwise divergence d_X_Y between populations X and Y:</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_divergence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/basics_divergence.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_divergence</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   x y   divergence</span></span>
<span><span class="co">#&gt; 1 A B 0.0002378613</span></span>
<span><span class="co">#&gt; 2 A C 0.0002375761</span></span>
<span><span class="co">#&gt; 3 A D 0.0002379385</span></span>
<span><span class="co">#&gt; 4 B C 0.0001088217</span></span>
<span><span class="co">#&gt; 5 B D 0.0001157056</span></span>
<span><span class="co">#&gt; 6 C D 0.0001100633</span></span></code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Value of the following
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>4</mn></msub><annotation encoding="application/x-tex">f_4</annotation></semantics></math>-statistic:</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed_f4</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"examples/basics_f4.tsv"</span>, package <span class="op">=</span> <span class="st">"demografr"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">observed_f4</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   W X Y Z            f4</span></span>
<span><span class="co">#&gt; 1 A B C D -3.262146e-06</span></span></code></pre>
<p>We will again bind them into a list:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  diversity  <span class="op">=</span> <span class="va">observed_diversity</span>,</span>
<span>  divergence <span class="op">=</span> <span class="va">observed_divergence</span>,</span>
<span>  f4 <span class="op">=</span> <span class="va">observed_f4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="three-competing-models">Three competing models<a class="anchor" aria-label="anchor" href="#three-competing-models"></a>
</h3>
<p>First, in order to perform model selection, we need to specify models
themselves. We do this by defining three separate <em>slendr</em>
functions, each of them encoding the three population relationships from
the diagrams above. Note that we’re not trying to infer the time of gene
flow, for simplicity, but we do fix the gene flow event to different
times (again, reflected in the alternative model sketches above), just
to make the models stand out from one another:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">modelX</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ne_A</span>, <span class="va">Ne_B</span>, <span class="va">Ne_C</span>, <span class="va">Ne_D</span>, <span class="va">T_1</span>, <span class="va">T_2</span>, <span class="va">T_3</span>, <span class="va">gf</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"A"</span>, time <span class="op">=</span> <span class="fl">1</span>,   N <span class="op">=</span> <span class="va">Ne_A</span><span class="op">)</span></span>
<span>  <span class="va">B</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"B"</span>, time <span class="op">=</span> <span class="va">T_1</span>, N <span class="op">=</span> <span class="va">Ne_B</span>, parent <span class="op">=</span> <span class="va">A</span><span class="op">)</span></span>
<span>  <span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"C"</span>, time <span class="op">=</span> <span class="va">T_2</span>, N <span class="op">=</span> <span class="va">Ne_C</span>, parent <span class="op">=</span> <span class="va">B</span><span class="op">)</span></span>
<span>  <span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"D"</span>, time <span class="op">=</span> <span class="va">T_3</span>, N <span class="op">=</span> <span class="va">Ne_D</span>, parent <span class="op">=</span> <span class="va">C</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">B</span>, to <span class="op">=</span> <span class="va">C</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="va">gf</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A</span>, <span class="va">B</span>, <span class="va">C</span>, <span class="va">D</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>    generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    direction <span class="op">=</span> <span class="st">"forward"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/schedule_sampling.html" class="external-link">schedule_sampling</a></span><span class="op">(</span></span>
<span>    <span class="va">model</span>, times <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">B</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">C</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">D</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">modelY</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ne_A</span>, <span class="va">Ne_B</span>, <span class="va">Ne_C</span>, <span class="va">Ne_D</span>, <span class="va">T_1</span>, <span class="va">T_2</span>, <span class="va">T_3</span>, <span class="va">gf</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"A"</span>, time <span class="op">=</span> <span class="fl">1</span>,   N <span class="op">=</span> <span class="va">Ne_A</span><span class="op">)</span></span>
<span>  <span class="va">B</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"B"</span>, time <span class="op">=</span> <span class="va">T_1</span>, N <span class="op">=</span> <span class="va">Ne_B</span>, parent <span class="op">=</span> <span class="va">A</span><span class="op">)</span></span>
<span>  <span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"C"</span>, time <span class="op">=</span> <span class="va">T_2</span>, N <span class="op">=</span> <span class="va">Ne_C</span>, parent <span class="op">=</span> <span class="va">A</span><span class="op">)</span></span>
<span>  <span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"D"</span>, time <span class="op">=</span> <span class="va">T_3</span>, N <span class="op">=</span> <span class="va">Ne_D</span>, parent <span class="op">=</span> <span class="va">A</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">A</span>, to <span class="op">=</span> <span class="va">B</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="va">gf</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A</span>, <span class="va">B</span>, <span class="va">C</span>, <span class="va">D</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>    generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    direction <span class="op">=</span> <span class="st">"forward"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/schedule_sampling.html" class="external-link">schedule_sampling</a></span><span class="op">(</span></span>
<span>    <span class="va">model</span>, times <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">B</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">C</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">D</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">modelZ</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ne_A</span>, <span class="va">Ne_B</span>, <span class="va">Ne_C</span>, <span class="va">Ne_D</span>, <span class="va">T_1</span>, <span class="va">T_2</span>, <span class="va">T_3</span>, <span class="va">gf</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"A"</span>, time <span class="op">=</span> <span class="fl">1</span>,   N <span class="op">=</span> <span class="va">Ne_A</span><span class="op">)</span></span>
<span>  <span class="va">B</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"B"</span>, time <span class="op">=</span> <span class="va">T_1</span>, N <span class="op">=</span> <span class="va">Ne_B</span>, parent <span class="op">=</span> <span class="va">A</span><span class="op">)</span></span>
<span>  <span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"C"</span>, time <span class="op">=</span> <span class="va">T_2</span>, N <span class="op">=</span> <span class="va">Ne_C</span>, parent <span class="op">=</span> <span class="va">A</span><span class="op">)</span></span>
<span>  <span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"D"</span>, time <span class="op">=</span> <span class="va">T_3</span>, N <span class="op">=</span> <span class="va">Ne_D</span>, parent <span class="op">=</span> <span class="va">C</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">C</span>, to <span class="op">=</span> <span class="va">D</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="va">gf</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A</span>, <span class="va">B</span>, <span class="va">C</span>, <span class="va">D</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>    generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    direction <span class="op">=</span> <span class="st">"forward"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/schedule_sampling.html" class="external-link">schedule_sampling</a></span><span class="op">(</span></span>
<span>    <span class="va">model</span>, times <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">B</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">C</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">D</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>    strict <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">model</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now, let’s <a href="vignette-02-priors.html">specify priors</a> using
<em>demografr</em>’s <a href="vignette-02-priors.html#prior-parameter-templates">templating
syntax</a>. This saves us a bit of typing, making the prior definition
code a bit more consise and easier to read:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">priors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">Ne...</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">T_1</span>   <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1</span>,    <span class="fl">4000</span><span class="op">)</span>,</span>
<span>  <span class="va">T_2</span>   <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">9000</span><span class="op">)</span>,</span>
<span>  <span class="va">T_3</span>   <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fl">10000</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="va">gf</span>    <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Let’s also put together a list of tree-sequence summary functions and
observed summary statistics:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compute_diversity</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_diversity.html" class="external-link">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">compute_divergence</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_divergence.html" class="external-link">ts_divergence</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">compute_f4</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span>  <span class="va">A</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"A"</span><span class="op">]</span>; <span class="va">B</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"B"</span><span class="op">]</span></span>
<span>  <span class="va">C</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"C"</span><span class="op">]</span>; <span class="va">D</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="st">"D"</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_f4ratio.html" class="external-link">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">A</span>, <span class="va">B</span>, <span class="va">C</span>, <span class="va">D</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  diversity <span class="op">=</span> <span class="va">compute_diversity</span>,</span>
<span>  divergence <span class="op">=</span> <span class="va">compute_divergence</span>,</span>
<span>  f4 <span class="op">=</span> <span class="va">compute_f4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Let’s validate the ABC setup of all three models – this is an
important check that the <em>slendr</em> model functions are defined
correctly (we set <code>quiet = TRUE</code> to surpress writing out the
full log output):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">modelX</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">modelY</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/validate_abc.html">validate_abc</a></span><span class="op">(</span><span class="va">modelZ</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<p>With that out of the way, we can proceed with generating simulated
data for inference using all three models. What we’ll do is perform
three runs and save them into appropriately named variables
<code>dataX</code>, <code>dataY</code>, and <code>dataZ</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataX</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span><span class="va">modelX</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                      sequence_length <span class="op">=</span> <span class="fl">10e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataY</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span><span class="va">modelY</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                      sequence_length <span class="op">=</span> <span class="fl">10e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_abc.html">simulate_abc</a></span><span class="op">(</span><span class="va">modelZ</span>, <span class="va">priors</span>, <span class="va">functions</span>, <span class="va">observed</span>, iterations <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                      sequence_length <span class="op">=</span> <span class="fl">10e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<p><strong>The total runtime for the ABC simulations was 1 hours 37
minutes 46 seconds parallelized across 96 CPUs.</strong></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abcX</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_abc.html">run_abc</a></span><span class="op">(</span><span class="va">dataX</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span>
<span><span class="va">abcY</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_abc.html">run_abc</a></span><span class="op">(</span><span class="va">dataY</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span>
<span><span class="va">abcZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_abc.html">run_abc</a></span><span class="op">(</span><span class="va">dataZ</span>, engine <span class="op">=</span> <span class="st">"abc"</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cross-validation">Cross-validation<a class="anchor" aria-label="anchor" href="#cross-validation"></a>
</h2>
<p>Before doing model selection, it’s important to perform
cross-validation to answer the question whether our ABC setup can even
distinguish between the competing models. Without having the power to do
this, trying to select the best model wouldn’t make much sense.</p>
<p>This can be done using <em>demografr</em>’s
<code><a href="../reference/cross_validate.html">cross_validate()</a></code> function which is built around
<em>abc</em>’s own function <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code>. We will not go
into too much detail, as this function simply calls
<code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code> under the hood, passing to it all specified
function arguments on behalf of a user to avoid unnecessary manual data
munging (creating the <code>index</code> vector, merging the summary
statistics into <code>sumstat</code> parameter, etc.). For more details,
read section “Model selection” in the <a href="https://cran.r-project.org/package=abc/vignettes/abcvignette.pdf" class="external-link">vignette</a>
of the <em>abc</em> R package.</p>
<p>The one difference between the two functions is that
<code><a href="../reference/cross_validate.html">cross_validate()</a></code> removes the need to prepare character
indices and bind together summary statistic matrices from different
models—given that <em>demografr</em>’s ABC output objects track all this
information along in their internals, this is redundant, and you can
perform cross-validation of different ABC models simply by calling
this:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">abcX</span>, <span class="va">abcY</span>, <span class="va">abcZ</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cv_selection</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cross_validate.html">cross_validate</a></span><span class="op">(</span><span class="va">models</span>, nval <span class="op">=</span> <span class="fl">100</span>, tol <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<p>If we print out the result, we get a quick summary of the resulting
confusion matrices and other diagnostic information:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv_selection</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Confusion matrix based on 100 samples for each model.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX     95      0      5</span></span>
<span><span class="co">#&gt; modelY      0     98      2</span></span>
<span><span class="co">#&gt; modelZ      0      1     99</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Mean model posterior probabilities (neuralnet)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX 0.8878 0.0597 0.0525</span></span>
<span><span class="co">#&gt; modelY 0.0467 0.9336 0.0197</span></span>
<span><span class="co">#&gt; modelZ 0.0097 0.0237 0.9667</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $conf.matrix</span></span>
<span><span class="co">#&gt; $conf.matrix$tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX     95      0      5</span></span>
<span><span class="co">#&gt; modelY      0     98      2</span></span>
<span><span class="co">#&gt; modelZ      0      1     99</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $probs</span></span>
<span><span class="co">#&gt; $probs$tol0.01</span></span>
<span><span class="co">#&gt;            modelX     modelY     modelZ</span></span>
<span><span class="co">#&gt; modelX 0.88777233 0.05968226 0.05254541</span></span>
<span><span class="co">#&gt; modelY 0.04670368 0.93357416 0.01972216</span></span>
<span><span class="co">#&gt; modelZ 0.00966135 0.02366910 0.96666955</span></span></code></pre>
<p>Similarly, you can use the <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> function to visualize
the result. This function, yet again, internally calls <em>abc</em>’s
own plotting method internall, with a bonus option to save a figure to a
PDF right from the <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> call (useful when working on a
remote server):</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cv_selection</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/plot_cv-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>Because we have three models, each of the three barplots shows how
often were summary statistics sampled from each model classified as
likely coming from one of the three models. In other words, with
absolutely perfect classification, each barplot would show just one of
the three colors. If a barplot (results for one model) shows multiple
colors, this means that some fraction of simulated statistics from that
model was incorrectly classified as another model. Again, for more
detail on interpretation, caveats, and best practices, please consult
the <em>abc</em> R package <a href="https://cran.r-project.org/package=abc/vignettes/abcvignette.pdf" class="external-link">vignette</a>
and a relevant statistical textbook.</p>
<p>The confusion matrices and the visualization all suggest that ABC can
distinguish between the three models very well. For instance,
<em>modelX</em> has been classified correctly in:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv_selection</span><span class="op">$</span><span class="va">estim</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="op">{</span> <span class="va">.</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">==</span> <span class="st">"modelX"</span><span class="op">]</span> <span class="op">}</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">==</span> <span class="va">.</span> <span class="op">}</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">sum</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] 95</span></span></code></pre>
<p>simulations out of the total of 100 cross-validation simulations,
with an overall misclassification rate for all models of only:</p>
<pre><code><span><span class="co">#&gt; [1] "2.7%"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="model-selection">Model selection<a class="anchor" aria-label="anchor" href="#model-selection"></a>
</h2>
<p>Armed with confidence in the ability of ABC to correctly identify the
correct model based on simulated data, we can proceed to selection of
the best model for our empirical data set. This can be done with the
function <code><a href="../reference/select_model.html">select_model()</a></code> which is <em>demografr</em>’s
convenience wrapper around <em>abc</em>’s own function
<code>postpr</code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">abcX</span>, <span class="va">abcY</span>, <span class="va">abcZ</span><span class="op">)</span></span>
<span></span>
<span><span class="va">modsel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_model.html">select_model</a></span><span class="op">(</span><span class="va">models</span>, tol <span class="op">=</span> <span class="fl">0.03</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<p>We can make a decision on the model selection by inspecting the
<code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code> of the produced result:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">modsel</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Call: </span></span>
<span><span class="co">#&gt; abc::postpr(target = parts$target, index = parts$index, sumstat = parts$sumstat, </span></span>
<span><span class="co">#&gt;     tol = tol, method = "neuralnet")</span></span>
<span><span class="co">#&gt; Data:</span></span>
<span><span class="co">#&gt;  postpr.out$values (900 posterior samples)</span></span>
<span><span class="co">#&gt; Models a priori:</span></span>
<span><span class="co">#&gt;  modelX, modelY, modelZ</span></span>
<span><span class="co">#&gt; Models a posteriori:</span></span>
<span><span class="co">#&gt;  modelX, modelY, modelZ</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Proportion of accepted simulations (rejection):</span></span>
<span><span class="co">#&gt; modelX modelY modelZ </span></span>
<span><span class="co">#&gt; 0.8711 0.0056 0.1233 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Bayes factors:</span></span>
<span><span class="co">#&gt;          modelX   modelY   modelZ</span></span>
<span><span class="co">#&gt; modelX   1.0000 156.8000   7.0631</span></span>
<span><span class="co">#&gt; modelY   0.0064   1.0000   0.0450</span></span>
<span><span class="co">#&gt; modelZ   0.1416  22.2000   1.0000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Posterior model probabilities (neuralnet):</span></span>
<span><span class="co">#&gt; modelX modelY modelZ </span></span>
<span><span class="co">#&gt;      1      0      0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Bayes factors:</span></span>
<span><span class="co">#&gt;             modelX      modelY      modelZ</span></span>
<span><span class="co">#&gt; modelX      1.0000 332871.6946 524021.4981</span></span>
<span><span class="co">#&gt; modelY      0.0000      1.0000      1.5742</span></span>
<span><span class="co">#&gt; modelZ      0.0000      0.6352      1.0000</span></span></code></pre>
<p>As we can see, <code>modelX</code> shows the highest rate of
acceptance among all simulations. Specifically, we see that the
proportion of acceptance is:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%.1f%%"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu">quiet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">modsel</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">rejection</span><span class="op">$</span><span class="va">Prob</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "87.1%"</span></span></code></pre>
<p>Similarly, looking at the Bayes factors, we see that the “modelX”
is</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%.1f"</span>, <span class="fu">quiet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">modsel</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">rejection</span><span class="op">$</span><span class="va">Bayes</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "156.8"</span></span></code></pre>
<p>times more likely than “modelY”, and</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%.1f"</span>, <span class="fu">quiet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">modsel</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">rejection</span><span class="op">$</span><span class="va">Bayes</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "7.1"</span></span></code></pre>
<p>times more likely than “modelZ”.</p>
<p>When the posterior probabilities are computed using a more elaborate
neural network method (again, see more details in the <em>abc</em> <a href="https://cran.r-project.org/package=abc/vignettes/abcvignette.pdf" class="external-link">vignette</a>),
we find that the probability of “modelX” is even more overwhelming. In
fact, the analysis shows that it has a 100% probabiliy of being the
correct model to explain our data.</p>
<p><strong>But how accurate are these conclusions?</strong> Well, if we
take a peek at the <em>slendr</em> model which was <a href="https://github.com/bodkan/demografr/blob/main/vignettes/vignette-01-basics.Rmd#L36-L4" class="external-link">internally
used</a> to generate the “observed” summary statistics, we see that the
data was indeed simulated by code nearly identical to the one shown
above as <code>modelX</code>. When plotted, the true model looks like
this:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"A"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"B"</span>, time <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="fl">800</span>, parent <span class="op">=</span> <span class="va">A</span><span class="op">)</span></span>
<span><span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"C"</span>, time <span class="op">=</span> <span class="fl">6000</span>, N <span class="op">=</span> <span class="fl">9000</span>, parent <span class="op">=</span> <span class="va">B</span><span class="op">)</span></span>
<span><span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/population.html" class="external-link">population</a></span><span class="op">(</span><span class="st">"D"</span>, time <span class="op">=</span> <span class="fl">8000</span>, N <span class="op">=</span> <span class="fl">4000</span>, parent <span class="op">=</span> <span class="va">C</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/gene_flow.html" class="external-link">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">B</span>, to <span class="op">=</span> <span class="va">C</span>, start <span class="op">=</span> <span class="fl">9000</span>, end <span class="op">=</span> <span class="fl">9301</span>, rate <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">example_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/compile_model.html" class="external-link">compile_model</a></span><span class="op">(</span></span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A</span>, <span class="va">B</span>, <span class="va">C</span>, <span class="va">D</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>  generation_time <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  simulation_length <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/slendr/man/plot_model.html" class="external-link">plot_model</a></span><span class="op">(</span><span class="va">example_model</span>, proportions <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/true_model-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>It appears that we can be quite confident in which of the three
models is most compatible with the data. However, before we move on to
inferring parameters of the model, we should check whether the best
selected model can indeed capture the most important features of the
data, which is in case of ABC represented by our summary statistics.</p>
<p>In order to do this, <em>demografr</em> provides a function call
<code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code> which performs posterior predictive check. Below,
we we also look at how to use functions <code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code> and
<code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code> implemented by the <em>abc</em> R package
itself.</p>
<p>Let’s start with the posterior predictive checks.</p>
</div>
<div class="section level2">
<h2 id="posterior-predictive-checks">Posterior predictive checks<a class="anchor" aria-label="anchor" href="#posterior-predictive-checks"></a>
</h2>
<p>Link:
<code>http://www.stat.columbia.edu/~gelman/research/published/A6n41.pdf</code></p>
<p>It is one thing to Simply speaking, when fitting a model and
estimating its parameters using ABC, we rely on a set of summary
statistics, which we assume capture the most important features of the
true model represented by the data we gathered. Performing posterior
predictive checks involves estimating the posterior distributions of the
parameters of interest from the data (i.e. doing ABC inference) –
something we’ve done above on our <code>modelX</code>,
<code>modelY</code>, and <code>modelZ</code> – and then sampling
parameters from these distributions and generating distributions of
simulated summary statistics from those estimated models. Then, we plot
the distributions of each of our <em>simulated</em> summary statistics
together with a vertical line representing the values of the
<em>observed</em> statistics. Naturally, only if the model can generate
statistics compatible with the values observed in real data can we
regard the model as sensible and usable for inference.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">predX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">abcX</span>, samples <span class="op">=</span> <span class="fl">1000</span>, posterior <span class="op">=</span> <span class="st">"unadj"</span><span class="op">)</span></span>
<span><span class="va">predY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">abcY</span>, samples <span class="op">=</span> <span class="fl">1000</span>, posterior <span class="op">=</span> <span class="st">"unadj"</span><span class="op">)</span></span>
<span><span class="va">predZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">abcZ</span>, samples <span class="op">=</span> <span class="fl">1000</span>, posterior <span class="op">=</span> <span class="st">"unadj"</span><span class="op">)</span></span></code></pre></div>
<p>Looking at the posterior predictive plots from the <em>model X</em>,
we can see that the observed statistics (vertical dashed lines) fall
very well within the distributions of statistics simulated from the
posterior parameters of the model. This suggests that are model
(simplified that it must be given the more complex reality), captures
the empoirical statistics quite well:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predX-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $divergence</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predX-2.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $f4</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predX-3.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>On the other hand, looking at the <em>model Z</em>, for instance, we
see that although the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mn>4</mn></msub><annotation encoding="application/x-tex">f_4</annotation></semantics></math>
gene-flow detection statistic fits rather nicely, the pairwise
divergences between populations are completely from the statistics
simulated from posterior distributions. This strongly suggests that the
model wasn’t able to capture the features of the data.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predY</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predY-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $divergence</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predY-2.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $f4</span></span></code></pre>
<p><img src="vignette-06-diagnostics_files/figure-html/predY-3.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>We can compare the divergences between models directly to see which
model gives the most reasonable posterior predictive check results more
clearly:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predX</span>, <span class="st">"divergence"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predY</span>, <span class="st">"divergence"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predZ</span>, <span class="st">"divergence"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/pred_divergence-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>We can see that posterior predictive checks for nucleotide diversity
are a little less clear as there are no extremely dramatic outliers as
it’s the case for divergence above:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predX</span>, <span class="st">"diversity"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predY</span>, <span class="st">"diversity"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predZ</span>, <span class="st">"diversity"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/pred_diversity-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predX</span>, <span class="st">"f4"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predY</span>, <span class="st">"f4"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_prediction.html">plot_prediction</a></span><span class="op">(</span><span class="va">predZ</span>, <span class="st">"f4"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/pred_f4-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>If more customization or a more detailed analyses is needed, you can
also extract data from the posterior simulations themselves:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_prediction.html">extract_prediction</a></span><span class="op">(</span><span class="va">predX</span>, <span class="st">"diversity"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $diversity</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4,000 × 12</span></span></span>
<span><span class="co">#&gt;      rep   T_1   T_2   T_3     gf  Ne_A  Ne_B  Ne_C  Ne_D summary  stat    value</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1 <span style="text-decoration: underline;">1</span>218. <span style="text-decoration: underline;">6</span>392. <span style="text-decoration: underline;">9</span>831. 0.062<span style="text-decoration: underline;">7</span> <span style="text-decoration: underline;">1</span>573.  987. <span style="text-decoration: underline;">5</span>493. <span style="text-decoration: underline;">5</span>879. diversi… dive… 5.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     1 <span style="text-decoration: underline;">1</span>218. <span style="text-decoration: underline;">6</span>392. <span style="text-decoration: underline;">9</span>831. 0.062<span style="text-decoration: underline;">7</span> <span style="text-decoration: underline;">1</span>573.  987. <span style="text-decoration: underline;">5</span>493. <span style="text-decoration: underline;">5</span>879. diversi… dive… 3.98<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     1 <span style="text-decoration: underline;">1</span>218. <span style="text-decoration: underline;">6</span>392. <span style="text-decoration: underline;">9</span>831. 0.062<span style="text-decoration: underline;">7</span> <span style="text-decoration: underline;">1</span>573.  987. <span style="text-decoration: underline;">5</span>493. <span style="text-decoration: underline;">5</span>879. diversi… dive… 9.35<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     1 <span style="text-decoration: underline;">1</span>218. <span style="text-decoration: underline;">6</span>392. <span style="text-decoration: underline;">9</span>831. 0.062<span style="text-decoration: underline;">7</span> <span style="text-decoration: underline;">1</span>573.  987. <span style="text-decoration: underline;">5</span>493. <span style="text-decoration: underline;">5</span>879. diversi… dive… 9.48<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     2  962. <span style="text-decoration: underline;">5</span>053. <span style="text-decoration: underline;">6</span>577. 0.074<span style="text-decoration: underline;">7</span> <span style="text-decoration: underline;">1</span>105. <span style="text-decoration: underline;">1</span>215. <span style="text-decoration: underline;">3</span>420. <span style="text-decoration: underline;">3</span>951. diversi… dive… 4.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     2  962. <span style="text-decoration: underline;">5</span>053. <span style="text-decoration: underline;">6</span>577. 0.074<span style="text-decoration: underline;">7</span> <span style="text-decoration: underline;">1</span>105. <span style="text-decoration: underline;">1</span>215. <span style="text-decoration: underline;">3</span>420. <span style="text-decoration: underline;">3</span>951. diversi… dive… 4.98<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     2  962. <span style="text-decoration: underline;">5</span>053. <span style="text-decoration: underline;">6</span>577. 0.074<span style="text-decoration: underline;">7</span> <span style="text-decoration: underline;">1</span>105. <span style="text-decoration: underline;">1</span>215. <span style="text-decoration: underline;">3</span>420. <span style="text-decoration: underline;">3</span>951. diversi… dive… 9.83<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     2  962. <span style="text-decoration: underline;">5</span>053. <span style="text-decoration: underline;">6</span>577. 0.074<span style="text-decoration: underline;">7</span> <span style="text-decoration: underline;">1</span>105. <span style="text-decoration: underline;">1</span>215. <span style="text-decoration: underline;">3</span>420. <span style="text-decoration: underline;">3</span>951. diversi… dive… 1.02<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     3  345. <span style="text-decoration: underline;">4</span>826. <span style="text-decoration: underline;">9</span>959. 0.294   473.  679. <span style="text-decoration: underline;">7</span>349. <span style="text-decoration: underline;">6</span>138. diversi… dive… 1.62<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>     3  345. <span style="text-decoration: underline;">4</span>826. <span style="text-decoration: underline;">9</span>959. 0.294   473.  679. <span style="text-decoration: underline;">7</span>349. <span style="text-decoration: underline;">6</span>138. diversi… dive… 2.53<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3,990 more rows</span></span></span></code></pre>
<p>However, note that <code>extract_prediction</code> is a convenience
function which only unnests, reformats, and renames the list-columns
with summary statistics stored in the <code>predictions</code> data
frame above. You can do all the processing based on the
<code>predictions</code> object above.</p>
<div class="section level3">
<h3 id="posterior-predictive-checks-using-new-statistics">Posterior predictive checks using new statistics<a class="anchor" aria-label="anchor" href="#posterior-predictive-checks-using-new-statistics"></a>
</h3>
<p>In situations in which you might be interested in performing a
posterior predictive check using a different set of statistics rather
than the summary statistics used in the posterior inference in the first
phase of your workflow (see, for instance, <a href="https://stats.stackexchange.com/a/125576" class="external-link">this</a> discussion and
associated references for more detail on why this might or might not be
a valuable thing to do), you can use the optional argument
<code>functions =</code> of the function <code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code>. Just
like any other instances of this argument in functions such as
<code><a href="../reference/simulate_abc.html">simulate_abc()</a></code>, this argument specifies a named list of
summary functions. You can then compare the posterior predictions of
these new summary statistics. Of course, as discussed in the link above,
this might be easier said than done in practice because when we do ABC,
we should be <em>already</em> using the best summary statistics we can,
but it’s worth mentioning this possibility here for cases in which this
might be relevant.</p>
<p>For instance, let’s say that our new summary statistics are AFS
vectors in each of our four populations, which can be computed by the
following tree-sequence summary statistics in R:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">afs_functions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  afs_A <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span> <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span><span class="op">[</span><span class="st">"A"</span><span class="op">]</span>; <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_afs.html" class="external-link">ts_afs</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span> <span class="op">}</span>,</span>
<span>  afs_B <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span> <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span><span class="op">[</span><span class="st">"B"</span><span class="op">]</span>; <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_afs.html" class="external-link">ts_afs</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span> <span class="op">}</span>,</span>
<span>  afs_C <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span> <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span><span class="op">[</span><span class="st">"C"</span><span class="op">]</span>; <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_afs.html" class="external-link">ts_afs</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span> <span class="op">}</span>,</span>
<span>  afs_D <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span> <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_names.html" class="external-link">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span><span class="op">[</span><span class="st">"D"</span><span class="op">]</span>; <span class="fu"><a href="https://rdrr.io/pkg/slendr/man/ts_afs.html" class="external-link">ts_afs</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="va">samples</span><span class="op">)</span> <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Again, it’s always useful to sanity check our functions definitions,
which we do it by computing these statistics from a single simulated
tree sequence (we make it tiny to make things faster—the point is to
test whether our new summary statistics functions even run the way we
want them to):</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_model.html">simulate_model</a></span><span class="op">(</span><span class="va">modelX</span>, <span class="va">priors</span>, sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">0</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarise_data.html">summarise_data</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">afs_functions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $afs_A</span></span>
<span><span class="co">#&gt;   [1] 2357  272   94   33   39   28   84    6   20    3   56   10    4    0    4</span></span>
<span><span class="co">#&gt;  [16]   24    4    0    3    0    0    0    0    0    3   35    0    0    0   17</span></span>
<span><span class="co">#&gt;  [31]    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [46]    0    0    0    0    0    8    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [61]    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [76]    0    0    0    0   55    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [91]    0    0    0    0   23    0    0    0    0    0   34</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $afs_B</span></span>
<span><span class="co">#&gt;   [1] 2530   93   84   37    8   26    6   11   24    6   19    9    1    2    0</span></span>
<span><span class="co">#&gt;  [16]    1    0    0    0   23  117    0    0    0    0    0    0    0    0  103</span></span>
<span><span class="co">#&gt;  [31]   17    0   18    0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [46]    0    0    0    0    0    0   47    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [61]    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [76]    0    0    0    0    0   25    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [91]    0    0    0    0    0    0    0    0    0    0    9</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $afs_C</span></span>
<span><span class="co">#&gt;   [1] 2061  245  199   28   38  164   20  148   15   43    0    0    0    0    5</span></span>
<span><span class="co">#&gt;  [16]    5    0    0    0    2    0    0    2   17   10   15    0    0    0    0</span></span>
<span><span class="co">#&gt;  [31]   14   92    0    0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [46]    0    0    0   43    0    0    0    0   16    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [61]    0    0   25    0    0    0    0    0    0    9    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [76]    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [91]    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $afs_D</span></span>
<span><span class="co">#&gt;   [1] 2174  255  135  106   50   62   70    8   22    9    0    0   71    2    9</span></span>
<span><span class="co">#&gt;  [16]    7    0    8   10    0    6   24    0   11    0    0   60    0    0    0</span></span>
<span><span class="co">#&gt;  [31]    2    0    5    0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [46]    0    0    0    0    0    0    0    6    0    0    0   25    0    0    0</span></span>
<span><span class="co">#&gt;  [61]    0    0    0    0    0    0    0    0    5    0    0    0    0    0   40</span></span>
<span><span class="co">#&gt;  [76]    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [91]    0    0    0    0    0    0    0    0    0    0   34</span></span></code></pre>
<p>Now we can finally get the posterior predictive distributions of the
new summary statistics (per-population AFS vectors), which are again
stored as <a href="https://tidyr.tidyverse.org/articles/nest.html" class="external-link">nested
list-columns</a>:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we simulate only 10 samples from the posterior distribution, but this is only</span></span>
<span><span class="co"># to save computational time!</span></span>
<span><span class="va">pred_afs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">abcX</span>, samples <span class="op">=</span> <span class="fl">10</span>, posterior <span class="op">=</span> <span class="st">"unadj"</span>, functions <span class="op">=</span> <span class="va">afs_functions</span><span class="op">)</span></span></code></pre></div>
<p>This means that each of those <code>afs_...</code> list columns
contains an AFS for the corresponding population for each given
replicate:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_afs</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 13</span></span></span>
<span><span class="co">#&gt;      rep    T_1   T_2   T_3     gf  Ne_A  Ne_B  Ne_C  Ne_D afs_A  afs_B  afs_C </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1 <span style="text-decoration: underline;">3</span>050.  <span style="text-decoration: underline;">7</span>577. <span style="text-decoration: underline;">8</span>936. 0.066<span style="text-decoration: underline;">3</span> <span style="text-decoration: underline;">2</span>365. <span style="text-decoration: underline;">1</span>222. <span style="text-decoration: underline;">3</span>197. <span style="text-decoration: underline;">2</span>109. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2   40.4 <span style="text-decoration: underline;">7</span>205. <span style="text-decoration: underline;">8</span>641. 0.669  <span style="text-decoration: underline;">1</span>333. <span style="text-decoration: underline;">1</span>531. <span style="text-decoration: underline;">3</span>170. <span style="text-decoration: underline;">6</span>227. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3 <span style="text-decoration: underline;">1</span>538.  <span style="text-decoration: underline;">4</span>675. <span style="text-decoration: underline;">9</span>914. 0.642  <span style="text-decoration: underline;">1</span>456.  664. <span style="text-decoration: underline;">5</span>659. <span style="text-decoration: underline;">4</span>147. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4 <span style="text-decoration: underline;">2</span>373.  <span style="text-decoration: underline;">6</span>711. <span style="text-decoration: underline;">9</span>685. 0.688  <span style="text-decoration: underline;">2</span>019. <span style="text-decoration: underline;">1</span>263. <span style="text-decoration: underline;">9</span>451. <span style="text-decoration: underline;">4</span>505. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5  967.  <span style="text-decoration: underline;">8</span>902. <span style="text-decoration: underline;">9</span>671. 0.886  <span style="text-decoration: underline;">1</span>851. <span style="text-decoration: underline;">2</span>284. <span style="text-decoration: underline;">3</span>933.  589. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6 <span style="text-decoration: underline;">1</span>217.  <span style="text-decoration: underline;">5</span>746. <span style="text-decoration: underline;">5</span>894. 0.240  <span style="text-decoration: underline;">1</span>456.  294. <span style="text-decoration: underline;">6</span>053. <span style="text-decoration: underline;">4</span>875. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7 <span style="text-decoration: underline;">2</span>758.  <span style="text-decoration: underline;">8</span>007. <span style="text-decoration: underline;">9</span>034. 0.524  <span style="text-decoration: underline;">2</span>810. <span style="text-decoration: underline;">1</span>341. <span style="text-decoration: underline;">5</span>109. <span style="text-decoration: underline;">8</span>394. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8 <span style="text-decoration: underline;">3</span>094.  <span style="text-decoration: underline;">7</span>611. <span style="text-decoration: underline;">7</span>852. 0.094<span style="text-decoration: underline;">9</span> <span style="text-decoration: underline;">1</span>701. <span style="text-decoration: underline;">1</span>381. <span style="text-decoration: underline;">7</span>170. <span style="text-decoration: underline;">7</span>504. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9 <span style="text-decoration: underline;">2</span>167.  <span style="text-decoration: underline;">7</span>024. <span style="text-decoration: underline;">9</span>708. 0.261  <span style="text-decoration: underline;">1</span>497. <span style="text-decoration: underline;">1</span>886. <span style="text-decoration: underline;">3</span>156. <span style="text-decoration: underline;">4</span>523. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10 <span style="text-decoration: underline;">1</span>644.  <span style="text-decoration: underline;">5</span>518. <span style="text-decoration: underline;">9</span>061. 0.018<span style="text-decoration: underline;">6</span> <span style="text-decoration: underline;">2</span>032.  864. <span style="text-decoration: underline;">4</span>695. <span style="text-decoration: underline;">1</span>399. <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span>  <span style="color: #949494;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: afs_D &lt;list&gt;</span></span></span></code></pre>
<p>Here’s one of them for the population “A” and the first simulated
posterior sample:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_afs</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">$</span><span class="va">afs_A</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;   [1] 10165  1073   465   309   232   199   191   153   124   126   102   110</span></span>
<span><span class="co">#&gt;  [13]   108   110    79    90    60    62    59    57    62    56    38    47</span></span>
<span><span class="co">#&gt;  [25]    43    36    41    40    28    26    25    33    34    40    30    28</span></span>
<span><span class="co">#&gt;  [37]    24    19    19    33    33    23    18    27    13    20    19    15</span></span>
<span><span class="co">#&gt;  [49]    29    19    14    15    20    13    20    15    11    19    19    28</span></span>
<span><span class="co">#&gt;  [61]    21    12     8     9    13    23     8    16     8    23    11    25</span></span>
<span><span class="co">#&gt;  [73]    13     6    15    17     6    16    18    10    14    17    11     9</span></span>
<span><span class="co">#&gt;  [85]    11    22    11    20    16    15    11    11     7    13    14     3</span></span>
<span><span class="co">#&gt;  [97]     4     2    12    11   320</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="additional-diagnostics-and-backwards-compatibility-with-the-abc-package">Additional diagnostics and backwards compatibility with the
<em>abc</em> package<a class="anchor" aria-label="anchor" href="#additional-diagnostics-and-backwards-compatibility-with-the-abc-package"></a>
</h2>
<p>As mentioned above, the functionality behind
<code><a href="../reference/cross_validate.html">cross_validate()</a></code> is implemented by reformatting the results
of <em>demografr</em> simulation and inference functionality in a form
necessary to run the functions of the <em>abc</em> package
<code><a href="https://rdrr.io/pkg/abc/man/cv4abc.html" class="external-link">cv4abc()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code> behind the scenes – a
task that normally requires a significant amount of potentially
error-prone bookkeeping on part of the user.</p>
<p>In addition to <em>demografr</em>’s own function
<code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code> which implements posterior predictive checks, the
<em>abc</em> package provides additional functionality for evaluating
the quality of the fit of an ABC model to the data, namely
<code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code>. In this section we
demonstrate how this functionality (and potentially other future
<em>abc</em> functions) can be applied on <em>demografr</em> data with
very little additional work.</p>
<p>First, <em>demografr</em> provides a function <code><a href="../reference/unpack.html">unpack()</a></code>
which – when applied on an object generated either by
<code><a href="../reference/simulate_abc.html">simulate_abc()</a></code> or <code><a href="../reference/run_abc.html">run_abc()</a></code> – <em>unpacks</em>
the object encapsulating all information describing the simulation and
inference which produced it, into individual components with names
matching the “low-level” function arguments of all relevant functions of
the <em>anc</em> package. This includes <code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code> and
<code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code> mentioned above, but also functions
<code><a href="https://rdrr.io/pkg/abc/man/cv4abc.html" class="external-link">cv4abc()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code>. Here is how
<code><a href="../reference/unpack.html">unpack()</a></code> works.</p>
<div class="section level3">
<h3 id="unpacking-a-demografr-object-to-low-level-components-used-by-the-abc-package">“Unpacking” a <em>demografr</em> object to low-level components used
by the <em>abc</em> package<a class="anchor" aria-label="anchor" href="#unpacking-a-demografr-object-to-low-level-components-used-by-the-abc-package"></a>
</h3>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">partsX</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unpack.html">unpack</a></span><span class="op">(</span><span class="va">abcX</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">partsX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">partsX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "sumstat" "index"   "param"   "target"</span></span></code></pre>
<p>As we can see, the <code><a href="../reference/unpack.html">unpack()</a></code> function produced a list
object with several elements. It’s not a coincidence that the names of
these list elements correspond to function arguments of various
diagnostic functions of the <em>abc</em> package. In this way,
<em>demografr</em> pre-generates the information required by those
functions (which normally need to be prepared manually), which can then
be run without any additional work whatsoever. For instance, we could
perform the cross-validation of the model selection using <em>abc</em>’s
own functionality like this.</p>
<p>First, because model-selection requires, well, multiple models to
select from, we will unpack information from all the three models:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">parts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unpack.html">unpack</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">abcX</span>, <span class="va">abcY</span>, <span class="va">abcZ</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">parts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">parts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "sumstat" "index"   "param"   "target"</span></span></code></pre>
<p>Then, we simply run <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code> as we normally would,
using the information pre-generated by <code><a href="../reference/unpack.html">unpack()</a></code>:</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="co"># cv4postpr() is a function from the abc R package, not demografr!</span></span>
<span></span>
<span><span class="va">cv4postpr_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr</a></span><span class="op">(</span>index <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">index</span>, sumstat <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">sumstat</span>, nval <span class="op">=</span> <span class="fl">10</span>, tols <span class="op">=</span> <span class="fl">0.01</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span></code></pre></div>
<p>Unlike the result of <code><a href="../reference/cross_validate.html">cross_validate()</a></code> above, typing out
the object <code>cv4postpr_res</code> will not help us, but we can get
the overview of the results by applying the function
<code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code>:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">cv4postpr_res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Confusion matrix based on 10 samples for each model.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX     10      0      0</span></span>
<span><span class="co">#&gt; modelY      0     10      0</span></span>
<span><span class="co">#&gt; modelZ      0      0     10</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Mean model posterior probabilities (neuralnet)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tol0.01</span></span>
<span><span class="co">#&gt;        modelX modelY modelZ</span></span>
<span><span class="co">#&gt; modelX 0.9193 0.0640 0.0167</span></span>
<span><span class="co">#&gt; modelY 0.0294 0.9635 0.0071</span></span>
<span><span class="co">#&gt; modelZ 0.0003 0.0199 0.9799</span></span></code></pre>
<p>As you can see, we can replicate the result of
<code><a href="../reference/cross_validate.html">cross_validate()</a></code> for model selection by combining the
functions <code><a href="../reference/unpack.html">unpack()</a></code>, <code><a href="https://rdrr.io/pkg/abc/man/cv4postpr.html" class="external-link">cv4postpr()</a></code> and
<code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code>. Of course, using <code><a href="../reference/cross_validate.html">cross_validate()</a></code>
allows us to do this all in one go in a straightforward way, but the
above hopefully demonstrates that there’s no magic behind
<em>demografr</em>’s convenient functionality. It just streamlines the
process of ABC diagnostics and avoids unnecessary (and potentially
error-prone) data processing.</p>
<p>On a similar note, we can also run <em>abc</em> functions
<code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code> and <code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code> (as described in <a href="https://CRAN.R-project.org/package=abc/vignettes/abcvignette.pdf" class="external-link">this
vignette</a>) in exactly the same way. We will focus just on the code,
and leave the interpretation to the discussion in the vignette:</p>
<p>First, we will unpack the results of the selected best-fitting model
from the appropriate <code>abc</code> object obtained by the
<em>demografr</em> function <code><a href="../reference/run_abc.html">run_abc()</a></code> above:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">partsX</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unpack.html">unpack</a></span><span class="op">(</span><span class="va">abcX</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">partsX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">partsX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "sumstat" "index"   "param"   "target"</span></span></code></pre>
<p>Then we use the elements of the parts list that we just created as
inputs for <code><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit()</a></code>:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="co"># gfit() is a function from the abc R package, not demografr!</span></span>
<span></span>
<span><span class="va">gfitX</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit</a></span><span class="op">(</span>target <span class="op">=</span> <span class="va">partsX</span><span class="op">$</span><span class="va">target</span>, sumstat <span class="op">=</span> <span class="va">partsX</span><span class="op">$</span><span class="va">sumstat</span>, statistic <span class="op">=</span> <span class="va">mean</span>, nb.replicate <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gfitX</span>, main<span class="op">=</span><span class="st">"Histogram under H0 (model X)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/gfitX-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gfitX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $pvalue</span></span>
<span><span class="co">#&gt; [1] 0.35</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $s.dist.sim</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   2.958   3.462   3.822   3.959   4.384   5.866 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dist.obs</span></span>
<span><span class="co">#&gt; [1] 4.076438</span></span></code></pre>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">partsY</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unpack.html">unpack</a></span><span class="op">(</span><span class="va">abcY</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gfitY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit</a></span><span class="op">(</span>target <span class="op">=</span> <span class="va">partsY</span><span class="op">$</span><span class="va">target</span>, sumstat <span class="op">=</span> <span class="va">partsY</span><span class="op">$</span><span class="va">sumstat</span>, statistic <span class="op">=</span> <span class="va">mean</span>, nb.replicate <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gfitY</span>, main<span class="op">=</span><span class="st">"Histogram under H0 (model Y)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/gfitY-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gfitY</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $pvalue</span></span>
<span><span class="co">#&gt; [1] 0.22</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $s.dist.sim</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   2.703   3.086   3.423   3.532   3.904   4.941 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dist.obs</span></span>
<span><span class="co">#&gt; [1] 3.968966</span></span></code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">partsZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unpack.html">unpack</a></span><span class="op">(</span><span class="va">abcZ</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gfitZ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/abc/man/gfit.html" class="external-link">gfit</a></span><span class="op">(</span>target <span class="op">=</span> <span class="va">partsZ</span><span class="op">$</span><span class="va">target</span>, sumstat <span class="op">=</span> <span class="va">partsZ</span><span class="op">$</span><span class="va">sumstat</span>, statistic <span class="op">=</span> <span class="va">mean</span>, nb.replicate <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gfitZ</span>, main<span class="op">=</span><span class="st">"Histogram under H0 (model Z)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/gfitZ-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gfitZ</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $pvalue</span></span>
<span><span class="co">#&gt; [1] 0.07</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $s.dist.sim</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   2.942   3.268   3.733   3.853   4.384   5.915 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dist.obs</span></span>
<span><span class="co">#&gt; [1] 5.032494</span></span></code></pre>
<p>For completeness, here is how we can run <code><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca()</a></code> on
<em>demografr</em> results:</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">abc</span><span class="op">)</span> <span class="co"># gfitpca() is a function from the abc R package, not demografr!</span></span>
<span></span>
<span><span class="va">parts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unpack.html">unpack</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">abcX</span>, <span class="va">abcY</span>, <span class="va">abcZ</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># this needs to be loaded because of a bug in abc dependency handling</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">locfit</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/abc/man/gfitpca.html" class="external-link">gfitpca</a></span><span class="op">(</span>target <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">target</span>, sumstat <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">sumstat</span>, index <span class="op">=</span> <span class="va">parts</span><span class="op">$</span><span class="va">index</span>, cprob <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>        xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">7</span>, <span class="fl">7</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">6</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/gfitpca-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>Based on the PCA, it appears that although model X and model Y are
able to generate summary statistics (the areas captured by the colored
“envelopes” in 2D space) compatible with the observed values (the
cross), modelZ clearly cannot.</p>
</div>
</div>
<div class="section level2">
<h2 id="abc-cross-validation">ABC cross-validation<a class="anchor" aria-label="anchor" href="#abc-cross-validation"></a>
</h2>
<p>Now we are finally at a point at which we can infer the parameters of
our model against the data. However, before we do so, we should first
check if our ABC setup can even estimate the model parameters at all. As
we have seen just above, the <em>abc</em> package provides a ABC
cross-validation function <code><a href="https://rdrr.io/pkg/abc/man/cv4abc.html" class="external-link">cv4abc()</a></code> for which
<em>demografr</em> implements a convenient interface in the form of the
method <code><a href="../reference/cross_validate.html">cross_validate()</a></code>. Above we have seen how to use this
method to perform model selection, if given a list of multiple ABC
objects as input. Instead, if given a single <em>demografr</em> ABC
object (generated by the function <code><a href="../reference/run_abc.html">run_abc()</a></code>), it performs
ABC cross-validation which we can use to gauge the accuracy of ABC and
the sensitivity of the parameter estimates to the tolerance rate.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv_abc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cross_validate.html">cross_validate</a></span><span class="op">(</span><span class="va">abcX</span>, nval <span class="op">=</span> <span class="fl">10</span>, tols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.005</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"neuralnet"</span><span class="op">)</span></span>
<span><span class="va">cv_abc_loclinear</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cross_validate.html">cross_validate</a></span><span class="op">(</span><span class="va">abcX</span>, nval <span class="op">=</span> <span class="fl">10</span>, tols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.005</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"loclinear"</span><span class="op">)</span></span>
<span><span class="va">cv_abc_rejection</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cross_validate.html">cross_validate</a></span><span class="op">(</span><span class="va">abcX</span>, nval <span class="op">=</span> <span class="fl">10</span>, tols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.005</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"rejection"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cv_abc</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Prediction error based on a cross-validation sample of 10</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;               T_1         T_2         T_3          gf        Ne_A        Ne_B</span></span>
<span><span class="co">#&gt; 0.005 0.066127677 0.059804264 0.127565665 0.176387818 0.016653242 0.033635941</span></span>
<span><span class="co">#&gt; 0.01  0.076075754 0.028500770 0.055767416 0.238912106 0.009378181 0.021096094</span></span>
<span><span class="co">#&gt; 0.05  0.054353933 0.062892707 0.113332733 0.330300985 0.026141790 0.087732732</span></span>
<span><span class="co">#&gt;              Ne_C        Ne_D</span></span>
<span><span class="co">#&gt; 0.005 0.425600692 0.268586461</span></span>
<span><span class="co">#&gt; 0.01  0.330747722 0.210105302</span></span>
<span><span class="co">#&gt; 0.05  0.296105927 0.171603983</span></span></code></pre>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cv_abc</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/cv_abc-1.png" class="r-plt" width="480" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-2.png" class="r-plt" width="480" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-3.png" class="r-plt" width="480" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-4.png" class="r-plt" width="480" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-5.png" class="r-plt" width="480" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-6.png" class="r-plt" width="480" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-7.png" class="r-plt" width="480" style="display: block; margin: auto;"><img src="vignette-06-diagnostics_files/figure-html/cv_abc-8.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="parameter-inference">Parameter inference<a class="anchor" aria-label="anchor" href="#parameter-inference"></a>
</h2>
<p>Now we can finally estimate the model parameters from the model which
seems to explain data the best – <code>modelX</code>. Because we
simulated the data from a (hidden) true, known model, we will also
visualize the true parameter values as vertical dashed lines to compare
their values against the posterior distributions of all parameters.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">abcX</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; abc::abc(target = observed, param = data$parameters, sumstat = simulated, </span></span>
<span><span class="co">#&gt;     tol = 0.01, method = "neuralnet")</span></span>
<span><span class="co">#&gt; Method:</span></span>
<span><span class="co">#&gt; Non-linear regression via neural networks</span></span>
<span><span class="co">#&gt; with correction for heteroscedasticity</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Parameters:</span></span>
<span><span class="co">#&gt; T_1, T_2, T_3, gf, Ne_A, Ne_B, Ne_C, Ne_D</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Statistics:</span></span>
<span><span class="co">#&gt; diversity_A, diversity_B, diversity_C, diversity_D, divergence_A_B, divergence_A_C, divergence_A_D, divergence_B_C, divergence_B_D, divergence_C_D, f4_A_B_C_D</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Total number of simulations 10000 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of accepted simulations:  100</span></span></code></pre>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_summary.html">extract_summary</a></span><span class="op">(</span><span class="va">abcX</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span>
<span><span class="co">#&gt; Warning in density.default(x, weights = weights): Selecting bandwidth *not*</span></span>
<span><span class="co">#&gt; using 'weights'</span></span></code></pre>
<pre><code><span><span class="co">#&gt;                             T_1      T_2      T_3          gf      Ne_A</span></span>
<span><span class="co">#&gt; Min.:                  1462.618 3030.968 6617.586 -0.10991103  679.2167</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:  1517.251 4319.252 7180.513 -0.07031312  890.9708</span></span>
<span><span class="co">#&gt; Weighted Median:       2148.144 5806.244 8460.379  0.12172614 2083.5104</span></span>
<span><span class="co">#&gt; Weighted Mean:         2128.986 5730.303 8432.278  0.14010521 2093.5091</span></span>
<span><span class="co">#&gt; Weighted Mode:         2217.808 5909.163 8424.770  0.10064241 1955.3505</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 2718.882 6404.353 9286.605  0.41296552 3102.2050</span></span>
<span><span class="co">#&gt; Max.:                  3030.757 6599.029 9532.209  0.64394792 4300.4994</span></span>
<span><span class="co">#&gt;                             Ne_B      Ne_C      Ne_D</span></span>
<span><span class="co">#&gt; Min.:                   274.4861  1251.380 -2265.807</span></span>
<span><span class="co">#&gt; Weighted 2.5 % Perc.:   338.0007  3580.354 -1039.344</span></span>
<span><span class="co">#&gt; Weighted Median:        856.8090  6934.996  2494.683</span></span>
<span><span class="co">#&gt; Weighted Mean:          881.8393  6803.139  2724.711</span></span>
<span><span class="co">#&gt; Weighted Mode:          857.0188  7408.730  2077.874</span></span>
<span><span class="co">#&gt; Weighted 97.5 % Perc.: 1482.2222 11379.856  6714.054</span></span>
<span><span class="co">#&gt; Max.:                  1578.6224 12019.281  7547.764</span></span></code></pre>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abcX</span>, param <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">6000</span>, <span class="fl">8000</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/posterior_Tsplit-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abcX</span>, param <span class="op">=</span> <span class="st">"Ne"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">800</span>, <span class="fl">9000</span>, <span class="fl">4000</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/posterior_Ne-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abcX</span>, param <span class="op">=</span> <span class="st">"gf"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.1</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/posterior_gf-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>Not bad, considering how few simulations we’ve thrown at the
inference in order to save computational resources in this small
vignette! We can see that the modes of the inferred posterior
distributions of the model parameters (the colorful densities plotted
above) line up pretty much perfectly with the true values (dashed
vertical lines).</p>
<p>To make this concluding point even clearer (and also to demonstrate
the importance of the various validation procedures described above,
such as the posterior predictive checks!) let’s say that we weren’t
careful enough with model selection and simply assumed that the
structure of <code>modelZ</code> is the most appropriate model for our
data and used it for inference of the split time parameters instead.
<strong>This is complete non-sense and we show this purely for
demonstration purposes!</strong></p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_posterior.html">plot_posterior</a></span><span class="op">(</span><span class="va">abcZ</span>, param <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">6000</span>, <span class="fl">8000</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-06-diagnostics_files/figure-html/posteriorZ_Tsplit-1.png" class="r-plt" width="480" style="display: block; margin: auto;"></p>
<p>We can clearly see that posterior distributions inferred by ABC for
this particular model completely failed to capture the true values.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
